// @flue/runtime@0.9.2/dist/cloudflare/index.mjs
import { d as getModelBinding, i as createSandboxSessionEnv, v as CLOUDFLARE_AI_BINDING_API } from "../sandbox-C2jFycj2.mjs";
import { c as invokeWorkflowAttached, m as InvalidRequestError, s as invokeDirectAttached } from "../handle-agent-C2C75lpq.mjs";
import { t as abortErrorFor } from "../abort-B-q3xjj8.mjs";
import { n as decodeRunCursor, r as encodeRunCursor, t as MAX_LIST_LIMIT } from "../run-registry-CwGOiYQh.mjs";
import { n as parseAgentWebSocketMessage, r as parseWorkflowWebSocketMessage, t as createWebSocketErrorMessage } from "../websocket-protocol-BSIpcV6C.mjs";
import { createAssistantMessageEventStream, parseStreamingJson } from "@earendil-works/pi-ai";
import { AsyncLocalStorage } from "node:async_hooks";
import { DurableObject } from "cloudflare:workers";
import { convertMessages } from "@earendil-works/pi-ai/openai-completions";
//#region src/cloudflare/virtual-sandbox.ts
function getVirtualSandbox(bucket, _options) {
	if (bucket === void 0) throw new Error("[flue] getVirtualSandbox() has been removed. Flue's default in-memory sandbox is already what you wanted — omit the `sandbox` field from createAgent(...) (or pass `false`) and you get it.");
	throw new Error("[flue] getVirtualSandbox(bucket) has been removed because R2 is not a live mounted agent filesystem. Run `flue add @cloudflare/shell`, import `getShellSandbox` and `hydrateFromBucket` from your generated `connectors/cloudflare-shell` file, hydrate the workspace, then pass `sandbox: getShellSandbox(...)` to createAgent().");
}
//#endregion
//#region src/cloudflare/cf-sandbox.ts
/** Wraps a @cloudflare/sandbox instance (from getSandbox()) into SessionEnv. */
async function cfSandboxToSessionEnv(sandbox, cwd = "/workspace") {
	return createSandboxSessionEnv({
		async readFile(path) {
			return (await sandbox.readFile(path)).content;
		},
		async readFileBuffer(path) {
			const file = await sandbox.readFile(path, { encoding: "base64" });
			const binary = atob(file.content);
			const bytes = new Uint8Array(binary.length);
			for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
			return bytes;
		},
		async writeFile(path, content) {
			if (typeof content === "string") await sandbox.writeFile(path, content);
			else {
				let binary = "";
				for (const byte of content) binary += String.fromCharCode(byte);
				const b64 = btoa(binary);
				await sandbox.writeFile(path, b64, { encoding: "base64" });
			}
		},
		async stat(path) {
			const result = await sandbox.exec(`stat -c '{"size":%s,"isDir":%F}' '${path.replace(/'/g, "'\\''")}'`);
			if (!result.success) throw new Error(`stat failed for ${path}: ${result.stderr}`);
			try {
				const raw = result.stdout.trim();
				const sizeMatch = raw.match(/"size":(\d+)/);
				const isDir = raw.includes("directory");
				return {
					isFile: !isDir,
					isDirectory: isDir,
					isSymbolicLink: false,
					size: parseInt(sizeMatch?.[1] ?? "0", 10),
					mtime: /* @__PURE__ */ new Date()
				};
			} catch {
				throw new Error(`Failed to parse stat output for ${path}: ${result.stdout}`);
			}
		},
		async readdir(path) {
			const result = await sandbox.exec(`ls -1 '${path.replace(/'/g, "'\\''")}'`);
			if (!result.success) throw new Error(`readdir failed for ${path}: ${result.stderr}`);
			return result.stdout.trim().split("\n").filter((s) => s.length > 0);
		},
		async exists(path) {
			return (await sandbox.exists(path)).exists;
		},
		async mkdir(path, opts) {
			await sandbox.mkdir(path, opts);
		},
		async rm(path, opts) {
			if (opts?.recursive || opts?.force) {
				const flags = [opts.recursive ? "-r" : "", opts.force ? "-f" : ""].filter(Boolean).join("");
				await sandbox.exec(`rm ${flags} '${path.replace(/'/g, "'\\''")}'`);
			} else await sandbox.deleteFile(path);
		},
		async exec(command, execOpts) {
			const externalSignal = execOpts?.signal;
			if (externalSignal?.aborted) throw abortErrorFor(externalSignal);
			const timeoutMs = typeof execOpts?.timeout === "number" ? execOpts.timeout * 1e3 : void 0;
			const result = await sandbox.exec(command, {
				cwd: execOpts?.cwd,
				env: execOpts?.env,
				timeout: timeoutMs
			});
			if (externalSignal?.aborted) throw abortErrorFor(externalSignal);
			return {
				stdout: result.stdout ?? "",
				stderr: result.stderr ?? "",
				exitCode: result.exitCode ?? (result.success ? 0 : 1)
			};
		}
	}, cwd);
}
//#endregion
//#region src/cloudflare/context.ts
/**
* Cloudflare environment context injection.
*
* Durable Objects are single-threaded, but async executions can still interleave
* at await points. AsyncLocalStorage keeps Cloudflare runtime primitives scoped
* to the request/fiber that invoked them instead of sharing a module global.
*/
const contextStorage = new AsyncLocalStorage();
function runWithCloudflareContext(ctx, fn) {
	return contextStorage.run(ctx, fn);
}
function getCloudflareContext() {
	const ctx = contextStorage.getStore();
	if (!ctx) throw new Error("[flue:cloudflare] Not running in a Cloudflare context. This function can only be called inside a Cloudflare Worker or Durable Object.");
	return ctx;
}
function getDurableObjectIdentity() {
	const ctx = getCloudflareContext();
	if (!ctx.durableObjectIdentity) throw new Error("[flue:cloudflare] Durable Object identity is not available in this Cloudflare context.");
	return ctx.durableObjectIdentity;
}
//#endregion
//#region src/cloudflare/registry-ops.ts
/** SQL-backed workflow-run registry operations and private REST router. */
function createRegistryOps(sql) {
	ensureRegistryTables(sql);
	return new SqlRegistryOps(sql);
}
async function handleRegistryRequest(ops, request) {
	const url = new URL(request.url);
	const segments = url.pathname.split("/").filter(Boolean);
	try {
		if (request.method === "GET" && segments[0] === "pointers" && segments.length === 2) {
			const runId = decodeURIComponent(segments[1] ?? "");
			if (!runId) return new Response("Missing runId.", { status: 404 });
			const pointer = ops.lookupRun(runId);
			if (!pointer) return new Response(null, { status: 404 });
			return jsonResponse(pointer);
		}
		if (request.method === "POST" && segments[0] === "pointers" && segments[2] === "start" && segments.length === 3) {
			const runId = decodeURIComponent(segments[1] ?? "");
			if (!runId) return new Response("Missing runId.", { status: 404 });
			const body = await request.json();
			ops.recordRunStart({
				...body,
				runId
			});
			return new Response(null, { status: 204 });
		}
		if (request.method === "POST" && segments[0] === "pointers" && segments[2] === "end" && segments.length === 3) {
			const runId = decodeURIComponent(segments[1] ?? "");
			if (!runId) return new Response("Missing runId.", { status: 404 });
			const body = await request.json();
			ops.recordRunEnd({
				...body,
				runId
			});
			return new Response(null, { status: 204 });
		}
		if (request.method === "GET" && segments[0] === "pointers" && segments.length === 1) return jsonResponse(ops.listRuns(parseListRunsOpts(url.searchParams)));
		return new Response(`Unknown registry endpoint: ${request.method} ${url.pathname}`, { status: 404 });
	} catch (err) {
		const message = err instanceof Error ? err.message : String(err);
		return new Response(JSON.stringify({ error: message }), {
			status: 500,
			headers: jsonHeaders()
		});
	}
}
var SqlRegistryOps = class {
	sql;
	constructor(sql) {
		this.sql = sql;
	}
	recordRunStart(input) {
		if (input.owner.instanceId !== input.runId) throw new Error("[flue] Workflow run owners must use the same instanceId as the pointer runId.");
		this.sql.exec(`INSERT OR IGNORE INTO flue_registry_runs
			 (run_id, owner_kind, instance_id, workflow_name, status, started_at, ended_at, duration_ms, is_error)
			 VALUES (?, 'workflow', ?, ?, 'active', ?, NULL, NULL, NULL)`, input.runId, input.owner.instanceId, input.owner.workflowName, input.startedAt);
	}
	recordRunEnd(input) {
		this.sql.exec(`UPDATE flue_registry_runs
			 SET status = ?, ended_at = ?, duration_ms = ?, is_error = ?
			 WHERE run_id = ? AND owner_kind = 'workflow'`, input.isError ? "errored" : "completed", input.endedAt, input.durationMs, input.isError ? 1 : 0, input.runId);
	}
	lookupRun(runId) {
		const row = this.sql.exec("SELECT * FROM flue_registry_runs WHERE run_id = ? AND owner_kind = 'workflow'", runId).toArray()[0];
		return row ? rowToRunPointer(row) : null;
	}
	listRuns(opts) {
		const limit = clampLimit(opts.limit);
		const cursor = decodeRunCursor(opts.cursor);
		const wheres = ["owner_kind = 'workflow'"];
		const bindings = [];
		if (opts.status) {
			wheres.push("status = ?");
			bindings.push(opts.status);
		}
		if (opts.workflowName) {
			wheres.push("workflow_name = ?");
			bindings.push(opts.workflowName);
		}
		if (cursor) {
			wheres.push("(started_at < ? OR (started_at = ? AND run_id < ?))");
			bindings.push(cursor.startedAt, cursor.startedAt, cursor.runId);
		}
		const rows = this.sql.exec(`SELECT * FROM flue_registry_runs WHERE ${wheres.join(" AND ")}
			 ORDER BY started_at DESC, run_id DESC LIMIT ?`, ...bindings, limit + 1).toArray();
		const hasMore = rows.length > limit;
		const page = (hasMore ? rows.slice(0, limit) : rows).map(rowToRunPointer);
		const last = page.at(-1);
		return {
			runs: page,
			nextCursor: hasMore && last ? encodeRunCursor(last) : void 0
		};
	}
};
function ensureRegistryTables(sql) {
	sql.exec(`CREATE TABLE IF NOT EXISTS flue_registry_runs (
		 run_id TEXT PRIMARY KEY,
		 owner_kind TEXT NOT NULL,
		 instance_id TEXT,
		 workflow_name TEXT,
		 status TEXT NOT NULL,
		 started_at TEXT NOT NULL,
		 ended_at TEXT,
		 duration_ms INTEGER,
		 is_error INTEGER
		)`);
	sql.exec("CREATE INDEX IF NOT EXISTS flue_registry_status_started_idx ON flue_registry_runs (status, started_at DESC)");
	sql.exec("CREATE INDEX IF NOT EXISTS flue_registry_workflow_started_idx ON flue_registry_runs (owner_kind, workflow_name, started_at DESC)");
}
function rowToRunPointer(row) {
	const runId = String(row.run_id);
	return {
		runId,
		owner: {
			kind: "workflow",
			workflowName: String(row.workflow_name),
			instanceId: String(row.instance_id ?? runId)
		},
		status: String(row.status),
		startedAt: String(row.started_at),
		endedAt: typeof row.ended_at === "string" ? row.ended_at : void 0,
		durationMs: typeof row.duration_ms === "number" ? row.duration_ms : void 0,
		isError: row.is_error === null || row.is_error === void 0 ? void 0 : Boolean(row.is_error)
	};
}
function parseListRunsOpts(params) {
	const opts = {};
	const status = params.get("status");
	if (status === "active" || status === "completed" || status === "errored") opts.status = status;
	const workflow = params.get("workflow");
	if (workflow) opts.workflowName = workflow;
	const limit = params.get("limit");
	if (limit !== null) opts.limit = Number.parseInt(limit, 10);
	const cursor = params.get("cursor");
	if (cursor) opts.cursor = cursor;
	return opts;
}
function clampLimit(limit) {
	if (!limit || !Number.isFinite(limit) || limit <= 0) return 100;
	return Math.min(limit, MAX_LIST_LIMIT);
}
function jsonHeaders() {
	return { "content-type": "application/json" };
}
function jsonResponse(body) {
	return new Response(JSON.stringify(body), { headers: jsonHeaders() });
}
//#endregion
//#region src/cloudflare/registry-do.ts
/** Singleton run-pointer index for the Cloudflare target. */
var FlueRegistry = class extends DurableObject {
	ops;
	constructor(state, env) {
		super(state, env);
		this.ops = createRegistryOps(state.storage.sql);
	}
	async fetch(request) {
		return handleRegistryRequest(this.ops, request);
	}
};
//#endregion
//#region src/cloudflare/run-registry.ts
function createCloudflareRunRegistry(namespace) {
	if (!namespace) return void 0;
	return new CloudflareRunRegistry(namespace);
}
const FLUE_REGISTRY_INSTANCE_NAME = "default";
const SYNTHETIC_BASE = "https://flue-registry.local";
var CloudflareRunRegistry = class {
	namespace;
	constructor(namespace) {
		this.namespace = namespace;
	}
	async recordRunStart(input) {
		const { runId, ...body } = input;
		await this.callExpectingNoContent(`/pointers/${encodeURIComponent(runId)}/start`, "POST", body);
	}
	async recordRunEnd(input) {
		const { runId, ...body } = input;
		await this.callExpectingNoContent(`/pointers/${encodeURIComponent(runId)}/end`, "POST", body);
	}
	async lookupRun(runId) {
		const response = await this.stub().fetch(new Request(`${SYNTHETIC_BASE}/pointers/${encodeURIComponent(runId)}`, { method: "GET" }));
		if (response.status === 404) return null;
		if (!response.ok) throw new Error(`[flue] FlueRegistry lookupRun(${runId}) failed: ${response.status} ${await response.text()}`);
		return await response.json();
	}
	async listRuns(opts = {}) {
		const params = new URLSearchParams();
		if (opts.status) params.set("status", opts.status);
		if (opts.workflowName) params.set("workflow", opts.workflowName);
		if (opts.limit !== void 0) params.set("limit", String(opts.limit));
		if (opts.cursor) params.set("cursor", opts.cursor);
		const qs = params.toString();
		const response = await this.stub().fetch(new Request(`${SYNTHETIC_BASE}/pointers${qs ? `?${qs}` : ""}`, { method: "GET" }));
		if (!response.ok) throw new Error(`[flue] FlueRegistry listRuns failed: ${response.status} ${await response.text()}`);
		return await response.json();
	}
	stub() {
		return this.namespace.get(this.namespace.idFromName(FLUE_REGISTRY_INSTANCE_NAME));
	}
	async callExpectingNoContent(path, method, body) {
		const response = await this.stub().fetch(new Request(`${SYNTHETIC_BASE}${path}`, {
			method,
			headers: { "content-type": "application/json" },
			body: JSON.stringify(body)
		}));
		if (!response.ok) throw new Error(`[flue] FlueRegistry ${method} ${path} failed: ${response.status} ${await response.text()}`);
	}
};
//#endregion
//#region src/cloudflare/session-store.ts
function copySessions(sessions) {
	return Object.assign(Object.create(null), sessions);
}
function store() {
	return {
		async save(id, data) {
			const { agentInstance } = getCloudflareContext();
			const sessions = copySessions(agentInstance.state?.sessions);
			sessions[id] = data;
			agentInstance.setState({
				...agentInstance.state,
				sessions
			});
		},
		async load(id) {
			const { agentInstance } = getCloudflareContext();
			const sessions = agentInstance.state?.sessions;
			return sessions && Object.hasOwn(sessions, id) ? sessions[id] ?? null : null;
		},
		async delete(id) {
			const { agentInstance } = getCloudflareContext();
			const sessions = copySessions(agentInstance.state?.sessions);
			delete sessions[id];
			agentInstance.setState({
				...agentInstance.state,
				sessions
			});
		}
	};
}
//#endregion
//#region src/cloudflare/websocket.ts
const MAX_MESSAGE_BYTES = 1024 * 1024;
function connectCloudflareAgentWebSocket(connection, options) {
	connection.serializeAttachment({
		version: 1,
		target: "agent",
		name: options.name,
		id: options.id,
		requestUrl: operationRequestUrl(options.requestUrl)
	});
	send(connection, {
		version: 1,
		type: "ready",
		target: "agent",
		name: options.name,
		instanceId: options.id
	});
}
function connectCloudflareWorkflowWebSocket(connection, options) {
	connection.serializeAttachment({
		version: 1,
		target: "workflow",
		name: options.name,
		runId: options.runId,
		requestUrl: operationRequestUrl(options.requestUrl),
		invoked: false
	});
	send(connection, {
		version: 1,
		type: "ready",
		target: "workflow",
		name: options.name
	});
}
async function messageCloudflareAgentWebSocket(connection, raw, options) {
	if (typeof raw !== "string") {
		sendError(connection, new InvalidRequestError({ reason: "Binary WebSocket messages are not supported." }));
		close(connection, 1003, "Binary messages are not supported");
		return;
	}
	if (messageBytes(raw) > MAX_MESSAGE_BYTES) {
		sendError(connection, new InvalidRequestError({ reason: "WebSocket messages must not exceed 1048576 bytes." }));
		close(connection, 1008, "Message too large");
		return;
	}
	let message;
	try {
		message = parseAgentWebSocketMessage(raw);
	} catch (error) {
		sendError(connection, error);
		return;
	}
	if (message.type === "ping") {
		send(connection, {
			version: 1,
			type: "pong",
			requestId: message.requestId
		});
		return;
	}
	await invokeAgentPrompt(connection, message, options);
}
async function messageCloudflareWorkflowWebSocket(connection, raw, options) {
	if (typeof raw !== "string") {
		sendError(connection, new InvalidRequestError({ reason: "Binary WebSocket messages are not supported." }));
		close(connection, 1003, "Binary messages are not supported");
		return;
	}
	if (messageBytes(raw) > MAX_MESSAGE_BYTES) {
		sendError(connection, new InvalidRequestError({ reason: "WebSocket messages must not exceed 1048576 bytes." }));
		close(connection, 1008, "Message too large");
		return;
	}
	let message;
	try {
		message = parseWorkflowWebSocketMessage(raw);
	} catch (error) {
		sendError(connection, error);
		return;
	}
	const attachment = connection.deserializeAttachment();
	if (!attachment || attachment.target !== "workflow" || attachment.invoked) {
		sendError(connection, new InvalidRequestError({ reason: "Workflow WebSocket connections accept one invocation only." }), message.requestId);
		close(connection, 1008, "Workflow accepts one invocation only");
		return;
	}
	connection.serializeAttachment({
		...attachment,
		invoked: true
	});
	await invokeWorkflow(connection, message, options);
}
async function invokeAgentPrompt(connection, message, options) {
	let didStart = false;
	try {
		await options.beforePrompt?.(message.session ?? "default");
		const result = await invokeDirectAttached({
			agentName: options.name,
			id: options.id,
			payload: {
				message: message.message,
				session: message.session
			},
			request: options.request,
			handler: options.handler,
			createContext: options.createContext,
			runHandler: options.runHandler,
			onEvent: (event) => {
				if (!didStart) {
					didStart = true;
					send(connection, {
						version: 1,
						type: "started",
						requestId: message.requestId
					});
				}
				send(connection, {
					version: 1,
					type: "event",
					requestId: message.requestId,
					event
				});
			},
			emitIdleOnComplete: true
		});
		send(connection, {
			version: 1,
			type: "result",
			requestId: message.requestId,
			result: result ?? null
		});
	} catch (error) {
		sendError(connection, error, message.requestId);
	}
}
async function invokeWorkflow(connection, message, options) {
	let didStart = false;
	const bufferedEvents = [];
	try {
		const invocation = await invokeWorkflowAttached({
			owner: {
				kind: "workflow",
				workflowName: options.name,
				instanceId: options.runId
			},
			id: options.runId,
			runId: options.runId,
			payload: message.payload,
			request: options.request,
			handler: options.handler,
			createContext: options.createContext,
			startWorkflowAdmission: options.startWorkflowAdmission,
			onAdmitted: () => {
				didStart = true;
				send(connection, {
					version: 1,
					type: "started",
					requestId: message.requestId,
					runId: options.runId
				});
				for (const event of bufferedEvents) send(connection, {
					version: 1,
					type: "event",
					requestId: message.requestId,
					runId: options.runId,
					event
				});
			},
			onEvent: (event) => {
				if (!didStart) {
					bufferedEvents.push(event);
					return;
				}
				send(connection, {
					version: 1,
					type: "event",
					requestId: message.requestId,
					runId: options.runId,
					event
				});
			},
			emitIdleOnComplete: true,
			runStore: options.runStore,
			runSubscribers: options.runSubscribers,
			runRegistry: options.runRegistry
		});
		send(connection, {
			version: 1,
			type: "result",
			requestId: message.requestId,
			runId: options.runId,
			result: invocation.result ?? null
		});
		close(connection, 1e3, "Workflow completed");
	} catch (error) {
		sendError(connection, error, message.requestId, options.runId);
		close(connection, 1011, "Workflow failed");
	}
}
function messageBytes(message) {
	return new TextEncoder().encode(message).byteLength;
}
function operationRequestUrl(requestUrl) {
	const url = new URL(requestUrl);
	url.search = "";
	url.hash = "";
	return url.toString();
}
function sendError(connection, error, requestId, runId) {
	send(connection, createWebSocketErrorMessage(error, requestId, runId));
}
function send(connection, message) {
	try {
		connection.send(JSON.stringify(message));
	} catch {
		return;
	}
}
function close(connection, code, reason) {
	try {
		connection.close(code, reason);
	} catch {
		return;
	}
}
//#endregion
//#region src/cloudflare/workers-ai-provider.ts
/**
* Mirrors pi-ai's `detectCompat('cloudflare-workers-ai')`. Hardcoded here
* because `convertMessages` requires a fully-resolved compat object and the
* binding's wire format matches `cloudflare-workers-ai` exactly. Re-mirror
* if pi-ai's detection logic changes upstream.
*/
const WORKERS_AI_COMPAT = {
	supportsStore: false,
	supportsDeveloperRole: false,
	supportsReasoningEffort: true,
	supportsUsageInStreaming: true,
	maxTokensField: "max_completion_tokens",
	requiresToolResultName: false,
	requiresAssistantAfterToolResult: false,
	requiresThinkingAsText: false,
	requiresReasoningContentOnAssistantMessages: false,
	thinkingFormat: "openai",
	openRouterRouting: {},
	vercelGatewayRouting: {},
	zaiToolStream: false,
	supportsStrictMode: true,
	cacheControlFormat: void 0,
	sendSessionAffinityHeaders: true,
	supportsLongCacheRetention: false
};
function convertTools(tools) {
	return tools.map((tool) => ({
		type: "function",
		function: {
			name: tool.name,
			description: tool.description,
			parameters: tool.parameters,
			...WORKERS_AI_COMPAT.supportsStrictMode !== false && { strict: false }
		}
	}));
}
function emptyUsage() {
	return {
		input: 0,
		output: 0,
		cacheRead: 0,
		cacheWrite: 0,
		totalTokens: 0,
		cost: {
			input: 0,
			output: 0,
			cacheRead: 0,
			cacheWrite: 0,
			total: 0
		}
	};
}
function parseChunkUsage(raw) {
	const cacheRead = raw.prompt_tokens_details?.cached_tokens ?? 0;
	const promptTokens = raw.prompt_tokens ?? 0;
	const completionTokens = raw.completion_tokens ?? 0;
	return {
		input: Math.max(0, promptTokens - cacheRead),
		output: completionTokens,
		cacheRead,
		cacheWrite: 0,
		totalTokens: raw.total_tokens ?? promptTokens + completionTokens,
		cost: {
			input: 0,
			output: 0,
			cacheRead: 0,
			cacheWrite: 0,
			total: 0
		}
	};
}
function mapStopReason(reason) {
	switch (reason) {
		case "stop":
		case "eos": return { stopReason: "stop" };
		case "length": return { stopReason: "length" };
		case "tool_calls":
		case "function_call": return { stopReason: "toolUse" };
		case "content_filter": return {
			stopReason: "error",
			errorMessage: "Provider stopped generation: content filter"
		};
		default: return {
			stopReason: "error",
			errorMessage: `Provider finish_reason: ${reason}`
		};
	}
}
async function* iterateSseChunks(body) {
	const reader = body.getReader();
	const decoder = new TextDecoder();
	let buffer = "";
	try {
		while (true) {
			const { done, value } = await reader.read();
			if (done) {
				buffer += decoder.decode();
				if (buffer.trim().length > 0) yield* parseSseEvents(buffer);
				return;
			}
			buffer += decoder.decode(value, { stream: true });
			let boundary = findSseBoundary(buffer);
			while (boundary) {
				const block = buffer.slice(0, boundary.index);
				buffer = buffer.slice(boundary.index + boundary.width);
				yield* parseSseEvents(block);
				boundary = findSseBoundary(buffer);
			}
		}
	} finally {
		try {
			reader.releaseLock();
		} catch {}
	}
}
function findSseBoundary(buffer) {
	const lf = buffer.indexOf("\n\n");
	const crlf = buffer.indexOf("\r\n\r\n");
	if (lf === -1 && crlf === -1) return null;
	if (lf === -1) return {
		index: crlf,
		width: 4
	};
	if (crlf === -1) return {
		index: lf,
		width: 2
	};
	return lf < crlf ? {
		index: lf,
		width: 2
	} : {
		index: crlf,
		width: 4
	};
}
function* parseSseEvents(block) {
	let start = 0;
	while (start <= block.length) {
		const newline = block.indexOf("\n", start);
		const end = newline === -1 ? block.length : newline;
		const lineEnd = end > start && block.charCodeAt(end - 1) === 13 ? end - 1 : end;
		const line = block.slice(start, lineEnd);
		if (line.startsWith("data:")) {
			const data = line.slice(5).trimStart();
			if (data !== "" && data !== "[DONE]") try {
				yield JSON.parse(data);
			} catch {}
		}
		if (newline === -1) return;
		start = newline + 1;
	}
}
function isAbortError(error) {
	return error instanceof Error && (error.name === "AbortError" || /aborted/i.test(error.message));
}
const streamCloudflareWorkersAi = (model, context, options) => {
	const stream = createAssistantMessageEventStream();
	(async () => {
		const output = {
			role: "assistant",
			content: [],
			api: model.api,
			provider: model.provider,
			model: model.id,
			usage: emptyUsage(),
			stopReason: "stop",
			timestamp: Date.now()
		};
		try {
			const ai = resolveBinding(model);
			const payload = {
				messages: convertMessages(model, context, WORKERS_AI_COMPAT),
				stream: true,
				stream_options: { include_usage: true }
			};
			if (context.tools && context.tools.length > 0) payload.tools = convertTools(context.tools);
			if (options?.maxTokens) payload.max_completion_tokens = options.maxTokens;
			if (options?.temperature !== void 0) payload.temperature = options.temperature;
			applyReasoningEffort(payload, model, options?.reasoning);
			const overridden = await options?.onPayload?.(payload, model);
			const finalPayload = overridden === void 0 ? payload : overridden;
			const extraHeaders = {};
			if (options?.sessionId) extraHeaders["x-session-affinity"] = options.sessionId;
			if (options?.headers) Object.assign(extraHeaders, options.headers);
			const gateway = model.gateway;
			const response = await ai.run(model.id, finalPayload, {
				returnRawResponse: true,
				...options?.signal ? { signal: options.signal } : {},
				...Object.keys(extraHeaders).length > 0 ? { extraHeaders } : {},
				...gateway ? { gateway } : {}
			});
			await options?.onResponse?.({
				status: response.status,
				headers: headersToRecord(response.headers)
			}, model);
			if (!response.ok) {
				const errorBody = await safeReadText(response);
				throw new Error(`Cloudflare AI binding returned ${response.status} ${response.statusText}` + (errorBody ? `: ${errorBody}` : ""));
			}
			if (!response.body) throw new Error("Cloudflare AI binding returned empty response body.");
			stream.push({
				type: "start",
				partial: output
			});
			let currentBlock = null;
			const blocks = output.content;
			const indexOf = (block) => block ? blocks.indexOf(block) : -1;
			const finishCurrentBlock = (block) => {
				if (!block) return;
				const contentIndex = indexOf(block);
				if (contentIndex === -1) return;
				if (block.type === "text") stream.push({
					type: "text_end",
					contentIndex,
					content: block.text,
					partial: output
				});
				else if (block.type === "thinking") stream.push({
					type: "thinking_end",
					contentIndex,
					content: block.thinking,
					partial: output
				});
				else if (block.type === "toolCall") {
					block.arguments = parseStreamingJson(block.partialArgs ?? "");
					delete block.partialArgs;
					delete block.streamIndex;
					stream.push({
						type: "toolcall_end",
						contentIndex,
						toolCall: block,
						partial: output
					});
				}
			};
			for await (const rawChunk of iterateSseChunks(response.body)) {
				const chunk = rawChunk;
				if (!chunk || typeof chunk !== "object") continue;
				output.responseId ||= chunk.id;
				if (typeof chunk.model === "string" && chunk.model.length > 0 && chunk.model !== model.id) output.responseModel ||= chunk.model;
				if (chunk.usage) output.usage = parseChunkUsage(chunk.usage);
				const choice = Array.isArray(chunk.choices) ? chunk.choices[0] : void 0;
				if (!choice) continue;
				if (!chunk.usage && choice.usage) output.usage = parseChunkUsage(choice.usage);
				if (choice.finish_reason) {
					const mapped = mapStopReason(choice.finish_reason);
					output.stopReason = mapped.stopReason;
					if (mapped.errorMessage) output.errorMessage = mapped.errorMessage;
				}
				const delta = choice.delta;
				if (!delta) continue;
				if (delta.content !== null && delta.content !== void 0 && delta.content.length > 0) {
					if (!currentBlock || currentBlock.type !== "text") {
						finishCurrentBlock(currentBlock);
						currentBlock = {
							type: "text",
							text: ""
						};
						blocks.push(currentBlock);
						stream.push({
							type: "text_start",
							contentIndex: indexOf(currentBlock),
							partial: output
						});
					}
					currentBlock.text += delta.content;
					stream.push({
						type: "text_delta",
						contentIndex: indexOf(currentBlock),
						delta: delta.content,
						partial: output
					});
				}
				const reasoningDelta = pickReasoning(delta);
				if (reasoningDelta) {
					if (!currentBlock || currentBlock.type !== "thinking") {
						finishCurrentBlock(currentBlock);
						currentBlock = {
							type: "thinking",
							thinking: "",
							thinkingSignature: reasoningDelta.field
						};
						blocks.push(currentBlock);
						stream.push({
							type: "thinking_start",
							contentIndex: indexOf(currentBlock),
							partial: output
						});
					}
					currentBlock.thinking += reasoningDelta.text;
					stream.push({
						type: "thinking_delta",
						contentIndex: indexOf(currentBlock),
						delta: reasoningDelta.text,
						partial: output
					});
				}
				if (delta.tool_calls) for (const toolCall of delta.tool_calls) {
					const streamIndex = typeof toolCall.index === "number" ? toolCall.index : void 0;
					if (!(currentBlock?.type === "toolCall" && (streamIndex !== void 0 && currentBlock.streamIndex === streamIndex || streamIndex === void 0 && !!toolCall.id && currentBlock.id === toolCall.id))) {
						finishCurrentBlock(currentBlock);
						currentBlock = {
							type: "toolCall",
							id: toolCall.id ?? "",
							name: toolCall.function?.name ?? "",
							arguments: {},
							partialArgs: "",
							streamIndex
						};
						blocks.push(currentBlock);
						stream.push({
							type: "toolcall_start",
							contentIndex: indexOf(currentBlock),
							partial: output
						});
					}
					const block = currentBlock?.type === "toolCall" ? currentBlock : null;
					if (block) {
						if (!block.id && toolCall.id) block.id = toolCall.id;
						if (!block.name && toolCall.function?.name) block.name = toolCall.function.name;
						if (block.streamIndex === void 0 && streamIndex !== void 0) block.streamIndex = streamIndex;
						let toolDelta = "";
						if (toolCall.function?.arguments) {
							toolDelta = toolCall.function.arguments;
							block.partialArgs = (block.partialArgs ?? "") + toolDelta;
							block.arguments = parseStreamingJson(block.partialArgs);
						}
						stream.push({
							type: "toolcall_delta",
							contentIndex: indexOf(block),
							delta: toolDelta,
							partial: output
						});
					}
				}
			}
			finishCurrentBlock(currentBlock);
			if (options?.signal?.aborted) throw new Error("Request was aborted");
			if (output.stopReason === "aborted") throw new Error("Request was aborted");
			if (output.stopReason === "error") throw new Error(output.errorMessage ?? "Provider returned an error stop reason");
			stream.push({
				type: "done",
				reason: output.stopReason,
				message: output
			});
			stream.end();
		} catch (error) {
			for (const block of output.content) if (block.type === "toolCall") {
				delete block.partialArgs;
				delete block.streamIndex;
			}
			output.stopReason = options?.signal?.aborted || isAbortError(error) ? "aborted" : "error";
			output.errorMessage = error instanceof Error ? error.message : JSON.stringify(error);
			stream.push({
				type: "error",
				reason: output.stopReason,
				error: output
			});
			stream.end();
		}
	})();
	return stream;
};
/**
* Read the binding extension carried on the resolved Model.
*/
function resolveBinding(model) {
	const ai = getModelBinding(model);
	if (!ai) throw new Error("[flue] Cloudflare AI binding not available. Models prefixed with \"cloudflare/\" require running on the Cloudflare target with `\"ai\": { \"binding\": \"AI\" }` declared in wrangler.jsonc. For URL-based access without the binding, use pi-ai's `cloudflare-workers-ai/...` or `cloudflare-ai-gateway/...` providers (both require Cloudflare API credentials in env vars).");
	return ai;
}
function pickReasoning(delta) {
	for (const field of ["reasoning_content", "reasoning"]) {
		const value = delta[field];
		if (typeof value === "string" && value.length > 0) return {
			field,
			text: value
		};
	}
	return null;
}
function applyReasoningEffort(payload, model, level) {
	if (!model.reasoning || level === void 0) return;
	payload.reasoning_effort = mapReasoningEffort(level);
}
function mapReasoningEffort(level) {
	switch (level) {
		case "minimal":
		case "low": return "low";
		case "medium": return "medium";
		case "high":
		case "xhigh": return "high";
	}
}
function headersToRecord(headers) {
	const out = {};
	headers.forEach((value, key) => {
		out[key] = value;
	});
	return out;
}
async function safeReadText(response) {
	try {
		return await response.text();
	} catch {
		return;
	}
}
/**
* Return the pi-ai `ApiProvider` definition for the Cloudflare AI binding.
*/
function getCloudflareAIBindingApiProvider() {
	return {
		api: CLOUDFLARE_AI_BINDING_API,
		stream: streamCloudflareWorkersAi,
		streamSimple: streamCloudflareWorkersAi
	};
}
//#endregion
//#region src/cloudflare/index.ts
const CLOUDFLARE_SHELL_CONNECTOR_MIGRATION = "Run `flue add @cloudflare/shell`, then import Cloudflare Shell helpers from your generated `connectors/cloudflare-shell` file.";
function getShellSandbox(..._args) {
	throw new Error(`[flue] getShellSandbox() is no longer implemented by @flue/runtime/cloudflare because Cloudflare Shell sandboxes are project-owned. ${CLOUDFLARE_SHELL_CONNECTOR_MIGRATION}`);
}
function getDefaultWorkspace(..._args) {
	throw new Error(`[flue] getDefaultWorkspace() is no longer implemented by @flue/runtime/cloudflare because Cloudflare Shell sandboxes are project-owned. ${CLOUDFLARE_SHELL_CONNECTOR_MIGRATION}`);
}
function hydrateFromBucket(..._args) {
	throw new Error(`[flue] hydrateFromBucket() is no longer implemented by @flue/runtime/cloudflare because hydration belongs to the Cloudflare Shell connector. ${CLOUDFLARE_SHELL_CONNECTOR_MIGRATION}`);
}
//#endregion
export { FlueRegistry, cfSandboxToSessionEnv, connectCloudflareAgentWebSocket, connectCloudflareWorkflowWebSocket, createCloudflareRunRegistry, getCloudflareAIBindingApiProvider, getCloudflareContext, getDefaultWorkspace, getDurableObjectIdentity, getShellSandbox, getVirtualSandbox, hydrateFromBucket, messageCloudflareAgentWebSocket, messageCloudflareWorkflowWebSocket, runWithCloudflareContext, store };
