// remult@3.3.12/esm/src/data-providers/json-entity-indexed-db-data-provider.js
export class JsonEntityIndexedDbStorage {
    dbName;
    storeName;
    constructor(dbName = 'db', storeName = 'jsonStore') {
        this.dbName = dbName;
        this.storeName = storeName;
    }
    supportsRawJson = true;
    //@internal
    db;
    getItem(entityDbName) {
        return this.run('readonly', (s) => s.get(entityDbName));
    }
    setItem(entityDbName, json) {
        return this.run('readwrite', (s) => s.put(json, entityDbName));
    }
    removeItem(entityDbName) {
        return this.run('readwrite', (s) => s.delete(entityDbName));
    }
    clear() {
        return this.run('readwrite', (s) => s.clear());
    }
    //@internal
    async run(mode, op) {
        const attempt = async () => {
            const db = await this.init();
            return new Promise((resolve, reject) => {
                const tx = db.transaction([this.storeName], mode);
                tx.onabort = () => reject(tx.error);
                const request = op(tx.objectStore(this.storeName));
                request.onerror = () => reject(request.error);
                request.onsuccess = () => resolve(request.result ?? null);
            });
        };
        try {
            return await attempt();
        }
        catch (err) {
            this.db = undefined;
            return await attempt();
        }
    }
    //@internal
    async init() {
        if (this.db)
            return this.db;
        this.db = await new Promise((resolve, reject) => {
            const request = indexedDB.open(this.dbName, 1);
            request.onerror = () => reject(request.error);
            request.onupgradeneeded = () => request.result.createObjectStore(this.storeName);
            request.onsuccess = () => {
                const db = request.result;
                const forget = () => {
                    if (this.db === db)
                        this.db = undefined;
                };
                db.onclose = forget;
                db.onversionchange = () => {
                    db.close();
                    forget();
                };
                resolve(db);
            };
        });
        return this.db;
    }
}
