// svelte-persisted-state@1.4.1/dist/index.svelte.js
import { getItem as idbGetItem, setItem as idbSetItem, closeDB } from './indexeddb-storage.js';
function getCookie(name) {
    const match = document.cookie.match(new RegExp('(^| )' + name + '=([^;]+)'));
    return match ? decodeURIComponent(match[2]) : null;
}
function setCookie(name, value, options = {}) {
    const { expireDays = 365, maxAge, path = '/', domain, secure = false, sameSite = 'Lax', httpOnly = false } = options;
    let cookieString = `${name}=${encodeURIComponent(value)}`;
    cookieString += `; path=${path}`;
    // Use max-age if specified, otherwise use expires
    if (maxAge !== undefined) {
        cookieString += `; max-age=${maxAge}`;
    }
    else {
        const expires = new Date(Date.now() + expireDays * 864e5).toUTCString();
        cookieString += `; expires=${expires}`;
    }
    if (domain) {
        cookieString += `; domain=${domain}`;
    }
    if (secure) {
        cookieString += `; secure`;
    }
    cookieString += `; samesite=${sameSite}`;
    if (httpOnly) {
        cookieString += `; httponly`;
    }
    document.cookie = cookieString;
}
function getStorage(type, cookieOptions = {}) {
    if (type === 'local')
        return {
            getItem: (k) => localStorage.getItem(k),
            setItem: (k, v) => localStorage.setItem(k, v)
        };
    if (type === 'session')
        return {
            getItem: (k) => sessionStorage.getItem(k),
            setItem: (k, v) => sessionStorage.setItem(k, v)
        };
    // cookie
    return {
        getItem: getCookie,
        setItem: (k, v) => setCookie(k, v, cookieOptions)
    };
}
export function persistedState(key, initialValue, options = {}) {
    const { storage = 'local', serializer = JSON, syncTabs = true, cookieExpireDays, cookieOptions = {}, onWriteError = console.error, onParseError = console.error, beforeRead = (v) => v, beforeWrite = (v) => v } = options;
    // Handle backward compatibility with cookieExpireDays
    const finalCookieOptions = {
        ...cookieOptions,
        ...(cookieExpireDays !== undefined && { expireDays: cookieExpireDays })
    };
    const browser = typeof window !== 'undefined' && typeof document !== 'undefined';
    const storageArea = browser ? getStorage(storage, finalCookieOptions) : null;
    let storedValue;
    try {
        const item = storageArea?.getItem(key);
        storedValue = item ? beforeRead(serializer.parse(item)) : initialValue;
    }
    catch (error) {
        onParseError(error);
        storedValue = initialValue;
    }
    let state = $state(storedValue);
    let lastSerialized = storageArea?.getItem(key) ?? null;
    function updateStorage(value) {
        try {
            const valueToStore = beforeWrite(value);
            const serialized = serializer.stringify(valueToStore);
            if (serialized !== lastSerialized) {
                storageArea?.setItem(key, serialized);
                lastSerialized = serialized;
            }
        }
        catch (error) {
            onWriteError(error);
        }
    }
    if (syncTabs && typeof window !== 'undefined' && storage === 'local') {
        window.addEventListener('storage', (event) => {
            if (event.key === key && event.storageArea === localStorage) {
                try {
                    const newValue = event.newValue ? serializer.parse(event.newValue) : initialValue;
                    state = beforeRead(newValue);
                }
                catch (error) {
                    onParseError(error);
                }
            }
        });
    }
    $effect.root(() => {
        $effect(() => {
            updateStorage(state);
        });
        return () => { };
    });
    return {
        /**
         * @deprecated Use current to align with Svelte conventions
         */
        get value() {
            return state;
        },
        /**
         * @deprecated Use current to align with Svelte conventions
         */
        set value(newValue) {
            state = newValue;
        },
        get current() {
            return state;
        },
        set current(newValue) {
            state = newValue;
        },
        reset() {
            state = initialValue;
        }
    };
}
export function persistedStateAsync(key, initialValue, options = {}) {
    const { indexedDB: indexedDBOptions = {}, serializer, syncTabs = true, onWriteError = console.error, onParseError = console.error, onHydrated, onHydrationError = console.error, beforeRead = (v) => v, beforeWrite = (v) => v } = options;
    const browser = typeof window !== 'undefined' && typeof indexedDB !== 'undefined';
    let state = $state(initialValue);
    let isLoading = $state(browser);
    let resolveReady;
    let rejectReady;
    const ready = new Promise((resolve, reject) => {
        resolveReady = resolve;
        rejectReady = reject;
    });
    let broadcastChannel = null;
    let skipNextWrite = false;
    async function hydrate() {
        if (!browser) {
            isLoading = false;
            resolveReady(initialValue);
            return;
        }
        try {
            const storedValue = await idbGetItem(key, indexedDBOptions);
            if (storedValue !== null) {
                try {
                    const value = serializer ? serializer.parse(storedValue) : storedValue;
                    state = beforeRead(value);
                }
                catch (error) {
                    onParseError(error);
                    state = initialValue;
                }
            }
            isLoading = false;
            onHydrated?.(state);
            resolveReady(state);
        }
        catch (error) {
            onHydrationError(error);
            isLoading = false;
            rejectReady(error);
        }
    }
    if (browser && syncTabs) {
        const channelName = `svelte-persisted-state:${indexedDBOptions.dbName ?? 'svelte-persisted-state'}`;
        broadcastChannel = new BroadcastChannel(channelName);
        broadcastChannel.onmessage = (event) => {
            if (event.data.key === key) {
                try {
                    skipNextWrite = true;
                    const value = serializer ? serializer.parse(event.data.value) : event.data.value;
                    state = beforeRead(value);
                }
                catch (error) {
                    skipNextWrite = false;
                    onParseError(error);
                }
            }
        };
    }
    hydrate();
    $effect.root(() => {
        let isFirstRun = true;
        $effect(() => {
            const snapshot = $state.snapshot(state);
            let valueToStore;
            try {
                const transformed = beforeWrite(snapshot);
                valueToStore = serializer ? serializer.stringify(transformed) : transformed;
            }
            catch (error) {
                onWriteError(error);
                return;
            }
            if (isFirstRun) {
                isFirstRun = false;
                return;
            }
            if (!isLoading && !skipNextWrite) {
                idbSetItem(key, valueToStore, indexedDBOptions)
                    .then(() => {
                    if (syncTabs && broadcastChannel) {
                        broadcastChannel.postMessage({ key, value: valueToStore });
                    }
                })
                    .catch(onWriteError);
            }
            if (skipNextWrite) {
                skipNextWrite = false;
            }
        });
        return () => {
            broadcastChannel?.close();
            if (browser) {
                closeDB(indexedDBOptions);
            }
        };
    });
    return {
        get current() {
            return state;
        },
        set current(newValue) {
            state = newValue;
        },
        get isLoading() {
            return isLoading;
        },
        get ready() {
            return ready;
        },
        reset() {
            state = initialValue;
        }
    };
}
