// remult@3.3.12/esm/src/live-query/SubscriptionChannel.js
import { Sort } from '../sort.js';
import { remult as defaultRemult } from '../remult-proxy.js';
import { getRepositoryInternals } from '../remult3/repository-internals.js';
export const streamUrl = 'stream';
//@internal
export class LiveQuerySubscriber {
    repo;
    query;
    sendDefaultState(onResult) {
        onResult(this.createReducerType(() => [...this.defaultQueryState], this.allItemsMessage(this.defaultQueryState)));
    }
    queryChannel;
    subscribeCode;
    unsubscribe = () => { };
    async setAllItems(result) {
        const items = await getRepositoryInternals(this.repo)._fromJsonArray(result, this.query.options);
        this.forListeners((listener) => {
            listener(() => {
                return items;
            });
        }, this.allItemsMessage(items));
    }
    allItemsMessage(items) {
        return [
            {
                type: 'all',
                data: items,
            },
        ];
    }
    forListeners(what, changes) {
        what((reducer) => {
            this.defaultQueryState = reducer(this.defaultQueryState);
            if (changes.find((c) => c.type === 'add' || c.type === 'replace')) {
                if (this.query.options.orderBy) {
                    const o = Sort.translateOrderByToSort(this.repo.metadata, this.query.options.orderBy);
                    this.defaultQueryState.sort((a, b) => o.compare(a, b));
                }
            }
        });
        for (const l of this.listeners) {
            what((reducer) => {
                l.next(this.createReducerType(reducer, changes));
            });
        }
    }
    createReducerType(applyChanges, changes) {
        return {
            applyChanges,
            changes,
            items: this.defaultQueryState,
        };
    }
    async handle(messages) {
        {
            let x = messages.filter(({ type }) => type == 'add' || type == 'replace');
            let loadedItems = await getRepositoryInternals(this.repo)._fromJsonArray(x.map((m) => m.data.item), this.query.options);
            for (let index = 0; index < x.length; index++) {
                const element = x[index];
                element.data.item = loadedItems[index];
            }
        }
        this.forListeners((listener) => {
            listener((items) => {
                if (!items)
                    items = [];
                for (const message of messages) {
                    switch (message.type) {
                        case 'all':
                            this.setAllItems(message.data);
                            break;
                        case 'replace': {
                            items = items.map((x) => this.repo.metadata.idMetadata.getId(x) === message.data.oldId
                                ? message.data.item
                                : x);
                            break;
                        }
                        case 'add':
                            items = items.filter((x) => this.repo.metadata.idMetadata.getId(x) !==
                                this.repo.metadata.idMetadata.getId(message.data.item));
                            items.push(message.data.item);
                            break;
                        case 'remove':
                            items = items.filter((x) => this.repo.metadata.idMetadata.getId(x) !== message.data.id);
                            break;
                    }
                }
                return items;
            });
        }, messages);
    }
    defaultQueryState = [];
    listeners = [];
    id = String(crypto.randomUUID());
    constructor(repo, query, userId) {
        this.repo = repo;
        this.query = query;
        this.queryChannel = `users:${userId}:queries:${this.id}`;
        this.id = this.queryChannel;
    }
}
export const liveQueryKeepAliveRoute = '_liveQueryKeepAlive';
/**
 * The `SubscriptionChannel` class is used to send messages from the backend to the frontend,
 * using the same mechanism used by live queries.
 *
 * @template messageType The type of the message that the channel will handle.
 * @example
 * ```ts
 * // Defined in code that is shared between the frontend and the backend
 * export const notif = new SubscriptionChannel<{ coolInfo: string }>("notif");
 *
 * // Backend: Publishing a message
 * notif.publish({ coolInfo: "Message coming from the backend" });
 *
 * // Frontend: Subscribing to messages
 * notif.subscribe((message) => {
 *     console.log(`Something exciting happened: ${message.coolInfo}`);
 * });
 *
 * // Note: If you want to publish from the frontend, use a BackendMethod for that.
 * ```
 *  ---
 *
 *  ### Use cases:
 *
 *  #### A user _(Paul)_ wants to notify everyone _(Eric & Simona)_
 *  ```ts
 *  // shared code between the frontend and the backend
 *  export const notif = new SubscriptionChannel<{ from: UserInfo, msg: string }>("notif");
 *
 *  // Frontend, Paul can call a BackendMethod to send his message
 *  NotificationContoller.send("Hello everyone!")
 *
 *  // Backend
 *  export class NotificationContoller {
 *    @BackendMethod({ allowed: true })
 *    static async send(message: string) {
 *      notif.publish({ from: remult.user, msg: message });
 *    }
 *  }
 *
 *  // Frontend: Eric & Simona can subscribe to messages
 *  notif.subscribe((message) => {
 *    if (message.from === remult.user) return;
 *    console.log(`Message from ${message.from.name}: ${message.msg}`);
 *  });
 *  ```
 *
 *  #### A complexe chart that needs to be updated on multiple changes
 *  ```ts
 *  // shared code between the frontend and the backend
 *  export const chart = new SubscriptionChannel<{ kind: "pie" | "bar" | "line", data: number[] }>("chart");
 *
 *  // Entity
 *  @Entity<Task>("tasks", {
 *    saved: async (e) => {
 *      if (e.completed) chart.publish({ kind: "pie", data: [1, 2, 3] });
 *    }
 *  })
 *  export class Task { ... }
 *
 *  // Frontend: in the chart component, we can subscribe to messages
 *  chart.subscribe((message) => {
 *    chartData = message.data;
 *  });
 *  ```
 *
 */
export class SubscriptionChannel {
    channelKey;
    /**
     * Constructs a new `SubscriptionChannel` instance.
     *
     * @param {string} channelKey The key that identifies the channel.
     */
    constructor(channelKey) {
        this.channelKey = channelKey;
    }
    /**
     * Publishes a message to the channel. This method should only be used on the backend.
     *
     * @param {messageType} message The message to be published.
     * @param {Remult} [remult] An optional instance of Remult to use for publishing the message.
     */
    publish(message, remult) {
        remult = remult || defaultRemult;
        remult.subscriptionServer.publishMessage(this.channelKey, message);
    }
    //@internal
    subscribe(next, remult) {
        remult = remult || defaultRemult;
        let listener = next;
        if (typeof next === 'function') {
            listener = {
                next,
            };
        }
        listener.error ??= () => { };
        listener.complete ??= () => { };
        return remult.liveQuerySubscriber.subscribeChannel(this.channelKey, listener);
    }
}
//TODO2 - consider moving the queued job mechanism into this.
