// @flue/runtime@0.9.2/dist/node/index.mjs
import { c as invokeWorkflowAttached, m as InvalidRequestError, s as invokeDirectAttached } from "../handle-agent-C2C75lpq.mjs";
import { c as registeredWorkflowsForTransport, s as registeredAgentsForTransport } from "../flue-app-NVPwXKDc.mjs";
import { r as generateWorkflowRunId } from "../ids-DiSJFsle.mjs";
import { t as abortErrorFor } from "../abort-B-q3xjj8.mjs";
import { n as parseAgentWebSocketMessage, r as parseWorkflowWebSocketMessage, t as createWebSocketErrorMessage } from "../websocket-protocol-BSIpcV6C.mjs";
import { exec } from "node:child_process";
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { promisify } from "node:util";
import { upgradeWebSocket } from "@hono/node-server";
import { WebSocket, WebSocketServer } from "ws";
//#region src/node/local-env.ts
/**
* Pure-Node `SessionEnv` backed by the host filesystem and `child_process`.
*
* Internal implementation behind the `local()` sandbox factory (see
* `./local.ts`). Not exported from `@flue/runtime/node` — user code reaches
* this through `local(...)`. `exec` shells out via `child_process.exec`;
* file methods call `node:fs/promises` directly.
*/
const execAsync = promisify(exec);
/**
* Shell-essential env vars inherited from `process.env` by default. Pulled
* once at sandbox construction.
*
* Invariant: nothing on this list should be sensitive on a typical host.
* Adding entries here is a security-relevant decision — secrets, tokens,
* cloud-provider creds, and agent sockets MUST NOT appear. To expose
* anything else, callers opt in explicitly via `options.env`.
*/
const DEFAULT_LOCAL_ENV_ALLOWLIST = [
	"PATH",
	"HOME",
	"USER",
	"LOGNAME",
	"HOSTNAME",
	"SHELL",
	"LANG",
	"LC_ALL",
	"LC_CTYPE",
	"TZ",
	"TERM",
	"TMPDIR",
	"TMP",
	"TEMP"
];
/**
* Snapshot `process.env` through the allowlist, then layer user overrides.
* Called once per sandbox; the result is captured in a closure and reused
* across every `exec()` so per-call cost stays minimal and the env shape
* is stable for the sandbox's lifetime (host mutations to `process.env`
* after construction are NOT picked up).
*/
function resolveBaseEnv(userEnv) {
	if (userEnv !== void 0 && (typeof userEnv !== "object" || Array.isArray(userEnv))) throw new TypeError("[flue] local() `env` must be a Record<string, string | undefined>. To inherit the full host env, pass `env: { ...process.env }`.");
	const base = {};
	for (const key of DEFAULT_LOCAL_ENV_ALLOWLIST) {
		const value = process.env[key];
		if (value !== void 0) base[key] = value;
	}
	if (!userEnv) return base;
	for (const [key, value] of Object.entries(userEnv)) if (value === void 0) delete base[key];
	else base[key] = value;
	return base;
}
function createLocalSessionEnv(options = {}) {
	const cwd = path.resolve(options.cwd ?? process.cwd());
	const baseEnv = resolveBaseEnv(options.env);
	const resolvePath = (p) => path.isAbsolute(p) ? p : path.resolve(cwd, p);
	return {
		async exec(command, opts) {
			const signal = opts?.signal;
			if (signal?.aborted) throw abortErrorFor(signal);
			const timeoutSignal = typeof opts?.timeout === "number" ? AbortSignal.timeout(opts.timeout * 1e3) : void 0;
			const mergedSignal = signal && timeoutSignal ? AbortSignal.any([signal, timeoutSignal]) : signal ?? timeoutSignal;
			try {
				const { stdout, stderr } = await execAsync(command, {
					cwd: opts?.cwd ? resolvePath(opts.cwd) : cwd,
					env: opts?.env ? {
						...baseEnv,
						...opts.env
					} : baseEnv,
					signal: mergedSignal,
					encoding: "utf8",
					maxBuffer: 64 * 1024 * 1024
				});
				if (signal?.aborted) throw abortErrorFor(signal);
				return {
					stdout,
					stderr,
					exitCode: 0
				};
			} catch (err) {
				if (signal?.aborted) throw abortErrorFor(signal);
				if (err && typeof err === "object" && "code" in err) return {
					stdout: typeof err.stdout === "string" ? err.stdout : "",
					stderr: typeof err.stderr === "string" ? err.stderr : String(err.message ?? ""),
					exitCode: typeof err.code === "number" ? err.code : 1
				};
				throw err;
			}
		},
		async readFile(p) {
			return fs.readFile(resolvePath(p), "utf8");
		},
		async readFileBuffer(p) {
			const buf = await fs.readFile(resolvePath(p));
			return new Uint8Array(buf.buffer, buf.byteOffset, buf.byteLength);
		},
		async writeFile(p, content) {
			const resolved = resolvePath(p);
			const dir = path.dirname(resolved);
			if (dir && dir !== resolved) await fs.mkdir(dir, { recursive: true });
			await fs.writeFile(resolved, content);
		},
		async stat(p) {
			const s = await fs.stat(resolvePath(p));
			return {
				isFile: s.isFile(),
				isDirectory: s.isDirectory(),
				isSymbolicLink: s.isSymbolicLink(),
				size: s.size,
				mtime: s.mtime
			};
		},
		async readdir(p) {
			return fs.readdir(resolvePath(p));
		},
		async exists(p) {
			try {
				await fs.access(resolvePath(p));
				return true;
			} catch {
				return false;
			}
		},
		async mkdir(p, opts) {
			await fs.mkdir(resolvePath(p), { recursive: opts?.recursive ?? false });
		},
		async rm(p, opts) {
			await fs.rm(resolvePath(p), {
				recursive: opts?.recursive ?? false,
				force: opts?.force ?? false
			});
		},
		cwd,
		resolvePath
	};
}
//#endregion
//#region src/node/local.ts
function local(options = {}) {
	return { createSessionEnv: async () => createLocalSessionEnv(options) };
}
//#endregion
//#region src/node/websocket.ts
function createNodeWebSocketTransport(options) {
	installErrorEvent();
	const server = new WebSocketServer({
		noServer: true,
		maxPayload: options.maxPayload ?? 1024 * 1024
	});
	const runtime = {
		target: "node",
		manifest: options.manifest
	};
	const agents = new Set(registeredAgentsForTransport(runtime, "websocket"));
	const workflows = new Set(registeredWorkflowsForTransport(runtime, "websocket"));
	return {
		server,
		agentRoute: upgradeWebSocket((c) => {
			const name = c.req.param("name") ?? "";
			const id = c.req.param("id") ?? "";
			const handler = options.agentHandlers[name];
			if (!agents.has(name) || !id || !handler) throw new Error("[flue] Node runtime is missing WebSocket agent handler configuration.");
			const target = {
				kind: "agent",
				name,
				id,
				handler
			};
			const request = c.req.raw;
			return {
				onOpen: (_event, socket) => openAgentSocket(socket, target),
				onMessage: (event, socket) => receiveAgentMessage(socket, request, target, event.data, options),
				onError: (_event, socket) => terminateSocket(socket)
			};
		}),
		workflowRoute: upgradeWebSocket((c) => {
			const name = c.req.param("name") ?? "";
			const handler = options.workflowHandlers[name];
			if (!workflows.has(name) || !handler) throw new Error("[flue] Node runtime is missing WebSocket workflow handler configuration.");
			const target = {
				kind: "workflow",
				name,
				handler
			};
			const request = c.req.raw;
			let invoked = false;
			return {
				onOpen: (_event, socket) => openWorkflowSocket(socket, target),
				onMessage: (event, socket) => {
					if (typeof event.data !== "string") {
						sendError(socket, new InvalidRequestError({ reason: "Binary WebSocket messages are not supported." }));
						socket.close(1003, "Binary messages are not supported");
						return;
					}
					let message;
					try {
						message = parseWorkflowWebSocketMessage(event.data);
					} catch (error) {
						sendError(socket, error);
						return;
					}
					if (invoked) {
						sendError(socket, new InvalidRequestError({ reason: "Workflow WebSocket connections accept one invocation only." }), message.requestId);
						socket.close(1008, "Workflow accepts one invocation only");
						return;
					}
					invoked = true;
					invokeWorkflow(socket, request, target, message, options);
				},
				onError: (_event, socket) => terminateSocket(socket)
			};
		}),
		async close() {
			for (const socket of server.clients) socket.terminate();
			await new Promise((resolve) => server.close(() => resolve()));
		}
	};
}
function openAgentSocket(socket, target) {
	send(socket, {
		version: 1,
		type: "ready",
		target: "agent",
		name: target.name,
		instanceId: target.id
	});
}
function receiveAgentMessage(socket, request, target, raw, options) {
	if (typeof raw !== "string") {
		sendError(socket, new InvalidRequestError({ reason: "Binary WebSocket messages are not supported." }));
		socket.close(1003, "Binary messages are not supported");
		return;
	}
	let message;
	try {
		message = parseAgentWebSocketMessage(raw);
	} catch (error) {
		sendError(socket, error);
		return;
	}
	if (message.type === "ping") {
		send(socket, {
			version: 1,
			type: "pong",
			requestId: message.requestId
		});
		return;
	}
	invokeAgentPrompt(socket, request, target, message, options);
}
async function invokeAgentPrompt(socket, request, target, message, options) {
	let didStart = false;
	try {
		const result = await invokeDirectAttached({
			agentName: target.name,
			id: target.id,
			payload: {
				message: message.message,
				session: message.session
			},
			request,
			handler: target.handler,
			createContext: options.createContext,
			runHandler: options.runHandler,
			onEvent: (event) => {
				if (!didStart) {
					didStart = true;
					send(socket, {
						version: 1,
						type: "started",
						requestId: message.requestId
					});
				}
				send(socket, {
					version: 1,
					type: "event",
					requestId: message.requestId,
					event
				});
			},
			emitIdleOnComplete: true
		});
		send(socket, {
			version: 1,
			type: "result",
			requestId: message.requestId,
			result: result ?? null
		});
	} catch (error) {
		sendError(socket, error, message.requestId);
	}
}
function openWorkflowSocket(socket, target) {
	send(socket, {
		version: 1,
		type: "ready",
		target: "workflow",
		name: target.name
	});
}
async function invokeWorkflow(socket, request, target, message, options) {
	const runId = generateWorkflowRunId(target.name);
	let didStart = false;
	const bufferedEvents = [];
	try {
		const invocation = await invokeWorkflowAttached({
			owner: {
				kind: "workflow",
				workflowName: target.name,
				instanceId: runId
			},
			id: runId,
			runId,
			payload: message.payload,
			request,
			handler: target.handler,
			createContext: options.createContext,
			startWorkflowAdmission: options.startWorkflowAdmission ?? ((_runId, run) => Promise.resolve().then(run)),
			onAdmitted: () => {
				didStart = true;
				send(socket, {
					version: 1,
					type: "started",
					requestId: message.requestId,
					runId
				});
				for (const event of bufferedEvents) send(socket, {
					version: 1,
					type: "event",
					requestId: message.requestId,
					runId,
					event
				});
			},
			onEvent: (event) => {
				if (!didStart) {
					bufferedEvents.push(event);
					return;
				}
				send(socket, {
					version: 1,
					type: "event",
					requestId: message.requestId,
					runId,
					event
				});
			},
			emitIdleOnComplete: true,
			runStore: options.runStore,
			runSubscribers: options.runSubscribers,
			runRegistry: options.runRegistry
		});
		send(socket, {
			version: 1,
			type: "result",
			requestId: message.requestId,
			runId,
			result: invocation.result ?? null
		});
		socket.close(1e3, "Workflow completed");
	} catch (error) {
		sendError(socket, error, message.requestId, runId);
		socket.close(1011, "Workflow failed");
	}
}
function installErrorEvent() {
	if (typeof globalThis.ErrorEvent !== "undefined") return;
	class NodeErrorEvent extends Event {
		error;
		constructor(type, init) {
			super(type);
			this.error = init?.error;
		}
	}
	Object.defineProperty(globalThis, "ErrorEvent", {
		configurable: true,
		value: NodeErrorEvent
	});
}
function terminateSocket(socket) {
	socket.raw?.terminate();
}
function sendError(socket, error, requestId, runId) {
	send(socket, createWebSocketErrorMessage(error, requestId, runId));
}
function send(socket, message) {
	if (socket.readyState === WebSocket.OPEN) socket.send(JSON.stringify(message));
}
//#endregion
export { createNodeWebSocketTransport, local };
