// remult@3.3.12/src/data-providers/json-entity-indexed-db-data-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonEntityIndexedDbStorage = void 0;
var tslib_1 = require("tslib");
var JsonEntityIndexedDbStorage = /** @class */ (function () {
    function JsonEntityIndexedDbStorage(dbName, storeName) {
        if (dbName === void 0) { dbName = 'db'; }
        if (storeName === void 0) { storeName = 'jsonStore'; }
        this.dbName = dbName;
        this.storeName = storeName;
        this.supportsRawJson = true;
    }
    JsonEntityIndexedDbStorage.prototype.getItem = function (entityDbName) {
        return this.run('readonly', function (s) { return s.get(entityDbName); });
    };
    JsonEntityIndexedDbStorage.prototype.setItem = function (entityDbName, json) {
        return this.run('readwrite', function (s) {
            return s.put(json, entityDbName);
        });
    };
    JsonEntityIndexedDbStorage.prototype.removeItem = function (entityDbName) {
        return this.run('readwrite', function (s) { return s.delete(entityDbName); });
    };
    JsonEntityIndexedDbStorage.prototype.clear = function () {
        return this.run('readwrite', function (s) { return s.clear(); });
    };
    //@internal
    JsonEntityIndexedDbStorage.prototype.run = function (mode, op) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var attempt, err_1;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        attempt = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            var db;
                            var _this = this;
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, this.init()];
                                    case 1:
                                        db = _a.sent();
                                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                                var tx = db.transaction([_this.storeName], mode);
                                                tx.onabort = function () { return reject(tx.error); };
                                                var request = op(tx.objectStore(_this.storeName));
                                                request.onerror = function () { return reject(request.error); };
                                                request.onsuccess = function () { var _a; return resolve((_a = request.result) !== null && _a !== void 0 ? _a : null); };
                                            })];
                                }
                            });
                        }); };
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 5]);
                        return [4 /*yield*/, attempt()];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        err_1 = _a.sent();
                        this.db = undefined;
                        return [4 /*yield*/, attempt()];
                    case 4: return [2 /*return*/, _a.sent()];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    //@internal
    JsonEntityIndexedDbStorage.prototype.init = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.db)
                            return [2 /*return*/, this.db];
                        _a = this;
                        return [4 /*yield*/, new Promise(function (resolve, reject) {
                                var request = indexedDB.open(_this.dbName, 1);
                                request.onerror = function () { return reject(request.error); };
                                request.onupgradeneeded = function () {
                                    return request.result.createObjectStore(_this.storeName);
                                };
                                request.onsuccess = function () {
                                    var db = request.result;
                                    var forget = function () {
                                        if (_this.db === db)
                                            _this.db = undefined;
                                    };
                                    db.onclose = forget;
                                    db.onversionchange = function () {
                                        db.close();
                                        forget();
                                    };
                                    resolve(db);
                                };
                            })];
                    case 1:
                        _a.db = _b.sent();
                        return [2 /*return*/, this.db];
                }
            });
        });
    };
    return JsonEntityIndexedDbStorage;
}());
exports.JsonEntityIndexedDbStorage = JsonEntityIndexedDbStorage;
