// remult@3.3.12/esm/src/remult3/pagedQueryResult.js
export function pagedQueryResult(pageSize, getPage, count) {
    const iterator = () => {
        let pageNumber = 1;
        let currentPage = [];
        let itemIndex = 0;
        return {
            next: async () => {
                if (itemIndex >= currentPage.length) {
                    currentPage = await getPage(pageNumber);
                    if (currentPage.length === 0) {
                        return { value: undefined, done: true };
                    }
                    pageNumber++;
                    itemIndex = 0;
                }
                const value = currentPage[itemIndex++];
                return { value, done: false };
            },
        };
    };
    return {
        count,
        getPage,
        forEach: async (what) => {
            let i = 0;
            const it = iterator();
            while (true) {
                const { value, done } = await it.next();
                if (done)
                    break;
                await what(value);
                i++;
            }
            return i;
        },
        [Symbol.asyncIterator]: iterator,
        paginator: async () => {
            const createPaginator = async (pageNumber) => {
                const items = await getPage(pageNumber);
                const hasNextPage = items.length === pageSize;
                const emptyPaginator = {
                    items: [],
                    hasNextPage: false,
                    count,
                    nextPage: async () => emptyPaginator,
                };
                return {
                    items,
                    hasNextPage,
                    count,
                    nextPage: async () => {
                        if (!hasNextPage) {
                            return emptyPaginator;
                        }
                        return createPaginator(pageNumber + 1);
                    },
                };
            };
            return createPaginator(1);
        },
    };
}
