// remult@3.3.12/src/remult3/pagedQueryResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pagedQueryResult = pagedQueryResult;
var tslib_1 = require("tslib");
function pagedQueryResult(pageSize, getPage, count) {
    var _a;
    var _this = this;
    var iterator = function () {
        var pageNumber = 1;
        var currentPage = [];
        var itemIndex = 0;
        return {
            next: function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var value;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!(itemIndex >= currentPage.length)) return [3 /*break*/, 2];
                            return [4 /*yield*/, getPage(pageNumber)];
                        case 1:
                            currentPage = _a.sent();
                            if (currentPage.length === 0) {
                                return [2 /*return*/, { value: undefined, done: true }];
                            }
                            pageNumber++;
                            itemIndex = 0;
                            _a.label = 2;
                        case 2:
                            value = currentPage[itemIndex++];
                            return [2 /*return*/, { value: value, done: false }];
                    }
                });
            }); },
        };
    };
    return _a = {
            count: count,
            getPage: getPage,
            forEach: function (what) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var i, it, _a, value, done;
                return tslib_1.__generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            i = 0;
                            it = iterator();
                            _b.label = 1;
                        case 1:
                            if (!true) return [3 /*break*/, 4];
                            return [4 /*yield*/, it.next()];
                        case 2:
                            _a = _b.sent(), value = _a.value, done = _a.done;
                            if (done)
                                return [3 /*break*/, 4];
                            return [4 /*yield*/, what(value)];
                        case 3:
                            _b.sent();
                            i++;
                            return [3 /*break*/, 1];
                        case 4: return [2 /*return*/, i];
                    }
                });
            }); }
        },
        _a[Symbol.asyncIterator] = iterator,
        _a.paginator = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var createPaginator;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                createPaginator = function (pageNumber) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                    var items, hasNextPage, emptyPaginator;
                    var _this = this;
                    return tslib_1.__generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, getPage(pageNumber)];
                            case 1:
                                items = _a.sent();
                                hasNextPage = items.length === pageSize;
                                emptyPaginator = {
                                    items: [],
                                    hasNextPage: false,
                                    count: count,
                                    nextPage: function () { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
                                        return [2 /*return*/, emptyPaginator];
                                    }); }); },
                                };
                                return [2 /*return*/, {
                                        items: items,
                                        hasNextPage: hasNextPage,
                                        count: count,
                                        nextPage: function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                            return tslib_1.__generator(this, function (_a) {
                                                if (!hasNextPage) {
                                                    return [2 /*return*/, emptyPaginator];
                                                }
                                                return [2 /*return*/, createPaginator(pageNumber + 1)];
                                            });
                                        }); },
                                    }];
                        }
                    });
                }); };
                return [2 /*return*/, createPaginator(1)];
            });
        }); },
        _a;
}
