// @flue/runtime@0.9.2/dist/events-BwRudw7t.mjs
import * as v from "valibot";
//#region src/agent-definition.ts
const AGENT_PROFILE_FIELDS = new Set([
	"name",
	"description",
	"model",
	"instructions",
	"skills",
	"tools",
	"subagents",
	"thinkingLevel",
	"compaction"
]);
const AGENT_RUNTIME_FIELDS = new Set([
	...AGENT_PROFILE_FIELDS,
	"profile",
	"cwd",
	"sandbox",
	"persist"
]);
const VALID_THINKING_LEVELS = {
	off: true,
	minimal: true,
	low: true,
	medium: true,
	high: true,
	xhigh: true
};
const AgentProfileSchema = v.looseObject({
	name: v.optional(v.string()),
	description: v.optional(v.string()),
	model: v.optional(v.union([v.string(), v.literal(false)])),
	instructions: v.optional(v.string()),
	skills: v.optional(v.array(v.unknown())),
	tools: v.optional(v.array(v.unknown())),
	subagents: v.optional(v.array(v.unknown())),
	thinkingLevel: v.optional(v.string()),
	compaction: v.optional(v.union([v.literal(false), v.looseObject({})]))
});
/**
* Validates and returns a reusable agent profile. Use profiles as the baseline
* for a created agent or as named subagents available to `session.task()`.
*
* Throws when the profile contains unknown fields, invalid capabilities,
* duplicate capability names, or circular subagents.
*/
function defineAgentProfile(profile) {
	assertAgentProfile(profile, "defineAgentProfile()", /* @__PURE__ */ new WeakSet());
	return profile;
}
/**
* Creates an agent initializer. Default-export the returned value from an
* `agents/<name>.ts` module to define an addressable agent, or pass it to
* `ctx.init()` inside a workflow.
*
* The initializer runs whenever the runtime initializes a harness from the
* created agent: when a workflow calls `ctx.init()`, and when the runtime
* prepares an addressable agent interaction. Do not treat it as a one-time
* constructor for a persistent agent instance id. Return a runtime config
* object with `model: '<provider>/<model>'`, `model: false`, or a profile with
* its own model field.
*/
function createAgent(initialize) {
	if (typeof initialize !== "function") throw new Error("[flue] createAgent() requires an initializer function.");
	return Object.freeze({
		__flueCreatedAgent: true,
		initialize
	});
}
function assertResolvedAgentProfile(profile, label) {
	assertAgentProfile(profile, label, /* @__PURE__ */ new WeakSet());
	return profile;
}
function resolveAgentProfile(options) {
	assertAgentRuntimeConfig(options);
	const profile = options?.profile;
	return {
		name: hasOwn(options, "name") ? options?.name : profile?.name,
		description: hasOwn(options, "description") ? options?.description : profile?.description,
		model: hasOwn(options, "model") ? options?.model : profile?.model,
		instructions: hasOwn(options, "instructions") ? options?.instructions : profile?.instructions,
		skills: mergeArrays(profile?.skills, options?.skills),
		tools: mergeArrays(profile?.tools, options?.tools),
		subagents: mergeArrays(profile?.subagents, options?.subagents),
		thinkingLevel: hasOwn(options, "thinkingLevel") ? options?.thinkingLevel : profile?.thinkingLevel,
		compaction: hasOwn(options, "compaction") ? options?.compaction : profile?.compaction
	};
}
function extendAgentProfile(profile, extensions) {
	return {
		...profile,
		skills: mergeArrays(profile.skills, extensions.skills),
		tools: mergeArrays(profile.tools, extensions.tools),
		subagents: mergeArrays(profile.subagents, extensions.subagents)
	};
}
function hasOwn(value, key) {
	return Boolean(value && Object.hasOwn(value, key));
}
function mergeArrays(base, additions) {
	if (base === void 0 && additions === void 0) return void 0;
	return [...base ?? [], ...additions ?? []];
}
function assertAgentRuntimeConfig(value) {
	if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error("[flue] createAgent() initializer must return an agent runtime config object.");
	for (const key of Object.keys(value)) if (!AGENT_RUNTIME_FIELDS.has(key)) throw new Error(`[flue] createAgent() initializer returned unknown runtime config field "${key}".`);
	if (value.profile !== void 0) assertAgentProfile(value.profile, "createAgent() profile", /* @__PURE__ */ new WeakSet());
}
function assertAgentProfile(value, label, activeDefinitions) {
	const parsed = v.safeParse(AgentProfileSchema, value);
	if (!parsed.success) throw new Error(`[flue] ${label} requires a valid agent profile: ${formatIssues(parsed.issues)}.`);
	const definition = parsed.output;
	const source = value;
	if (activeDefinitions.has(source)) throw new Error(`[flue] ${label} must not contain circular subagents.`);
	activeDefinitions.add(source);
	assertKnownFields(definition, label);
	if (definition.name !== void 0) assertAgentName(definition.name, `${label} name`);
	if (definition.description !== void 0) assertNonEmptyString(definition.description, `${label} description`);
	assertThinkingLevel(definition.thinkingLevel, label);
	assertCompaction(definition.compaction, label);
	assertTools(definition.tools, label);
	assertSkills(definition.skills, label);
	assertSubagents(definition.subagents, label, activeDefinitions);
	assertUniqueNames(definition.tools, `${label} tools`, "tool");
	assertUniqueNames(definition.skills, `${label} skills`, "skill");
	assertUniqueNames(definition.subagents, `${label} subagents`, "subagent");
	activeDefinitions.delete(source);
}
function assertKnownFields(definition, label) {
	for (const key of Object.keys(definition)) if (!AGENT_PROFILE_FIELDS.has(key)) throw new Error(`[flue] ${label} received unknown agent profile field "${key}".`);
}
function assertThinkingLevel(value, label) {
	if (value !== void 0 && !(value in VALID_THINKING_LEVELS)) throw new Error(`[flue] ${label} thinkingLevel must be one of: ${Object.keys(VALID_THINKING_LEVELS).join(", ")}.`);
}
function assertCompaction(definition, label) {
	if (definition === void 0 || definition === false) return;
	for (const key of Object.keys(definition)) if (key !== "reserveTokens" && key !== "keepRecentTokens" && key !== "model") throw new Error(`[flue] ${label} compaction received unknown field "${key}".`);
	assertTokenCount(definition.reserveTokens, `${label} compaction.reserveTokens`);
	assertTokenCount(definition.keepRecentTokens, `${label} compaction.keepRecentTokens`);
	if (definition.model !== void 0 && typeof definition.model !== "string") throw new Error(`[flue] ${label} compaction.model must be a string.`);
}
function assertTokenCount(value, label) {
	if (value === void 0) return;
	if (!Number.isFinite(value) || !Number.isInteger(value) || value < 0) throw new Error(`[flue] ${label} must be a non-negative integer.`);
}
function assertTools(values, label) {
	for (const [index, value] of values?.entries() ?? []) {
		if (!value || typeof value !== "object") throw new Error(`[flue] ${label} tools[${index}] must be a tool definition object.`);
		const tool = value;
		assertNonEmptyString(tool.name, `${label} tools[${index}].name`);
		assertNonEmptyString(tool.description, `${label} tools[${index}].description`);
		if (!tool.parameters || typeof tool.parameters !== "object") throw new Error(`[flue] ${label} tools[${index}].parameters is required.`);
		if (typeof tool.execute !== "function") throw new Error(`[flue] ${label} tools[${index}].execute must be a function.`);
	}
}
function assertSkills(values, label) {
	for (const [index, value] of values?.entries() ?? []) {
		if (!value || typeof value !== "object") throw new Error(`[flue] ${label} skills[${index}] must be a skill definition object.`);
		const skill = value;
		assertNonEmptyString(skill.name, `${label} skills[${index}].name`);
		assertNonEmptyString(skill.description, `${label} skills[${index}].description`);
	}
}
function assertSubagents(values, label, activeDefinitions) {
	for (const [index, value] of values?.entries() ?? []) {
		if (!value || typeof value !== "object") throw new Error(`[flue] ${label} subagents[${index}] must be an agent definition object.`);
		assertAgentName(value.name, `${label} subagents[${index}].name`);
		assertAgentProfile(value, `${label} subagents[${index}]`, activeDefinitions);
	}
}
function assertAgentName(value, label) {
	assertNonEmptyString(value, label);
	if (!/^[A-Za-z][A-Za-z0-9_-]*$/.test(value)) throw new Error(`[flue] ${label} must start with a letter and contain only letters, numbers, "_", or "-".`);
}
function assertNonEmptyString(value, label) {
	if (typeof value !== "string" || value.trim().length === 0) throw new Error(`[flue] ${label} must be a non-empty string.`);
}
function assertUniqueNames(values, label, kind) {
	if (!values) return;
	const seen = /* @__PURE__ */ new Set();
	for (const value of values) {
		const name = value.name;
		if (!name) continue;
		if (seen.has(name)) throw new Error(`[flue] ${label} must not contain duplicate ${kind} name "${name}".`);
		seen.add(name);
	}
}
function formatIssues(issues) {
	return issues.map((issue) => issue.message).join("; ");
}
//#endregion
//#region src/runtime/events.ts
const subscribers = /* @__PURE__ */ new Set();
/**
* Subscribe to live workflow-run or agent-interaction activity emitted in this isolate.
* The subscription does not replay durable workflow history or aggregate events
* across processes or Cloudflare Durable Object isolates.
*
* Usage (typically at the top of `app.ts`):

*
*     import { observe } from '@flue/runtime';
*
*     observe((event, ctx) => {
*       if (event.type === 'run_end' && event.isError) {
*         // ship to your error reporter, metrics sink, etc.
*       }
*     });
*
* The returned function unsubscribes the listener. Most error
* reporting and telemetry use cases register once at startup and
* never unsubscribe — the returned function is provided for tests
* and dynamic-wiring scenarios.
*
* Subscribers are invoked synchronously from the event emit path with an
* isolated JSON snapshot. They should be cheap and side-effect-only; returned
* promises are observed for rejection but are not awaited. Queue substantial
* work outside the callback rather than blocking emission.
*/
function observe(subscriber) {
	subscribers.add(subscriber);
	return () => {
		subscribers.delete(subscriber);
	};
}
/**
* Internal: dispatch a single event to every registered subscriber.
* Called from `createFlueContext`'s `emitEvent` after the per-context
* subscribers have run.
*/
function dispatchGlobalEvent(event, ctx) {
	if (subscribers.size === 0) return;
	let serializedEvent;
	try {
		serializedEvent = JSON.stringify(event);
		if (serializedEvent === void 0) throw new Error("Event snapshot serialization returned undefined.");
	} catch (error) {
		reportSubscriberFailure(error);
		return;
	}
	for (const subscriber of [...subscribers]) try {
		Promise.resolve(subscriber(JSON.parse(serializedEvent), ctx)).catch(reportSubscriberFailure);
	} catch (error) {
		reportSubscriberFailure(error);
	}
}
function reportSubscriberFailure(error) {
	console.error("[flue:observe] subscriber failed:", error);
}
//#endregion
export { defineAgentProfile as a, createAgent as i, observe as n, extendAgentProfile as o, assertResolvedAgentProfile as r, resolveAgentProfile as s, dispatchGlobalEvent as t };
