// @flue/runtime@0.9.2/dist/flue-app-NVPwXKDc.mjs
import { S as toHttpResponse, T as validateWorkflowRequest, a as handleAgentRequest, b as ValidationError, d as handleRunRouteRequest, g as RouteNotFoundError, h as MethodNotAllowedError, o as handleWorkflowRequest, v as RunNotFoundError, w as validateAgentRequest, x as configureErrorRendering, y as RunRegistryUnavailableError } from "./handle-agent-C2C75lpq.mjs";
import { r as generateWorkflowRunId } from "./ids-DiSJFsle.mjs";
import * as v from "valibot";
import { Hono } from "hono";
import { describeRoute, openAPIRouteHandler, resolver, validator } from "hono-openapi";
//#region src/runtime/dispatch.ts
async function enqueueDispatch(options) {
	const agent = options.request.agent;
	const input = validateAndCloneDispatchRequest(options.request, agent, options.rt);
	const session = options.request.session ?? "default";
	return options.dispatchQueue.enqueue({
		dispatchId: crypto.randomUUID(),
		agent,
		id: options.request.id,
		session,
		input,
		acceptedAt: (/* @__PURE__ */ new Date()).toISOString()
	});
}
function validateAndCloneDispatchRequest(request, agent, rt) {
	if (typeof agent !== "string" || agent.trim() === "") throw new Error("[flue] dispatch() requires a non-empty target agent.");
	if (typeof request.id !== "string" || request.id.trim() === "") throw new Error("[flue] dispatch() requires a non-empty \"id\" target agent instance id.");
	if (request.session !== void 0 && (typeof request.session !== "string" || request.session.trim() === "")) throw new Error("[flue] dispatch() requires a non-empty \"session\" target session id when provided.");
	if (request.input === void 0) throw new Error("[flue] dispatch() requires an \"input\" payload. Use null for an intentional empty payload.");
	if (!agentExists(rt, agent)) throw new Error(`[flue] dispatch() target agent "${agent}" is not registered.`);
	return cloneJsonSerializable(request.input, "dispatch().input");
}
function agentExists(rt, agentName) {
	return (rt.manifest?.agents ?? []).some((agent) => agent.name === agentName);
}
function cloneJsonSerializable(value, label) {
	assertJsonLike(value, label, /* @__PURE__ */ new WeakSet());
	let json;
	try {
		json = JSON.stringify(value);
	} catch (error) {
		throw new Error(`[flue] ${label} must be JSON-serializable: ${error instanceof Error ? error.message : String(error)}`);
	}
	return JSON.parse(json);
}
function assertJsonLike(value, path, seen) {
	if (value === null) return;
	const type = typeof value;
	if (type === "string" || type === "number" || type === "boolean") {
		if (type === "number" && !Number.isFinite(value)) throw new Error(`[flue] ${path} must not contain non-finite numbers.`);
		return;
	}
	if (type === "undefined" || type === "function" || type === "symbol" || type === "bigint") throw new Error(`[flue] ${path} must not contain ${type} values.`);
	if (typeof value !== "object") return;
	if (seen.has(value)) throw new Error(`[flue] ${path} must not contain circular references.`);
	seen.add(value);
	if (Array.isArray(value)) {
		for (let i = 0; i < value.length; i++) assertJsonLike(value[i], `${path}[${i}]`, seen);
		seen.delete(value);
		return;
	}
	if (Object.getPrototypeOf(value) !== Object.prototype && Object.getPrototypeOf(value) !== null) throw new Error(`[flue] ${path} must contain only plain JSON objects, arrays, strings, numbers, booleans, or null.`);
	for (const [key, child] of Object.entries(value)) assertJsonLike(child, `${path}.${key}`, seen);
	seen.delete(value);
}
//#endregion
//#region src/runtime/schemas.ts
const RunStatusSchema = v.picklist([
	"active",
	"completed",
	"errored"
]);
const ErrorEnvelopeSchema = v.object({ error: v.object({
	type: v.string(),
	message: v.string(),
	details: v.string(),
	dev: v.optional(v.string()),
	meta: v.optional(v.record(v.string(), v.unknown()))
}) });
const RunOwnerSchema = v.object({
	kind: v.literal("workflow"),
	workflowName: v.string(),
	instanceId: v.string()
});
const RunRecordSchema = v.object({
	runId: v.string(),
	owner: RunOwnerSchema,
	status: RunStatusSchema,
	startedAt: v.string(),
	payload: v.optional(v.unknown()),
	endedAt: v.optional(v.string()),
	isError: v.optional(v.boolean()),
	durationMs: v.optional(v.number()),
	result: v.optional(v.unknown()),
	error: v.optional(v.unknown())
});
const RunPointerSchema = v.object({
	runId: v.string(),
	owner: RunOwnerSchema,
	status: RunStatusSchema,
	startedAt: v.string(),
	endedAt: v.optional(v.string()),
	durationMs: v.optional(v.number()),
	isError: v.optional(v.boolean())
});
const EventBaseSchema = {
	runId: v.optional(v.string()),
	instanceId: v.optional(v.string()),
	dispatchId: v.optional(v.string()),
	eventIndex: v.optional(v.number()),
	timestamp: v.optional(v.string()),
	session: v.optional(v.string()),
	parentSession: v.optional(v.string()),
	taskId: v.optional(v.string()),
	harness: v.optional(v.string()),
	operationId: v.optional(v.string()),
	turnId: v.optional(v.string())
};
const flueEvent = (entries) => v.looseObject({
	...EventBaseSchema,
	...entries
});
const PromptUsageSchema = v.object({
	input: v.number(),
	output: v.number(),
	cacheRead: v.number(),
	cacheWrite: v.number(),
	totalTokens: v.number(),
	cost: v.object({
		input: v.number(),
		output: v.number(),
		cacheRead: v.number(),
		cacheWrite: v.number(),
		total: v.number()
	})
});
const LlmTextContentSchema = v.object({
	type: v.literal("text"),
	text: v.string(),
	textSignature: v.optional(v.string())
});
const LlmThinkingContentSchema = v.object({
	type: v.literal("thinking"),
	thinking: v.string(),
	thinkingSignature: v.optional(v.string()),
	redacted: v.optional(v.boolean())
});
const LlmImageContentSchema = v.object({
	type: v.literal("image"),
	data: v.string(),
	mimeType: v.string()
});
const LlmToolCallSchema = v.object({
	type: v.literal("toolCall"),
	id: v.string(),
	name: v.string(),
	arguments: v.record(v.string(), v.unknown()),
	thoughtSignature: v.optional(v.string())
});
const LlmUserMessageSchema = v.object({
	role: v.literal("user"),
	content: v.union([v.string(), v.array(v.union([LlmTextContentSchema, LlmImageContentSchema]))])
});
const LlmAssistantMessageSchema = v.object({
	role: v.literal("assistant"),
	content: v.array(v.union([
		LlmTextContentSchema,
		LlmThinkingContentSchema,
		LlmToolCallSchema
	]))
});
const LlmToolResultMessageSchema = v.object({
	role: v.literal("toolResult"),
	toolCallId: v.string(),
	toolName: v.string(),
	content: v.array(v.union([LlmTextContentSchema, LlmImageContentSchema])),
	isError: v.boolean()
});
const LlmMessageSchema = v.union([
	LlmUserMessageSchema,
	LlmAssistantMessageSchema,
	LlmToolResultMessageSchema
]);
const LlmToolSchema = v.object({
	name: v.string(),
	description: v.string(),
	parameters: v.unknown()
});
const FLUE_EVENT_TYPES = [
	"run_start",
	"run_resume",
	"agent_start",
	"agent_end",
	"turn_start",
	"turn_request",
	"turn_end",
	"message_start",
	"message_update",
	"message_end",
	"tool_execution_start",
	"tool_execution_update",
	"tool_execution_end",
	"text_delta",
	"thinking_start",
	"thinking_delta",
	"thinking_end",
	"tool_start",
	"tool_call",
	"turn",
	"task_start",
	"task",
	"compaction_start",
	"compaction",
	"operation_start",
	"operation",
	"log",
	"idle",
	"run_end"
];
const FlueEventSchema = v.union([
	flueEvent({
		type: v.literal("run_start"),
		runId: v.string(),
		owner: v.object({
			kind: v.literal("workflow"),
			workflowName: v.string(),
			instanceId: v.string()
		}),
		instanceId: v.string(),
		workflowName: v.string(),
		startedAt: v.string(),
		payload: v.unknown()
	}),
	flueEvent({
		type: v.literal("run_resume"),
		runId: v.string(),
		owner: v.object({
			kind: v.literal("workflow"),
			workflowName: v.string(),
			instanceId: v.string()
		}),
		instanceId: v.string(),
		workflowName: v.string(),
		startedAt: v.string()
	}),
	flueEvent({ type: v.literal("agent_start") }),
	flueEvent({
		type: v.literal("agent_end"),
		messages: v.array(v.any())
	}),
	flueEvent({
		type: v.literal("turn_start"),
		turnId: v.string(),
		purpose: v.picklist([
			"agent",
			"compaction",
			"compaction_prefix"
		])
	}),
	flueEvent({
		type: v.literal("turn_request"),
		turnId: v.string(),
		purpose: v.picklist([
			"agent",
			"compaction",
			"compaction_prefix"
		]),
		model: v.string(),
		provider: v.string(),
		api: v.string(),
		input: v.object({
			systemPrompt: v.optional(v.string()),
			messages: v.array(LlmMessageSchema),
			tools: v.optional(v.array(LlmToolSchema))
		}),
		reasoning: v.optional(v.string())
	}),
	flueEvent({
		type: v.literal("turn_end"),
		turnId: v.string(),
		purpose: v.picklist([
			"agent",
			"compaction",
			"compaction_prefix"
		]),
		message: v.any(),
		toolResults: v.array(v.any())
	}),
	flueEvent({
		type: v.literal("message_start"),
		message: v.any()
	}),
	flueEvent({
		type: v.literal("message_update"),
		message: v.any(),
		assistantMessageEvent: v.unknown()
	}),
	flueEvent({
		type: v.literal("message_end"),
		message: v.any()
	}),
	flueEvent({
		type: v.literal("tool_execution_start"),
		toolCallId: v.string(),
		toolName: v.string(),
		args: v.unknown()
	}),
	flueEvent({
		type: v.literal("tool_execution_update"),
		toolCallId: v.string(),
		toolName: v.string(),
		args: v.unknown(),
		partialResult: v.unknown()
	}),
	flueEvent({
		type: v.literal("tool_execution_end"),
		toolCallId: v.string(),
		toolName: v.string(),
		result: v.unknown(),
		isError: v.boolean()
	}),
	flueEvent({
		type: v.literal("text_delta"),
		text: v.string()
	}),
	flueEvent({ type: v.literal("thinking_start") }),
	flueEvent({
		type: v.literal("thinking_delta"),
		delta: v.string()
	}),
	flueEvent({
		type: v.literal("thinking_end"),
		content: v.string()
	}),
	flueEvent({
		type: v.literal("tool_start"),
		toolName: v.string(),
		toolCallId: v.string(),
		args: v.optional(v.unknown())
	}),
	flueEvent({
		type: v.literal("tool_call"),
		toolName: v.string(),
		toolCallId: v.string(),
		isError: v.boolean(),
		result: v.optional(v.unknown()),
		durationMs: v.number()
	}),
	flueEvent({
		type: v.literal("turn"),
		turnId: v.string(),
		purpose: v.picklist([
			"agent",
			"compaction",
			"compaction_prefix"
		]),
		durationMs: v.number(),
		model: v.optional(v.string()),
		provider: v.optional(v.string()),
		api: v.optional(v.string()),
		output: v.optional(LlmAssistantMessageSchema),
		usage: v.optional(PromptUsageSchema),
		stopReason: v.optional(v.string()),
		isError: v.boolean(),
		error: v.optional(v.unknown())
	}),
	flueEvent({
		type: v.literal("task_start"),
		taskId: v.string(),
		prompt: v.string(),
		agent: v.optional(v.string()),
		cwd: v.optional(v.string())
	}),
	flueEvent({
		type: v.literal("task"),
		taskId: v.string(),
		agent: v.optional(v.string()),
		isError: v.boolean(),
		result: v.optional(v.unknown()),
		durationMs: v.number()
	}),
	flueEvent({
		type: v.literal("compaction_start"),
		reason: v.picklist([
			"threshold",
			"overflow",
			"manual"
		]),
		estimatedTokens: v.number()
	}),
	flueEvent({
		type: v.literal("compaction"),
		messagesBefore: v.number(),
		messagesAfter: v.number(),
		durationMs: v.number(),
		usage: v.optional(PromptUsageSchema)
	}),
	flueEvent({
		type: v.literal("operation_start"),
		operationId: v.string(),
		operationKind: v.picklist([
			"prompt",
			"skill",
			"task",
			"shell",
			"compact"
		])
	}),
	flueEvent({
		type: v.literal("operation"),
		operationId: v.string(),
		operationKind: v.picklist([
			"prompt",
			"skill",
			"task",
			"shell",
			"compact"
		]),
		durationMs: v.number(),
		isError: v.boolean(),
		error: v.optional(v.unknown()),
		result: v.optional(v.unknown()),
		usage: v.optional(PromptUsageSchema)
	}),
	flueEvent({
		type: v.literal("log"),
		level: v.picklist([
			"info",
			"warn",
			"error"
		]),
		message: v.string(),
		attributes: v.optional(v.record(v.string(), v.unknown()))
	}),
	flueEvent({ type: v.literal("idle") }),
	flueEvent({
		type: v.literal("run_end"),
		runId: v.string(),
		result: v.optional(v.unknown()),
		isError: v.boolean(),
		error: v.optional(v.unknown()),
		durationMs: v.number()
	})
]);
const RunEventListResponseSchema = v.object({ events: v.array(FlueEventSchema) });
const AgentInvocationResponseSchema = v.object({ result: v.unknown() });
const WorkflowInvocationResponseSchema = v.object({
	result: v.unknown(),
	_meta: v.object({ runId: v.string() })
});
const WorkflowAdmissionResponseSchema = v.object({
	status: v.literal("accepted"),
	runId: v.string()
});
const integerString = (message) => v.pipe(v.string(), v.regex(/^\d+$/, message));
const eventTypesPattern = new RegExp(`^(${FLUE_EVENT_TYPES.join("|")})(,(${FLUE_EVENT_TYPES.join("|")}))*$`);
const RunEventsQuerySchema = v.object({
	after: v.optional(integerString("after must be a non-negative integer.")),
	types: v.optional(v.pipe(v.string(), v.regex(eventTypesPattern, "types must be a comma-separated list of known event type names."))),
	limit: v.optional(v.pipe(integerString("limit must be an integer between 1 and 1000."), v.transform(Number), v.minValue(1, "limit must be at least 1."), v.maxValue(1e3, "limit must be at most 1000.")))
});
const RunIdParamSchema = v.object({ runId: v.string() });
const WorkflowRouteParamSchema = v.object({ name: v.string() });
const WorkflowInvocationQuerySchema = v.object({ wait: v.optional(v.literal("result")) });
const AgentRouteParamSchema = v.object({
	name: v.string(),
	id: v.string()
});
const AgentManifestEntrySchema = v.object({
	name: v.string(),
	transports: v.object({
		http: v.optional(v.literal(true)),
		websocket: v.optional(v.literal(true))
	}),
	created: v.boolean()
});
const ListAgentsResponseSchema = v.object({ items: v.array(AgentManifestEntrySchema) });
const ListRunsResponseSchema = v.object({
	items: v.array(RunPointerSchema),
	nextCursor: v.optional(v.string())
});
const ListLimitSchema = v.optional(v.pipe(integerString("limit must be an integer between 1 and 1000."), v.transform(Number), v.minValue(1, "limit must be at least 1."), v.maxValue(1e3, "limit must be at most 1000.")));
const AdminRunsQuerySchema = v.object({
	status: v.optional(RunStatusSchema),
	workflowName: v.optional(v.string()),
	cursor: v.optional(v.string()),
	limit: ListLimitSchema
});
//#endregion
//#region src/runtime/flue-app.ts
const RUN_ROUTES_BY_ID = [
	["/runs/:runId", "get"],
	["/runs/:runId/events", "events"],
	["/runs/:runId/stream", "stream"]
];
async function dispatch(agentOrRequest, maybeRequest) {
	const rt = runtimeConfig;
	if (!rt) throw new Error("[flue] dispatch() called before runtime was configured. This usually means it was used outside a Flue-built server entry.");
	if (!rt.dispatchQueue) throw new Error("[flue] dispatch() cannot be accepted because no dispatch queue is configured.");
	return enqueueDispatch({
		request: isCreatedAgentValue(agentOrRequest) ? resolveCreatedAgentDispatchRequest(agentOrRequest, maybeRequest, rt) : agentOrRequest,
		dispatchQueue: rt.dispatchQueue,
		rt
	});
}
function isCreatedAgentValue(value) {
	return "__flueCreatedAgent" in value && value.__flueCreatedAgent === true && typeof value.initialize === "function";
}
function resolveCreatedAgentDispatchRequest(agent, request, rt) {
	if (!request) throw new Error("[flue] dispatch(agent, request) requires a dispatch request.");
	const name = rt.resolveDispatchAgentName?.(agent);
	if (!name) throw new Error("[flue] dispatch() target created agent is not a discovered default-exported agent in this built application.");
	return {
		agent: name,
		id: request.id,
		session: request.session,
		input: request.input
	};
}
let runtimeConfig;
/**
* Not part of the public API — exposed via `@flue/runtime/internal` only
* because the generated entry imports it from a stable bare specifier.
*/
function configureFlueRuntime(cfg) {
	runtimeConfig = cfg;
	configureErrorRendering({ devMode: cfg.devMode ?? false });
}
function resetFlueRuntimeForTests() {
	runtimeConfig = void 0;
	configureErrorRendering({ devMode: false });
}
function getFlueRuntime() {
	return runtimeConfig;
}
/**
* Creates a mountable Hono sub-app for Flue's public HTTP and WebSocket API.
* Routes are relative to the application-chosen mount prefix.
*
* The mounted sub-app exposes:
*
* - `GET /openapi.json`
* - `POST /agents/:name/:id` and `GET /agents/:name/:id` WebSocket upgrades
* - `POST /workflows/:name` and `GET /workflows/:name` WebSocket upgrades
* - `GET /runs/:runId`
* - `GET /runs/:runId/events`
* - `GET /runs/:runId/stream`
*
* Agent and workflow routes are available only when the corresponding module
* opts into that transport. Run routes inspect workflow runs only and may
* expose payloads, results, errors, and events; applications publishing them
* should authorize access to the selected run.
*/
function flue() {
	const app = new Hono();
	app.get("/openapi.json", lazyOpenApiRouteHandler(app, publicOpenApiOptions));
	app.post("/workflows/:name", describeRoute(workflowRouteSpec()), validated("param", WorkflowRouteParamSchema), validated("query", WorkflowInvocationQuerySchema), workflowRouteHandler);
	app.get("/workflows/:name", workflowSocketRouteHandler);
	app.all("/workflows/:name", workflowRouteHandler);
	app.post("/agents/:name/:id", describeRoute(agentRouteSpec()), validated("param", AgentRouteParamSchema), agentRouteHandler);
	app.get("/agents/:name/:id", agentSocketRouteHandler);
	app.all("/agents/:name/:id", agentRouteHandler);
	for (const [routePath, action] of RUN_ROUTES_BY_ID) {
		if (action === "events") app.get(routePath, describeRoute(runRouteSpec(action)), validated("param", RunIdParamSchema), validated("query", RunEventsQuerySchema), runByIdRouteHandler(action));
		else app.get(routePath, describeRoute(runRouteSpec(action)), validated("param", RunIdParamSchema), runByIdRouteHandler(action));
		app.all(routePath, runByIdRouteHandler(action));
	}
	app.onError((err) => toHttpResponse(err));
	return app;
}
/**
* Build the default outer Hono app used when no user `app.ts` is
* present. Mounts `flue()` at root, renders canonical Flue envelopes
* for unmatched paths and any thrown errors.
*
* Lives in @flue/runtime rather than the generated entry so that user
* projects on the Cloudflare target — whose `node_modules` does not
* declare `hono` directly — don't have to add it themselves just to
* keep the no-`app.ts` default behavior working. When a user does
* write an `app.ts`, they own this composition and must `pnpm add
* hono` (or equivalent) themselves.
*/
function createDefaultFlueApp() {
	const app = new Hono();
	app.route("/", flue());
	app.notFound((c) => {
		throw new RouteNotFoundError({
			method: c.req.method,
			path: new URL(c.req.url).pathname
		});
	});
	app.onError((err) => toHttpResponse(err));
	return app;
}
function publicOpenApiOptions() {
	return { documentation: {
		info: {
			title: "Flue Public API",
			version: runtimeConfig?.runtimeVersion ?? "0.0.0",
			description: "Public Flue agent invocation and workflow run inspection API."
		},
		servers: []
	} };
}
function validated(target, schema) {
	return validator(target, schema, (result) => {
		if (result.success) return;
		throw new ValidationError({
			details: `Invalid ${target} parameters.`,
			issues: result.error
		});
	});
}
function jsonResponse(schema, description) {
	return {
		description,
		content: { "application/json": { schema: resolver(schema) } }
	};
}
function errorResponses() {
	return {
		400: jsonResponse(ErrorEnvelopeSchema, "Validation or request-shape error."),
		404: jsonResponse(ErrorEnvelopeSchema, "Resource or route not found."),
		405: jsonResponse(ErrorEnvelopeSchema, "HTTP method is not allowed."),
		415: jsonResponse(ErrorEnvelopeSchema, "Request body must be JSON."),
		500: jsonResponse(ErrorEnvelopeSchema, "Internal server error."),
		501: jsonResponse(ErrorEnvelopeSchema, "Runtime feature is not configured.")
	};
}
function workflowRouteSpec() {
	return {
		tags: ["workflows"],
		operationId: "invokeWorkflow",
		summary: "Start a workflow run",
		description: "Starts the named HTTP-exposed workflow through one admitted execution path. By default it returns an accepted run id; use ?wait=result for a synchronous JSON observation or Accept: text/event-stream to observe live run events while the connection remains available.",
		requestBody: {
			required: false,
			content: { "application/json": { schema: {
				type: "object",
				additionalProperties: true,
				description: "Workflow-defined payload. Consult the target workflow documentation."
			} } }
		},
		responses: {
			202: jsonResponse(WorkflowAdmissionResponseSchema, "Workflow run accepted."),
			200: {
				description: "Workflow result envelope or server-sent events stream, depending on the requested mode.",
				content: {
					"application/json": { schema: resolver(WorkflowInvocationResponseSchema) },
					"text/event-stream": { schema: {
						type: "string",
						description: "SSE-framed FlueEvent values. A terminal stream-infrastructure event: error frame has data { error: { type, message, details, dev?, meta? } }."
					} }
				}
			},
			...errorResponses()
		},
		"x-flue-invocation-modes": [
			"accepted",
			"wait-result",
			"stream"
		],
		"x-flue-user-defined": true
	};
}
function agentRouteSpec() {
	return {
		tags: ["agents"],
		operationId: "invokeAgent",
		summary: "Invoke an agent instance",
		description: "Prompts the named agent instance as an attached interaction. Use dispatch(...) from application code for asynchronous delivery.",
		requestBody: {
			required: true,
			content: { "application/json": { schema: {
				type: "object",
				required: ["message"],
				properties: {
					message: { type: "string" },
					session: {
						type: "string",
						minLength: 1,
						pattern: ".*\\S.*"
					}
				}
			} } }
		},
		responses: {
			200: {
				description: "Attached prompt result or server-sent events stream, depending on the requested mode.",
				content: {
					"application/json": { schema: resolver(AgentInvocationResponseSchema) },
					"text/event-stream": { schema: {
						type: "string",
						description: "SSE frames for attached agent events correlated by instanceId without workflow run identity. A terminal event: error frame has data { type: \"error\", instanceId, error: { type, message, details, dev?, meta? } }."
					} }
				}
			},
			...errorResponses()
		},
		"x-flue-invocation-modes": ["sync", "stream"],
		"x-flue-user-defined": true
	};
}
function runRouteSpec(action) {
	if (action === "stream") return {
		tags: ["runs"],
		operationId: "streamRunEvents",
		summary: "Stream workflow run events",
		responses: {
			200: {
				description: "Server-sent events stream of workflow run lifecycle and nested agent events.",
				content: { "text/event-stream": { schema: {
					type: "string",
					description: "SSE-framed workflow run FlueEvent values. A terminal stream-infrastructure event: error frame has data { error: { type, message, details, dev?, meta? } }."
				} } }
			},
			...errorResponses()
		},
		"x-flue-streaming": true
	};
	return {
		tags: ["runs"],
		operationId: action === "get" ? "getRun" : "listRunEvents",
		summary: action === "get" ? "Get a workflow run record" : "List workflow run events",
		responses: {
			200: jsonResponse(action === "get" ? RunRecordSchema : RunEventListResponseSchema, action === "get" ? "Run record." : "Persisted run event page."),
			...errorResponses()
		}
	};
}
const workflowSocketRouteHandler = async (c, next) => {
	if (!isWebSocketUpgrade(c.req.raw)) return next();
	const rt = requiredRuntime();
	const name = c.req.param("name") ?? "";
	if (!registeredWorkflowsForTransport(rt, "websocket").includes(name)) throw new RouteNotFoundError({
		method: c.req.method,
		path: new URL(c.req.url).pathname
	});
	return runAttachedMiddleware(c, rt.workflowWebSocketMiddleware?.[name], async () => {
		if (rt.target === "node") {
			if (!rt.nodeWebSocketWorkflowRoute) throw new Error("[flue] Node runtime is missing WebSocket workflow routing.");
			return await rt.nodeWebSocketWorkflowRoute(c, next) ?? void 0;
		}
		if (!rt.routeWorkflowRequest) throw new Error("[flue] Cloudflare runtime is missing workflow route forwarding.");
		const response = await rt.routeWorkflowRequest(normalizeAttachedRequest(c.req.raw, `/workflows/${encodeURIComponent(name)}`), c.env, {
			workflowName: name,
			instanceId: generateWorkflowRunId(name)
		});
		if (response) return response;
		throw new RouteNotFoundError({
			method: c.req.method,
			path: new URL(c.req.url).pathname
		});
	});
};
const agentSocketRouteHandler = async (c, next) => {
	if (!isWebSocketUpgrade(c.req.raw)) return next();
	const rt = requiredRuntime();
	const name = c.req.param("name") ?? "";
	const id = c.req.param("id") ?? "";
	if (!registeredAgentsForTransport(rt, "websocket").includes(name) || id.trim() === "") throw new RouteNotFoundError({
		method: c.req.method,
		path: new URL(c.req.url).pathname
	});
	return runAttachedMiddleware(c, rt.agentWebSocketMiddleware?.[name], async () => {
		if (rt.target === "node") {
			if (!rt.nodeWebSocketAgentRoute) throw new Error("[flue] Node runtime is missing WebSocket agent routing.");
			return await rt.nodeWebSocketAgentRoute(c, next) ?? void 0;
		}
		if (!rt.routeAgentRequest) throw new Error("[flue] Cloudflare runtime is missing agent route forwarding.");
		const response = await rt.routeAgentRequest(normalizeAttachedRequest(c.req.raw, `/agents/${encodeURIComponent(name)}/${encodeURIComponent(id)}`), c.env);
		if (response) return response;
		throw new RouteNotFoundError({
			method: c.req.method,
			path: new URL(c.req.url).pathname
		});
	});
};
const workflowRouteHandler = async (c) => {
	const rt = runtimeConfig;
	if (!rt) throw new Error("[flue] flue() route invoked before runtime was configured. This usually means flue() was used outside a Flue-built server entry.");
	const name = c.req.param("name") ?? "";
	const workflows = rt.manifest?.workflows ?? [];
	validateWorkflowRequest({
		method: c.req.method,
		name,
		registeredWorkflows: workflows.map((workflow) => workflow.name),
		httpWorkflows: registeredWorkflowsForTransport(rt, "http")
	});
	const request = c.req.raw.clone();
	return runAttachedMiddleware(c, rt.workflowRouteMiddleware?.[name], async () => {
		if (rt.target === "node") {
			const handler = rt.workflowHandlers?.[name];
			const createContext = rt.createContext;
			if (!handler || !createContext) throw new Error("[flue] Node runtime is missing workflow handler configuration.");
			return handleWorkflowRequest({
				request,
				workflowName: name,
				handler,
				createContext,
				startWorkflowAdmission: rt.startWorkflowAdmission,
				runStore: rt.runStore,
				runSubscribers: rt.runSubscribers,
				runRegistry: rt.runRegistry
			});
		}
		if (!rt.routeWorkflowRequest) throw new Error("[flue] Cloudflare runtime is missing workflow route forwarding.");
		const response = await rt.routeWorkflowRequest(request, c.env, {
			workflowName: name,
			instanceId: generateWorkflowRunId(name)
		});
		if (response) return response;
		throw new RouteNotFoundError({
			method: c.req.method,
			path: new URL(c.req.url).pathname
		});
	});
};
const agentRouteHandler = async (c) => {
	const rt = runtimeConfig;
	if (!rt) throw new Error("[flue] flue() route invoked before runtime was configured. This usually means flue() was used outside a Flue-built server entry.");
	const name = c.req.param("name") ?? "";
	const id = c.req.param("id") ?? "";
	validateAgentRequest({
		method: c.req.method,
		name,
		id,
		registeredAgents: registeredAgentsForTransport(rt, "http")
	});
	const request = c.req.raw.clone();
	return runAttachedMiddleware(c, rt.agentRouteMiddleware?.[name], async () => {
		if (rt.target === "node") {
			const handler = rt.handlers?.[name];
			const createContext = rt.createContext;
			if (!handler || !createContext) throw new Error("[flue] Node runtime is missing agent handler configuration.");
			return handleAgentRequest({
				request,
				agentName: name,
				id,
				handler,
				createContext,
				runHandler: rt.runHandler
			});
		}
		if (!rt.routeAgentRequest) throw new Error("[flue] Cloudflare runtime is missing agent route forwarding.");
		const response = await rt.routeAgentRequest(request, c.env);
		if (response) return response;
		throw new RouteNotFoundError({
			method: c.req.method,
			path: new URL(c.req.url).pathname
		});
	});
};
function runByIdRouteHandler(action) {
	return async (c) => {
		const rt = runtimeConfig;
		if (!rt) throw new Error("[flue] flue() route invoked before runtime was configured. This usually means flue() was used outside a Flue-built server entry.");
		if (c.req.method !== "GET") throw new MethodNotAllowedError({
			method: c.req.method,
			allowed: ["GET"]
		});
		const runId = c.req.param("runId") || void 0;
		if (!runId) throw new RouteNotFoundError({
			method: c.req.method,
			path: new URL(c.req.url).pathname
		});
		return handleRunById({
			rt,
			request: c.req.raw,
			env: c.env,
			runId,
			action
		});
	};
}
async function handleRunById(opts) {
	const { rt, request, env, runId, action } = opts;
	if (rt.target === "cloudflare") {
		if (!rt.createRunRegistryForRequest || !rt.routeRunRequest) throw new RunRegistryUnavailableError();
		const registry = rt.createRunRegistryForRequest(env);
		if (!registry) throw new RunRegistryUnavailableError();
		const pointer = await registry.lookupRun(runId);
		if (!pointer) throw new RunNotFoundError({ runId });
		const response = await rt.routeRunRequest(normalizeRunRequest(request, runId, action), env, pointer.owner);
		if (response) return response;
		throw new RouteNotFoundError({
			method: request.method,
			path: new URL(request.url).pathname
		});
	}
	if (!rt.runRegistry) throw new RunRegistryUnavailableError();
	const pointer = await rt.runRegistry.lookupRun(runId);
	if (!pointer) throw new RunNotFoundError({ runId });
	return handleRunRouteRequest({
		request,
		runStore: rt.runStore,
		runSubscribers: rt.runSubscribers,
		owner: pointer.owner,
		runId,
		action
	});
}
function lazyOpenApiRouteHandler(app, getOptions) {
	return (c, next) => openAPIRouteHandler(app, getOptions())(c, next);
}
function normalizeRunRequest(request, runId, action) {
	const url = new URL(request.url);
	url.pathname = action === "events" ? `/runs/${encodeURIComponent(runId)}/events` : action === "stream" ? `/runs/${encodeURIComponent(runId)}/stream` : `/runs/${encodeURIComponent(runId)}`;
	return new Request(url, request);
}
function requiredRuntime() {
	if (!runtimeConfig) throw new Error("[flue] flue() route invoked before runtime was configured. This usually means flue() was used outside a Flue-built server entry.");
	return runtimeConfig;
}
async function runAttachedMiddleware(c, middleware, handle) {
	if (!middleware) return handle();
	const finalizedBefore = c.finalized;
	const responseBefore = finalizedBefore ? c.res : void 0;
	let continued = false;
	const response = await middleware(c, async () => {
		if (continued) throw new Error("next() called multiple times");
		continued = true;
		const handled = await handle();
		if (handled) c.res = handled;
	});
	if (response) return response;
	if (continued || c.finalized && (!finalizedBefore || c.res !== responseBefore)) return c.res;
	throw new Error("Context is not finalized. Did you forget to return a Response object or await next()?");
}
function isWebSocketUpgrade(request) {
	return request.method === "GET" && request.headers.get("upgrade")?.toLowerCase() === "websocket";
}
function normalizeAttachedRequest(request, pathname) {
	const url = new URL(request.url);
	url.pathname = pathname;
	return new Request(url, request);
}
function registeredAgentsForTransport(rt, transport) {
	return (rt.manifest?.agents ?? []).filter((agent) => agent.transports[transport] === true).map((agent) => agent.name);
}
function registeredWorkflowsForTransport(rt, transport) {
	return (rt.manifest?.workflows ?? []).filter((workflow) => workflow.transports[transport] === true).map((workflow) => workflow.name);
}
//#endregion
export { getFlueRuntime as a, registeredWorkflowsForTransport as c, ErrorEnvelopeSchema as d, ListAgentsResponseSchema as f, RunRecordSchema as h, flue as i, resetFlueRuntimeForTests as l, RunIdParamSchema as m, createDefaultFlueApp as n, handleRunById as o, ListRunsResponseSchema as p, dispatch as r, registeredAgentsForTransport as s, configureFlueRuntime as t, AdminRunsQuerySchema as u };
