// @flue/runtime@0.9.2/dist/handle-agent-C2C75lpq.mjs
import { r as generateWorkflowRunId } from "./ids-DiSJFsle.mjs";
//#region src/errors.ts
/**
* Complete error framework for Flue.
*
* This file contains both the error vocabulary (concrete error classes) and
* the framework utilities (renderers, type guards, request parsing helpers).
* Previously split across `errors.ts` and `error-utils.ts`, but consolidated
* for better LLM comprehension.
*
* ──── Why this file exists ────────────────────────────────────────────────
*
* Concentrating every error in one file is deliberate. When all errors are
* visible together, it's easy to:
*
*   - Keep message tone and detail level consistent across the codebase.
*   - Notice duplicates ("oh, we already have an error for this case").
*   - Establish norms by example — when adding a new error, look at the
*     neighbors above and copy the pattern.
*
* Application code throughout the codebase should reach for one of these
* classes rather than constructing a `FlueError` ad hoc. If no existing class
* fits, add one here. That's the entire convention.
*
* ──── Two audiences: caller vs. developer ─────────────────────────────────
*
* The reader of an error message is one of two distinct audiences:
*
*   - The *caller*: an HTTP client. Possibly third-party, possibly hostile,
*     possibly an end user who shouldn't even know we're built on Flue.
*     Sees `message` and `details` always.
*
*   - The *developer*: the human running the service (`flue dev`, `flue run`,
*     local debugging). Sees `dev` in addition, but only when the generated
*     runtime is configured for local development.
*
* Every error class must classify its prose by audience. The required-but-
* possibly-empty shape of both `details` and `dev` is the discipline:
* forgetting either field is a TypeScript error, and writing `''` is a
* deliberate "I have nothing for that audience" decision.
*
* Concretely:
*
*   - `message`     One sentence. Caller-safe. Always rendered.
*   - `details`     Longer caller-safe prose. About the request itself, the
*                   contract, what the caller can do to fix it. Always
*                   rendered. NEVER includes:
*                     - sibling/neighbor enumeration (leaks namespace)
*                     - filesystem paths or "agents/" / "skills/" / etc.
*                       (leaks framework internals)
*                     - source-code-level fix instructions ("add ... to your
*                       agent definition") (caller can't act on these)
*                     - build-time or runtime mechanics
*   - `dev`         Longer dev-audience prose. Available alternatives,
*                   filesystem layout, framework guidance, source-code-level
*                   fix instructions. Rendered ONLY in local development.
*
* When in doubt, put information in `dev`. The default is conservative.
*
* ──── Conventions for new error classes ───────────────────────────────────
*
*   - Class name: PascalCase, suffixed with `Error`. E.g. `AgentNotFoundError`.
*   - The class owns its `type` constant (snake_case). Set once in the
*     subclass constructor, never passed by callers. Renaming the wire type
*     is then a one-line change.
*   - Constructor takes ONLY structured input data (the values used to build
*     the message). The constructor assembles `message`, `details`, and
*     `dev` from that data, so call sites never reinvent phrasing.
*   - `details` and `dev` are both required strings. Pass `''` only when
*     there's genuinely nothing more to say for that audience.
*   - For HTTP errors, the class sets its own `status` (and `headers` where
*     relevant). Callers do not pick HTTP status codes ad-hoc.
*
* Worked example (matches `AgentNotFoundError` below):
*
*     new AgentNotFoundError({ name, available });
*     // builds:
*     //   message: `Agent "foo" is not registered.`
*     //   details: `Verify the agent name is correct.`
*     //   dev:     `Available agents: "echo", "greeter". Agents are
*     //            loaded from the project root's "agents/" directory at
*     //            build time. ...`
*
* The wire response in production omits `dev`; in `flue dev` / `flue run`
* it includes `dev`. That separation is what lets the dev field be richly
* helpful without leaking namespace state to public callers.
*
* Counter-example to avoid:
*
*     class AgentNotFoundError extends FlueHttpError {
*       constructor(message: string) {                       // ✗ free-form
*         super({                                            // ✗ wrong type
*           type: 'agent_error',
*           message,
*           details: 'Available: "x", "y", "z"',             // ✗ leaks names
*           dev: '',                                         // ✗ wasted field
*           status: 500,                                     // ✗ wrong status
*         });
*       }
*     }
*
* The structured-constructor pattern below is what prevents that drift.
*/
/**
* Format a list of items for inclusion in error details. Empty lists render
* as the supplied fallback (default `(none)`), so messages read naturally
* regardless of whether anything is registered.
*
* Module-private: only used by the concrete error subclasses below. Promote
* to `export` if/when a real cross-file caller appears.
*/
function formatList(items, fallback = "(none)") {
	if (items.length === 0) return fallback;
	return items.map((item) => `"${String(item)}"`).join(", ");
}
/**
* Base class for every error Flue throws. Do not instantiate directly in
* application code — extend it via a subclass below. If a use case isn't
* covered, add a new subclass here rather than throwing a raw `FlueError`.
*/
var FlueError = class extends Error {
	type;
	details;
	dev;
	meta;
	cause;
	constructor(options) {
		super(options.message);
		this.name = "FlueError";
		this.type = options.type;
		this.details = options.details;
		this.dev = options.dev;
		this.meta = options.meta;
		this.cause = options.cause;
	}
};
/**
* Base class for HTTP-layer errors. Adds `status` and optional `headers`.
* Subclasses set these in the `super({...})` call so the call site doesn't
* have to think about HTTP semantics.
*/
var FlueHttpError = class extends FlueError {
	status;
	headers;
	constructor(options) {
		super(options);
		this.name = "FlueHttpError";
		this.status = options.status;
		this.headers = options.headers;
	}
};
var MethodNotAllowedError = class extends FlueHttpError {
	constructor({ method, allowed }) {
		super({
			type: "method_not_allowed",
			message: `HTTP method ${method} is not allowed on this endpoint.`,
			details: `This endpoint accepts ${formatList(allowed)} only.`,
			dev: "",
			status: 405,
			headers: { Allow: allowed.join(", ") }
		});
	}
};
var UnsupportedMediaTypeError = class extends FlueHttpError {
	constructor({ received }) {
		const detailLines = [];
		if (received) detailLines.push(`Received Content-Type: "${received}".`);
		else detailLines.push(`No Content-Type header was sent.`);
		detailLines.push("Send the request body as JSON with the header \"Content-Type: application/json\", or omit the body entirely (and the Content-Type header) if the request doesn't have a payload.");
		super({
			type: "unsupported_media_type",
			message: `Request body must be sent as application/json.`,
			details: detailLines.join("\n"),
			dev: "",
			status: 415
		});
	}
};
var InvalidJsonError = class extends FlueHttpError {
	constructor({ parseError }) {
		super({
			type: "invalid_json",
			message: `Request body is not valid JSON.`,
			details: `The JSON parser reported: ${parseError}\nVerify the body is well-formed JSON, or omit the body entirely if the request doesn't have a payload.`,
			dev: "",
			status: 400
		});
	}
};
var AgentNotFoundError = class extends FlueHttpError {
	constructor({ name, available }) {
		super({
			type: "agent_not_found",
			message: `Agent "${name}" is not registered.`,
			details: `Verify the agent name is correct.`,
			dev: `Available agents: ${formatList(available)}.\nAgents are loaded from the project root's "agents/" directory at build time. Verify the agent file is present in the project root being served.`,
			status: 404
		});
	}
};
var WorkflowNotFoundError = class extends FlueHttpError {
	constructor({ name, available }) {
		super({
			type: "workflow_not_found",
			message: `Workflow "${name}" is not registered.`,
			details: `Verify the workflow name is correct.`,
			dev: `Available workflows: ${formatList(available)}.\nWorkflows are loaded from the project root's "workflows/" directory at build time.`,
			status: 404
		});
	}
};
var WorkflowNotHttpError = class extends FlueHttpError {
	constructor({ name }) {
		super({
			type: "workflow_not_http",
			message: `Workflow "${name}" is not web-accessible.`,
			details: `This endpoint is not exposed over HTTP.`,
			dev: `To expose it, export route middleware and call await next() to enter the workflow handler.`,
			status: 404
		});
	}
};
var RouteNotFoundError = class extends FlueHttpError {
	constructor({ method, path }) {
		super({
			type: "route_not_found",
			message: `No route matches ${method} ${path}.`,
			details: `Agents are served at POST /agents/<name>/<id>.`,
			dev: "",
			status: 404
		});
	}
};
var RunNotFoundError = class extends FlueHttpError {
	constructor({ runId }) {
		super({
			type: "run_not_found",
			message: `Run "${runId}" was not found.`,
			details: "Verify the run id is correct and its history is still available.",
			dev: "",
			status: 404
		});
	}
};
var RunStoreUnavailableError = class extends FlueHttpError {
	constructor() {
		super({
			type: "run_store_unavailable",
			message: "Run history is not available in this runtime.",
			details: "This endpoint requires the generated runtime to be configured with a run store.",
			dev: "",
			status: 501
		});
	}
};
var RunEventTooLargeError = class extends Error {
	constructor() {
		super("[flue:run-store] event payload exceeds the 1 MB persistence limit and will not be persisted.");
		this.name = "RunEventTooLargeError";
	}
};
var RunRegistryUnavailableError = class extends FlueHttpError {
	constructor() {
		super({
			type: "run_registry_unavailable",
			message: "Run lookup is not available in this runtime.",
			details: "This endpoint requires the generated runtime to be configured with a run registry.",
			dev: "",
			status: 501
		});
	}
};
var InvalidRequestError = class extends FlueHttpError {
	constructor({ reason }) {
		super({
			type: "invalid_request",
			message: `Request is malformed.`,
			details: reason,
			dev: "",
			status: 400
		});
	}
};
var ValidationError = class extends FlueHttpError {
	constructor({ details, issues }) {
		super({
			type: "validation_failed",
			message: "Request validation failed.",
			details,
			dev: "",
			meta: { issues },
			status: 400
		});
	}
};
/**
* Error framework utilities: renderers, type guards, request parsing helpers.
*
* Wire envelope (HTTP body + SSE `data:` payload for error events):
*
*     {
*       "error": {
*         "type":    "...",
*         "message": "...",
*         "details": "...",
*         "dev":     "..."   // present only in local/dev mode AND when non-empty
*       }
*     }
*
* Field rules:
*   - `type`, `message`, `details` are always present on the wire.
*   - `dev` is gated by explicit generated-runtime configuration. Even in
*     local development, `dev` is omitted when the error class set it to
*     `''` — so its presence is not a reliable signal of mode by itself;
*     clients should not depend on it that way.
*     See the error classes above for the two-audience rationale.
*   - `meta` is included on the wire only when an error subclass sets it
*     (rare).
*   - `cause` is never included on the wire (it's logged server-side only).
*/
function isFlueError(value) {
	return value instanceof FlueError;
}
/**
* Module-private for now: when an external call site appears we can promote
* to `export` and decide the right shape for `warn`/`info` (FlueError
* subclasses with severity? plain strings? structured data?) — rather than
* committing to a shape now without any usage to validate it.
*/
function formatForLog(prefix, err) {
	if (isFlueError(err)) {
		const lines = [`${prefix} [${err.type}] ${err.message}`];
		if (err.details) for (const line of err.details.split("\n")) lines.push(`  ${line}`);
		if (err.dev) for (const line of err.dev.split("\n")) lines.push(`  ${line}`);
		if (err.cause !== void 0) lines.push(`  cause: ${err.cause instanceof Error ? err.cause.stack ?? err.cause.message : String(err.cause)}`);
		return lines.join("\n");
	}
	if (err instanceof Error) return `${prefix} ${err.stack ?? err.message}`;
	return `${prefix} ${String(err)}`;
}
const flueLog = { error(err) {
	console.error(formatForLog("[flue]", err));
} };
let devMode = false;
function configureErrorRendering(options) {
	devMode = options.devMode;
}
function envelope(err) {
	const out = { error: {
		type: err.type,
		message: err.message,
		details: err.details
	} };
	if (devMode && err.dev) out.error.dev = err.dev;
	if (err.meta) out.error.meta = err.meta;
	return out;
}
const GENERIC_INTERNAL = { error: {
	type: "internal_error",
	message: "An internal error occurred.",
	details: "The server encountered an unexpected error while handling this request."
} };
function toPublicError(err) {
	if (isFlueError(err)) {
		if (!(err instanceof FlueHttpError)) flueLog.error(err);
		return envelope(err).error;
	}
	flueLog.error(err);
	return GENERIC_INTERNAL.error;
}
/**
* Render any thrown value into a `Response` with the canonical Flue error
* envelope. Unknown / non-Flue errors are logged in full and rendered as a
* generic 500 with no message leaked.
*/
function toHttpResponse(err) {
	if (isFlueError(err)) {
		const isHttp = err instanceof FlueHttpError;
		const status = isHttp ? err.status : 500;
		const headers = { "content-type": "application/json" };
		if (isHttp && err.headers) Object.assign(headers, err.headers);
		if (!isHttp) flueLog.error(err);
		return new Response(JSON.stringify(envelope(err)), {
			status,
			headers
		});
	}
	flueLog.error(err);
	return new Response(JSON.stringify(GENERIC_INTERNAL), {
		status: 500,
		headers: { "content-type": "application/json" }
	});
}
/**
* Parse a request body as JSON. Returns `{}` for genuinely empty bodies
* (Content-Length: 0 or missing) so workflow invocations that do not accept
* a payload can be invoked without one.
*
* Throws `UnsupportedMediaTypeError` if a body is present without
* `application/json` content-type, and `InvalidJsonError` if the body is
* present but unparseable.
*/
async function parseJsonBody(request) {
	const contentLengthHeader = request.headers.get("content-length");
	const contentLength = contentLengthHeader === null ? null : Number(contentLengthHeader);
	const contentType = request.headers.get("content-type");
	if (contentLength === 0 || contentLengthHeader === null && contentType === null) return {};
	if (!contentType?.toLowerCase().includes("application/json")) throw new UnsupportedMediaTypeError({ received: contentType });
	let text;
	try {
		text = await request.clone().text();
	} catch (err) {
		throw new InvalidJsonError({ parseError: err instanceof Error ? err.message : String(err) });
	}
	if (text.trim() === "") return {};
	try {
		return JSON.parse(text);
	} catch (err) {
		throw new InvalidJsonError({ parseError: err instanceof Error ? err.message : String(err) });
	}
}
function validateWorkflowRequest(opts) {
	if (opts.method !== "POST") throw new MethodNotAllowedError({
		method: opts.method,
		allowed: ["POST"]
	});
	if (opts.name.trim() === "") throw new InvalidRequestError({ reason: "Workflow URLs must have the shape /workflows/<name> with a non-empty segment." });
	if (!opts.registeredWorkflows.includes(opts.name)) throw new WorkflowNotFoundError({
		name: opts.name,
		available: opts.registeredWorkflows
	});
	if (!opts.httpWorkflows.includes(opts.name)) throw new WorkflowNotHttpError({ name: opts.name });
}
function validateAgentRequest(opts) {
	if (opts.method !== "POST") throw new MethodNotAllowedError({
		method: opts.method,
		allowed: ["POST"]
	});
	if (opts.name.trim() === "" || opts.id.trim() === "") throw new InvalidRequestError({ reason: "Agent URLs must have the shape /agents/<name>/<id> with non-empty segments." });
	if (!opts.registeredAgents.includes(opts.name)) throw new AgentNotFoundError({
		name: opts.name,
		available: opts.registeredAgents
	});
}
//#endregion
//#region src/runtime/dispatch-queue.ts
function assertCurrentDispatchInput(value) {
	if (value && typeof value === "object" && "targetAgent" in value) throw new Error("[flue] Legacy dispatch metadata is unsupported. Clear persisted dispatch state created by an earlier Flue beta.");
}
/**
* Process-lifetime dispatch queue. Acceptance currently means the dispatch was
* validated, appended to this in-memory queue, and scheduled for the configured
* processor. Queued items are lost if the process/isolate exits before the
* placeholder processor runs. Durable backends should implement DispatchQueue
* without changing dispatch() authoring semantics.
*/
var InMemoryDispatchQueue = class {
	processor;
	pending = [];
	scheduled = false;
	constructor(processor = noopDispatchProcessor) {
		this.processor = processor;
	}
	async enqueue(input) {
		this.pending.push(input);
		this.scheduleDrain();
		return {
			dispatchId: input.dispatchId,
			acceptedAt: input.acceptedAt
		};
	}
	scheduleDrain() {
		if (this.scheduled) return;
		this.scheduled = true;
		queueMicrotask(() => {
			this.drain().catch((error) => {
				console.error("[flue:dispatch] Dispatch processor failed:", error);
			});
		});
	}
	async drain() {
		try {
			while (this.pending.length > 0) {
				const input = this.pending.shift();
				if (!input) continue;
				try {
					await this.processor.process(input);
				} catch (error) {
					console.error("[flue:dispatch] Dispatch processor failed:", error);
				}
			}
		} finally {
			this.scheduled = false;
			if (this.pending.length > 0) this.scheduleDrain();
		}
	}
};
const noopDispatchProcessor = { process() {} };
//#endregion
//#region src/runtime/handle-run-routes.ts
/** Run-history HTTP endpoints shared by the Node and Cloudflare targets. */
const EVENTS_DEFAULT_LIMIT = 100;
const EVENTS_MAX_LIMIT = 1e3;
const SSE_HEARTBEAT_MS$1 = 15e3;
/** Buffer cap for events published while a live stream is replaying history. */
const REPLAY_BUFFER_CAP = 1e3;
async function handleRunRouteRequest(opts) {
	const store = opts.runStore;
	if (!store) throw new RunStoreUnavailableError();
	switch (opts.action) {
		case "get": return getRun(store, requireRunId(opts.runId), opts.owner);
		case "events": return getRunEvents(opts.request, store, requireRunId(opts.runId), opts.owner);
		case "stream": return streamRunEvents(opts.request, store, opts.runSubscribers, requireRunId(opts.runId), opts.owner);
	}
}
async function getRun(store, runId, owner) {
	return json(await getRunForOwner(store, runId, owner));
}
async function getRunEvents(request, store, runId, owner) {
	await getRunForOwner(store, runId, owner);
	const url = new URL(request.url);
	const after = parseEventIndex(url.searchParams.get("after"));
	const types = parseTypes(url.searchParams.get("types"));
	const limit = parseLimit(url.searchParams.get("limit"), EVENTS_DEFAULT_LIMIT, EVENTS_MAX_LIMIT);
	let events = await store.getEvents(runId, after === void 0 ? void 0 : after + 1);
	if (types) events = events.filter((event) => types.has(event.type));
	return json({ events: events.slice(0, limit) });
}
/**
* Replay durable history, then tail live events for active runs.
* Subscribe-before-replay avoids dropping events produced during the
* store read; eventIndex dedup handles overlap.
*/
async function streamRunEvents(request, store, subscribers, runId, owner) {
	const run = await getRunForOwner(store, runId, owner);
	const lastEventId = parseLastEventId(request.headers.get("last-event-id"));
	const fromIndex = lastEventId === void 0 ? void 0 : lastEventId + 1;
	if (isTerminal(run)) return sseResponse(encodeSseEvents(await store.getEvents(runId, fromIndex)));
	if (!subscribers) throw new Error("[flue] Active run streaming requires a run subscriber registry, but none was configured for this target. Wire one through HandleRunRouteOptions.runSubscribers.");
	return streamReplayThenTail({
		store,
		subscribers,
		runId,
		fromIndex
	});
}
function streamActiveRunEvents(store, subscribers, runId) {
	return streamReplayThenTail({
		store,
		subscribers,
		runId,
		fromIndex: void 0
	});
}
function streamReplayThenTail(opts) {
	const { store, subscribers, runId, fromIndex } = opts;
	const encoder = new TextEncoder();
	let buffer = [];
	let bufferOverflowed = false;
	let replayDone = false;
	let lastSentIndex = fromIndex === void 0 ? void 0 : fromIndex - 1;
	let closed = false;
	let onLiveEvent;
	let onClose;
	const subscriberListener = (event) => {
		if (closed) return;
		if (!replayDone) {
			if (buffer.length >= REPLAY_BUFFER_CAP) {
				bufferOverflowed = true;
				return;
			}
			buffer.push(event);
			return;
		}
		onLiveEvent?.(event);
	};
	const unsubscribe = subscribers.subscribe(runId, subscriberListener);
	return sseResponse(new ReadableStream({
		start(controller) {
			const heartbeat = setInterval(() => {
				if (closed) return;
				try {
					controller.enqueue(encoder.encode(": heartbeat\n\n"));
				} catch {}
			}, SSE_HEARTBEAT_MS$1);
			const close = () => {
				if (closed) return;
				closed = true;
				clearInterval(heartbeat);
				unsubscribe();
				try {
					controller.close();
				} catch {}
			};
			onClose = close;
			const write = (event) => {
				if (closed) return;
				if (typeof event.eventIndex === "number" && lastSentIndex !== void 0 && event.eventIndex <= lastSentIndex) {
					if (event.type === "run_end") close();
					return;
				}
				try {
					controller.enqueue(encoder.encode(encodeSseEvent(event)));
				} catch {
					close();
					return;
				}
				if (typeof event.eventIndex === "number") lastSentIndex = event.eventIndex;
				if (event.type === "run_end") close();
			};
			onLiveEvent = write;
			(async () => {
				try {
					await runReplayPhase({
						store,
						runId,
						fromIndex,
						write,
						drainBuffer: () => {
							const drained = buffer;
							buffer = [];
							return drained;
						},
						getBufferOverflowed: () => bufferOverflowed,
						resetBufferOverflowed: () => {
							bufferOverflowed = false;
						},
						getLastSentIndex: () => lastSentIndex,
						markReplayDone: () => {
							replayDone = true;
						}
					});
				} catch (error) {
					if (closed) return;
					try {
						controller.enqueue(encoder.encode(encodeSseError(error, lastSentIndex)));
					} catch {}
					close();
				}
			})();
		},
		cancel() {
			onClose?.();
		}
	}));
}
async function runReplayPhase(opts) {
	const { store, runId, fromIndex, write, drainBuffer, getBufferOverflowed, resetBufferOverflowed, getLastSentIndex, markReplayDone } = opts;
	const replay = await store.getEvents(runId, fromIndex);
	for (const event of replay) write(event);
	while (getBufferOverflowed()) {
		resetBufferOverflowed();
		const lastSent = getLastSentIndex();
		const refetchFrom = lastSent === void 0 ? void 0 : lastSent + 1;
		const refetched = await store.getEvents(runId, refetchFrom);
		for (const event of refetched) write(event);
	}
	const buffered = drainBuffer();
	for (const event of buffered) write(event);
	markReplayDone();
}
async function getRunForOwner(store, runId, owner) {
	const run = await store.getRun(runId);
	if (!run) throw new RunNotFoundError({ runId });
	if (!sameOwner(run.owner, owner)) throw new RunNotFoundError({ runId });
	return run;
}
function sameOwner(left, right) {
	return left.workflowName === right.workflowName && left.instanceId === right.instanceId;
}
function isTerminal(run) {
	return run.status === "completed" || run.status === "errored";
}
function encodeSseEvents(events) {
	return events.map(encodeSseEvent).join("");
}
function encodeSseEvent(event) {
	const id = typeof event.eventIndex === "number" ? event.eventIndex : 0;
	return [
		`event: ${event.type}`,
		`id: ${id}`,
		`data: ${JSON.stringify(event)}`,
		"",
		""
	].join("\n");
}
function encodeSseError(error, lastSentIndex) {
	const data = { error: toPublicError(error) };
	return [
		`event: error`,
		`id: ${lastSentIndex ?? 0}`,
		`data: ${JSON.stringify(data)}`,
		"",
		""
	].join("\n");
}
function sseResponse(body) {
	return new Response(body, { headers: {
		"content-type": "text/event-stream",
		"cache-control": "no-cache",
		connection: "keep-alive"
	} });
}
function requireRunId(runId) {
	if (!runId) throw new InvalidRequestError({ reason: "Run id is required for this endpoint." });
	return runId;
}
function parseTypes(value) {
	if (!value) return void 0;
	const types = value.split(",").map((type) => type.trim()).filter(Boolean);
	return types.length > 0 ? new Set(types) : void 0;
}
function parseLimit(value, defaultLimit, maxLimit) {
	if (!value) return defaultLimit;
	const parsed = Number.parseInt(value, 10);
	if (!Number.isFinite(parsed) || parsed <= 0) return defaultLimit;
	return Math.min(parsed, maxLimit);
}
function parseEventIndex(value) {
	if (!value) return void 0;
	const parsed = Number.parseInt(value, 10);
	if (!Number.isFinite(parsed) || parsed < 0) return void 0;
	return parsed;
}
/**
* `Last-Event-ID` is the standard SSE reconnect header. Browsers send the
* last `id:` field they saw; the server uses it to resume from that point.
* Malformed values are ignored — equivalent to no header.
*/
function parseLastEventId(value) {
	if (!value || !/^\d+$/.test(value)) return void 0;
	const parsed = Number(value);
	if (!Number.isSafeInteger(parsed)) return void 0;
	return parsed;
}
function json(data) {
	return new Response(JSON.stringify(data), { headers: { "content-type": "application/json" } });
}
//#endregion
//#region src/runtime/handle-agent.ts
function createAgentDispatchProcessor(options) {
	return { async process(input) {
		assertCurrentDispatchInput(input);
		const agent = options.agents[input.agent];
		if (!agent) throw new Error(`[flue] dispatch target agent "${input.agent}" has no created agent.`);
		const releaseSessionLock = await reserveDispatchAgentSession({
			agentName: input.agent,
			instanceId: input.id
		}, input);
		try {
			const ctx = options.createContext(input.id, void 0, input, dispatchRequest(), void 0, input.dispatchId);
			await createDispatchAgentHandler(agent, input)(ctx);
		} finally {
			releaseSessionLock();
		}
	} };
}
async function validateAgentDispatchAdmission(options) {
	const { input } = options;
	assertCurrentDispatchInput(input);
	if (!isDispatchInput(input)) throw new Error("[flue] Internal dispatch admission received an invalid payload.");
	return {
		dispatchId: input.dispatchId,
		acceptedAt: input.acceptedAt
	};
}
function createDispatchAgentHandler(agent, input) {
	return (ctx) => processAgentDispatch(ctx, agent, input);
}
async function reserveDispatchAgentSession(target, payload) {
	return waitForAgentSessionLock(target, payload);
}
async function processAgentDispatch(ctx, agent, input) {
	const session = await (await ctx.initializeCreatedAgent(agent, void 0)).session(input.session);
	if (!isDispatchSession(session)) throw new Error("[flue] Internal session does not support dispatch input processing.");
	return session.processDispatchInput(input);
}
function isDispatchInput(value) {
	if (!value || typeof value !== "object") return false;
	const input = value;
	return typeof input.dispatchId === "string" && input.dispatchId.trim() !== "" && typeof input.agent === "string" && input.agent.trim() !== "" && typeof input.id === "string" && input.id.trim() !== "" && typeof input.session === "string" && input.session.trim() !== "" && input.input !== void 0 && typeof input.acceptedAt === "string" && input.acceptedAt.trim() !== "";
}
function dispatchRequest() {
	return new Request("http://flue.local/_dispatch", { method: "POST" });
}
function isDispatchSession(value) {
	return !!value && typeof value === "object" && typeof value.processDispatchInput === "function";
}
function createDirectAgentHandler(agent) {
	return async (ctx) => {
		const payload = parseDirectAgentPayload(ctx.payload);
		const session = await (await ctx.initializeCreatedAgent(agent, void 0)).session(payload.session);
		if (!isDirectRequestSession(session)) throw new Error("[flue] Internal session does not support direct input processing.");
		return session.processDirectInput({ message: payload.message });
	};
}
function isDirectRequestSession(value) {
	return !!value && typeof value === "object" && typeof value.processDirectInput === "function";
}
function parseDirectAgentPayload(payload) {
	const expected = "Direct agent requests must use JSON object body { \"message\": string, \"session\"?: string }.";
	if (!payload || typeof payload !== "object" || Array.isArray(payload)) throw new InvalidRequestError({ reason: expected });
	const value = payload;
	if (typeof value.message !== "string") throw new InvalidRequestError({ reason: expected });
	if (value.session !== void 0 && (typeof value.session !== "string" || value.session.trim() === "")) throw new InvalidRequestError({ reason: "Direct agent request \"session\" must be a non-empty string when provided." });
	return {
		message: value.message,
		session: value.session
	};
}
/**
* Handle one attached `/agents/:name/:id` prompt interaction.
*
* `Accept: text/event-stream` returns attached event streaming; otherwise the
* response is synchronous JSON `{ result }`. Former `X-Webhook: true` agent
* requests are rejected because asynchronous delivery uses `dispatch(...)`.
*
* Errors thrown before streaming starts are returned as regular HTTP error
* responses; errors thrown after SSE begins are framed as stream errors.
*/
async function handleAgentRequest(opts) {
	const { request, agentName, id, handler, createContext } = opts;
	const runHandler = opts.runHandler ?? defaultRunHandler;
	try {
		const rawPayload = await parseJsonBody(request);
		if (request.headers.get("x-webhook") === "true") throw new InvalidRequestError({ reason: "Direct agent prompts are attached interactions. Use dispatch(...) for asynchronous delivery." });
		const directOptions = {
			agentName,
			id,
			handler,
			payload: parseDirectAgentPayload(rawPayload),
			request,
			createContext,
			runHandler
		};
		if ((request.headers.get("accept") || "").includes("text/event-stream")) return runDirectSseMode(directOptions);
		return runDirectSyncMode(directOptions);
	} catch (err) {
		return toHttpResponse(err);
	}
}
async function handleWorkflowRequest(opts) {
	const { request, workflowName, handler, createContext, runStore, runSubscribers, runRegistry } = opts;
	const startWorkflowAdmission = opts.startWorkflowAdmission ?? defaultStartWorkflowAdmission;
	const runId = opts.runId ?? generateWorkflowRunId(workflowName);
	const instanceId = runId;
	try {
		const payload = await parseJsonBody(request);
		const isSSE = (request.headers.get("accept") || "").includes("text/event-stream");
		const wait = new URL(request.url).searchParams.get("wait");
		const owner = {
			kind: "workflow",
			workflowName,
			instanceId
		};
		if (isSSE && wait !== "result" && !runSubscribers) throw new Error("[flue] Workflow SSE requires a run subscriber registry.");
		const execution = await prepareWorkflowExecution({
			owner,
			id: runId,
			runId,
			handler,
			payload,
			request,
			createContext,
			startWorkflowAdmission,
			runStore,
			runSubscribers,
			runRegistry
		});
		if (isSSE && wait !== "result") return await runSseMode(execution);
		if (wait === "result") return await runSyncMode(execution);
		return await runWorkflowAdmissionMode(execution);
	} catch (err) {
		const response = toHttpResponse(err);
		response.headers.set("X-Flue-Run-Id", runId);
		return response;
	}
}
const activeAttachedAgentSessions = /* @__PURE__ */ new Map();
async function waitForAgentSessionLock(target, payload) {
	while (true) try {
		return acquireDirectAgentSessionLock(target.agentName, target.instanceId, payload) ?? (() => {});
	} catch (error) {
		if (!(error instanceof InvalidRequestError) || error.details !== "This agent session already has an active prompt.") throw error;
		await new Promise((resolve) => setTimeout(resolve, 0));
	}
}
async function prepareWorkflowExecution(opts) {
	const { owner, id, runId, handler, payload, request, createContext, startWorkflowAdmission, runStore, runSubscribers, runRegistry, onAdmitted, onEvent, emitIdleOnComplete } = opts;
	if (!runStore) throw new RunStoreUnavailableError();
	return {
		runId,
		runStore,
		runSubscribers,
		lifecycle: await createWorkflowRunLifecycle({
			owner,
			id,
			runId,
			payload,
			request,
			createContext,
			runStore,
			runSubscribers,
			runRegistry,
			requirePersistedAdmission: true
		}),
		startWorkflowAdmission,
		handler,
		onAdmitted,
		onEvent,
		emitIdleOnComplete
	};
}
function startWorkflowExecution(execution) {
	if (execution.completion) return execution.completion;
	const { runId, lifecycle, handler, startWorkflowAdmission, onEvent, emitIdleOnComplete } = execution;
	let didRun = false;
	let didEmitIdle = false;
	if (onEvent || emitIdleOnComplete) lifecycle.ctx.setEventCallback((event) => {
		if (event.type === "idle") didEmitIdle = true;
		return onEvent?.(event);
	});
	const run = async () => {
		didRun = true;
		try {
			return await withWorkflowRunLifecycle(lifecycle, async () => {
				execution.onAdmitted?.(runId);
				try {
					return await handler(lifecycle.ctx);
				} finally {
					if (emitIdleOnComplete && !didEmitIdle) lifecycle.ctx.emitEvent({ type: "idle" });
				}
			});
		} finally {
			lifecycle.ctx.setEventCallback(void 0);
		}
	};
	let scheduled;
	try {
		scheduled = startWorkflowAdmission(runId, run);
	} catch (error) {
		lifecycle.ctx.setEventCallback(void 0);
		execution.completion = emitRunEnd(lifecycle, {
			isError: true,
			error
		}).then(() => Promise.reject(error));
		execution.completion.catch(() => void 0);
		throw error;
	}
	execution.completion = scheduled.catch(async (error) => {
		if (!didRun) {
			await emitRunEnd(lifecycle, {
				isError: true,
				error
			});
			lifecycle.ctx.setEventCallback(void 0);
		}
		throw error;
	});
	return execution.completion;
}
async function runWorkflowAdmissionMode(execution) {
	try {
		startWorkflowExecution(execution);
	} catch (error) {
		await execution.completion?.catch(() => void 0);
		throw error;
	}
	return new Response(JSON.stringify({
		status: "accepted",
		runId: execution.runId
	}), {
		status: 202,
		headers: {
			"content-type": "application/json",
			"X-Flue-Run-Id": execution.runId
		}
	});
}
async function failRecoveredRun(opts) {
	const events = opts.runStore ? await opts.runStore.getEvents(opts.runId) : [];
	const terminalEvent = findTerminalRunEvent(events);
	const run = await opts.runStore?.getRun(opts.runId);
	if (terminalEvent || run && run.status !== "active") {
		await reconcileTerminalRun(opts, run, terminalEvent, events);
		return;
	}
	if (run) await safeRegistry("recordRunStart(recovery)", () => opts.runRegistry?.recordRunStart({
		runId: opts.runId,
		owner: run.owner,
		startedAt: run.startedAt
	}));
	const initialEventIndex = nextEventIndex(events);
	const startedAt = run?.startedAt ?? (/* @__PURE__ */ new Date()).toISOString();
	const startedAtMs = Date.parse(startedAt);
	const startEvent = events.find((event) => event.type === "run_start");
	const payload = run?.payload !== void 0 ? run.payload : startEvent?.payload;
	const lifecycle = {
		...opts,
		payload,
		ctx: opts.createContext(opts.id, opts.runId, payload, opts.request, initialEventIndex),
		startedAt,
		startedAtMs: Number.isFinite(startedAtMs) ? startedAtMs : Date.now()
	};
	if (startEvent) {
		const flushFanout = subscribeRunFanout(lifecycle);
		emitRunResume(lifecycle);
		await flushFanout();
	}
	await emitRunEnd(lifecycle, {
		isError: true,
		error: opts.error
	});
}
async function reconcileTerminalRun(opts, run, terminalEvent, events) {
	const isError = terminalEvent?.isError ?? run?.isError ?? false;
	const result = terminalEvent?.result ?? run?.result;
	const error = terminalEvent?.error ?? run?.error;
	const endedAt = terminalEvent?.timestamp ?? run?.endedAt ?? (/* @__PURE__ */ new Date()).toISOString();
	const durationMs = terminalEvent?.durationMs ?? run?.durationMs ?? 0;
	if (!terminalEvent && run && run.status !== "active") try {
		await opts.runStore?.appendEvent(opts.runId, {
			type: "run_end",
			runId: opts.runId,
			result: result === void 0 ? null : result,
			isError,
			error,
			durationMs,
			eventIndex: nextEventIndex(events),
			timestamp: endedAt
		});
	} catch (eventError) {
		console.error("[flue:run-store] appendEvent(run_end recovery) failed:", eventError);
	}
	if (terminalEvent && (!run || run.status === "active")) await opts.runStore?.endRun({
		runId: opts.runId,
		endedAt,
		isError,
		durationMs,
		result,
		error
	});
	await safeRegistry("recordRunStart(recovery)", () => opts.runRegistry?.recordRunStart({
		runId: opts.runId,
		owner: run?.owner ?? opts.owner,
		startedAt: run?.startedAt ?? endedAt
	}));
	await safeRegistry("recordRunEnd(recovery)", () => opts.runRegistry?.recordRunEnd({
		runId: opts.runId,
		endedAt,
		durationMs,
		isError
	}));
	opts.runSubscribers?.complete(opts.runId);
}
function findTerminalRunEvent(events) {
	return [...events].reverse().find((event) => event.type === "run_end");
}
function nextEventIndex(events) {
	return events.reduce((next, event) => Math.max(next, (event.eventIndex ?? -1) + 1), 0);
}
const SSE_HEARTBEAT_MS = 15e3;
function runDirectSseMode(opts) {
	const { readable, writable } = new TransformStream();
	const writer = writable.getWriter();
	const encoder = new TextEncoder();
	let closed = false;
	const writeSSE = async (data, eventType) => {
		if (closed) return;
		const eventIndex = getEventIndex(data) ?? 0;
		const lines = [
			`event: ${eventType}`,
			`id: ${eventIndex}`,
			`data: ${typeof data === "string" ? data : JSON.stringify(data)}`,
			"",
			""
		];
		try {
			await writer.write(encoder.encode(lines.join("\n")));
		} catch {}
	};
	const heartbeat = setInterval(() => {
		if (!closed) writer.write(encoder.encode(": heartbeat\n\n")).catch(() => {});
	}, SSE_HEARTBEAT_MS);
	(async () => {
		try {
			await invokeDirectAttached({
				...opts,
				onEvent: (event) => writeSSE(event, event.type),
				emitIdleOnComplete: true
			});
		} catch (error) {
			await writeSSE({
				type: "error",
				instanceId: opts.id,
				error: toPublicError(error)
			}, "error");
		} finally {
			clearInterval(heartbeat);
			closed = true;
			try {
				await writer.close();
			} catch {}
		}
	})();
	return new Response(readable, { headers: {
		"content-type": "text/event-stream",
		"cache-control": "no-cache",
		connection: "keep-alive"
	} });
}
async function runDirectSyncMode(opts) {
	const result = await invokeDirectAttached(opts);
	return new Response(JSON.stringify({ result: result === void 0 ? null : result }), { headers: { "content-type": "application/json" } });
}
async function invokeDirectAttached(opts) {
	const sessionLock = acquireDirectAgentSessionLock(opts.agentName, opts.id, opts.payload);
	try {
		const ctx = opts.createContext(opts.id, void 0, opts.payload, opts.request);
		const runHandler = opts.runHandler ?? defaultRunHandler;
		let didEmitIdle = false;
		if (opts.onEvent || opts.emitIdleOnComplete) ctx.setEventCallback((event) => {
			if (event.type === "run_start" || event.type === "run_end") return;
			if (event.type === "idle") didEmitIdle = true;
			const attachedEvent = {
				...event,
				instanceId: opts.id
			};
			delete attachedEvent.runId;
			return opts.onEvent?.(attachedEvent);
		});
		try {
			return await runHandler(ctx, async (innerCtx) => {
				try {
					return await opts.handler(innerCtx);
				} finally {
					if (opts.emitIdleOnComplete && !didEmitIdle) innerCtx.emitEvent({ type: "idle" });
				}
			});
		} finally {
			ctx.setEventCallback(void 0);
		}
	} finally {
		sessionLock?.();
	}
}
async function runSseMode(execution) {
	if (!execution.runSubscribers) throw new Error("[flue] Workflow SSE requires a run subscriber registry.");
	const response = streamActiveRunEvents(execution.runStore, execution.runSubscribers, execution.runId);
	response.headers.set("X-Flue-Run-Id", execution.runId);
	try {
		startWorkflowExecution(execution);
	} catch (error) {
		await execution.completion?.catch(() => void 0);
		await response.body?.cancel();
		throw error;
	}
	return response;
}
async function runSyncMode(execution) {
	let result;
	try {
		result = await startWorkflowExecution(execution);
	} catch (error) {
		await execution.completion?.catch(() => void 0);
		throw error;
	}
	return new Response(JSON.stringify({
		result: result === void 0 ? null : result,
		_meta: { runId: execution.runId }
	}), { headers: {
		"content-type": "application/json",
		"X-Flue-Run-Id": execution.runId
	} });
}
async function invokeWorkflowAttached(opts) {
	if (!opts.startWorkflowAdmission) return invokeWorkflowAttachedUnlocked(opts);
	const execution = await prepareWorkflowExecution({
		owner: opts.owner,
		id: opts.id,
		runId: opts.runId,
		handler: opts.handler,
		payload: opts.payload,
		request: opts.request,
		createContext: opts.createContext,
		startWorkflowAdmission: opts.startWorkflowAdmission,
		runStore: opts.runStore,
		runSubscribers: opts.runSubscribers,
		runRegistry: opts.runRegistry,
		onAdmitted: opts.onAdmitted,
		onEvent: opts.onEvent,
		emitIdleOnComplete: opts.emitIdleOnComplete
	});
	let result;
	try {
		result = await startWorkflowExecution(execution);
	} catch (error) {
		await execution.completion?.catch(() => void 0);
		throw error;
	}
	return {
		runId: opts.runId,
		result
	};
}
async function invokeWorkflowAttachedUnlocked(opts) {
	const lifecycle = await createWorkflowRunLifecycle({
		owner: opts.owner,
		id: opts.id,
		runId: opts.runId,
		payload: opts.payload,
		request: opts.request,
		createContext: opts.createContext,
		runStore: opts.runStore,
		runSubscribers: opts.runSubscribers,
		runRegistry: opts.runRegistry
	});
	const { ctx } = lifecycle;
	let didEmitIdle = false;
	if (opts.onEvent || opts.emitIdleOnComplete) ctx.setEventCallback((event) => {
		if (event.type === "idle") didEmitIdle = true;
		return opts.onEvent?.(event);
	});
	try {
		const result = await withWorkflowRunLifecycle(lifecycle, async () => {
			try {
				return await opts.handler(ctx);
			} finally {
				if (opts.emitIdleOnComplete && !didEmitIdle) ctx.emitEvent({ type: "idle" });
			}
		});
		return {
			runId: opts.runId,
			result
		};
	} finally {
		ctx.setEventCallback(void 0);
	}
}
function acquireDirectAgentSessionLock(agentName, instanceId, input) {
	const payload = input;
	const key = `${agentName}\0${instanceId}\0${typeof payload?.session === "string" && payload.session.trim() !== "" ? payload.session : "default"}`;
	if (activeAttachedAgentSessions.has(key)) throw new InvalidRequestError({ reason: "This agent session already has an active prompt." });
	const token = Symbol(key);
	activeAttachedAgentSessions.set(key, token);
	return () => {
		if (activeAttachedAgentSessions.get(key) === token) activeAttachedAgentSessions.delete(key);
	};
}
async function createWorkflowRunLifecycle(options) {
	const startedAtMs = Date.now();
	const startedAt = new Date(startedAtMs).toISOString();
	const ctx = options.createContext(options.id, options.runId, options.payload, options.request);
	const runStore = options.runStore;
	const owner = options.owner;
	let didCreateRun = false;
	try {
		didCreateRun = runStore ? await persistRunAdmission("createRun", options.requirePersistedAdmission === true, () => runStore.createRun({
			runId: options.runId,
			owner,
			startedAt,
			payload: options.payload
		})) : false;
	} catch (error) {
		console.error("[flue] Workflow admission error:", {
			workflowName: owner.workflowName,
			runId: options.runId,
			operation: "createRun",
			outcome: "admission_failed"
		}, error);
		throw error;
	}
	if (didCreateRun) await safeRegistry("recordRunStart", () => options.runRegistry?.recordRunStart({
		runId: options.runId,
		owner,
		startedAt
	}));
	return {
		...options,
		ctx,
		startedAt,
		startedAtMs
	};
}
/**
* Wrap all workflow invocation modes with the same run-start/run-end envelope.
*/
async function withWorkflowRunLifecycle(lifecycle, body) {
	const flushFanout = subscribeRunFanout(lifecycle);
	emitRunStart(lifecycle);
	let didFlushFanout = false;
	let result;
	try {
		result = await body();
		await flushFanout();
		didFlushFanout = true;
	} catch (error) {
		if (!didFlushFanout) try {
			await flushFanout();
		} catch {}
		await emitRunEnd(lifecycle, {
			isError: true,
			error
		});
		throw error;
	}
	await emitRunEnd(lifecycle, {
		result,
		isError: false
	});
	return result;
}
function emitRunStart(lifecycle) {
	lifecycle.ctx.emitEvent({
		type: "run_start",
		runId: lifecycle.runId,
		owner: lifecycle.owner,
		instanceId: lifecycle.owner.instanceId,
		workflowName: lifecycle.owner.workflowName,
		startedAt: lifecycle.startedAt,
		payload: lifecycle.payload
	});
}
function emitRunResume(lifecycle) {
	lifecycle.ctx.emitEvent({
		type: "run_resume",
		runId: lifecycle.runId,
		owner: lifecycle.owner,
		instanceId: lifecycle.owner.instanceId,
		workflowName: lifecycle.owner.workflowName,
		startedAt: lifecycle.startedAt
	});
}
/**
* Emit `run_end` and finalize the run.
*
* Terminal ordering matters for `/runs/:runId/stream`: append `run_end`
* before marking the run terminal, then publish and close subscribers.
*/
async function emitRunEnd(lifecycle, input) {
	const endedAtMs = Date.now();
	const endedAt = new Date(endedAtMs).toISOString();
	const durationMs = endedAtMs - lifecycle.startedAtMs;
	const result = input.isError ? void 0 : input.result;
	const error = input.isError ? serializeError(input.error) : void 0;
	const normalizedResult = result === void 0 ? null : result;
	const { runStore, runSubscribers, runRegistry, runId } = lifecycle;
	const decorated = lifecycle.ctx.emitEvent({
		type: "run_end",
		runId,
		result: normalizedResult,
		isError: input.isError,
		error,
		durationMs
	});
	let appendError;
	try {
		await persistRunEvent("appendEvent(run_end)", () => runStore?.appendEvent(runId, decorated));
	} catch (error) {
		appendError = error;
	}
	runSubscribers?.publish(runId, decorated);
	if (runStore ? await safeRunStore("endRun", () => runStore.endRun({
		runId,
		endedAt,
		isError: input.isError,
		durationMs,
		result: input.isError ? result : normalizedResult,
		error
	})) : false) await safeRegistry("recordRunEnd", () => runRegistry?.recordRunEnd({
		runId,
		endedAt,
		durationMs,
		isError: input.isError
	}));
	runSubscribers?.complete(runId);
	if (appendError) throw appendError;
}
/**
* Persist non-terminal events before publishing them to live subscribers.
* `run_end` is handled separately by {@link emitRunEnd}.
*/
function subscribeRunFanout(lifecycle) {
	const { ctx, runStore, runSubscribers, runId } = lifecycle;
	if (!runStore && !runSubscribers) return async () => {};
	let chain = Promise.resolve();
	const unsubscribe = ctx.subscribeEvent((event) => {
		if (event.type === "run_end") return;
		chain = chain.then(() => fanOutEvent(runStore, runSubscribers, runId, event));
	});
	return () => {
		unsubscribe();
		return chain;
	};
}
async function fanOutEvent(runStore, runSubscribers, runId, event) {
	if (runStore) await persistRunEvent("appendEvent", () => runStore.appendEvent(runId, event));
	runSubscribers?.publish(runId, event);
}
async function persistRunEvent(label, fn) {
	try {
		await fn();
		return true;
	} catch (error) {
		if (error instanceof RunEventTooLargeError) throw error;
		console.error(`[flue:run-store] ${label} failed:`, error);
		return false;
	}
}
async function persistRunAdmission(label, required, fn) {
	try {
		await fn();
		return true;
	} catch (error) {
		console.error(`[flue:run-store] ${label} failed:`, error);
		if (required) throw error;
		return false;
	}
}
async function safeRunStore(label, fn) {
	return persistRunAdmission(label, false, fn);
}
async function safeRegistry(label, fn) {
	try {
		await fn();
	} catch (error) {
		console.error(`[flue:run-registry] ${label} failed:`, error);
	}
}
function serializeError(error) {
	if (error instanceof Error) return {
		name: error.name,
		message: error.message
	};
	return error;
}
function getEventIndex(data) {
	if (typeof data !== "object" || data === null) return void 0;
	const value = data.eventIndex;
	return typeof value === "number" ? value : void 0;
}
/**
* Default background workflow runner: invoke `run()` directly so the workflow
* executes in the current process. Used by the Node target. The Cloudflare
* target overrides this with a `runFiber` wrapper for crash-recoverable
* execution across DO hibernation.
*/
const defaultStartWorkflowAdmission = (_runId, run) => Promise.resolve().then(run);
/**
* Default direct-agent foreground handler runner: invoke directly. Used by the
* Node target. The Cloudflare target overrides this with a `runFiber` wrapper.
*/
const defaultRunHandler = (ctx, handler) => handler(ctx);
//#endregion
export { toPublicError as C, toHttpResponse as S, validateWorkflowRequest as T, RunEventTooLargeError as _, handleAgentRequest as a, ValidationError as b, invokeWorkflowAttached as c, handleRunRouteRequest as d, InMemoryDispatchQueue as f, RouteNotFoundError as g, MethodNotAllowedError as h, failRecoveredRun as i, reserveDispatchAgentSession as l, InvalidRequestError as m, createDirectAgentHandler as n, handleWorkflowRequest as o, assertCurrentDispatchInput as p, createDispatchAgentHandler as r, invokeDirectAttached as s, createAgentDispatchProcessor as t, validateAgentDispatchAdmission as u, RunNotFoundError as v, validateAgentRequest as w, configureErrorRendering as x, RunRegistryUnavailableError as y };
