// @flue/runtime@0.9.2/dist/index.mjs
import { a as defineAgentProfile, i as createAgent, n as observe } from "./events-BwRudw7t.mjs";
import { t as connectMcpServer } from "./mcp-CNpDbStB.mjs";
import { h as registerProvider, i as createSandboxSessionEnv, m as registerApiProvider, u as configureProvider, y as ResultUnavailableError } from "./sandbox-C2jFycj2.mjs";
import { r as dispatch } from "./flue-app-NVPwXKDc.mjs";
import { Type } from "@earendil-works/pi-ai";
//#region src/tool.ts
/**
* Validates a custom model-callable tool and returns a shallow-frozen copy.
*
* This validates the required definition fields. Tool names are checked for
* collisions with other active tools when a session assembles its tool list.
*/
function defineTool(tool) {
	if (!tool || typeof tool !== "object") throw new Error("[flue] defineTool() requires a tool definition object.");
	if (typeof tool.name !== "string" || tool.name.trim().length === 0) throw new Error("[flue] defineTool({ name }) must be a non-empty string.");
	if (typeof tool.description !== "string" || tool.description.trim().length === 0) throw new Error("[flue] defineTool({ description }) must be a non-empty string.");
	if (!tool.parameters || typeof tool.parameters !== "object") throw new Error("[flue] defineTool({ parameters }) is required.");
	if (typeof tool.execute !== "function") throw new Error("[flue] defineTool({ execute }) must be a function.");
	return Object.freeze({ ...tool });
}
//#endregion
export { ResultUnavailableError, Type, configureProvider, connectMcpServer, createAgent, createSandboxSessionEnv, defineAgentProfile, defineTool, dispatch, observe, registerApiProvider, registerProvider };
