// @flue/runtime@0.9.2/dist/internal.mjs
import { o as extendAgentProfile, r as assertResolvedAgentProfile, s as resolveAgentProfile, t as dispatchGlobalEvent } from "./events-BwRudw7t.mjs";
import { _ as resolveRegisteredModel, a as isBashLike, b as parseSkillMarkdown, c as deleteSessionTree, f as getProviderConfiguration, g as resetProvidersForTests, l as discoverSessionContext, n as createCwdSessionEnv, o as InMemorySessionStore, p as hasRegisteredProvider, r as createFlueFs, s as Session, t as bashFactoryToSessionEnv, x as formatBashResult } from "./sandbox-C2jFycj2.mjs";
import { _ as RunEventTooLargeError, a as handleAgentRequest, c as invokeWorkflowAttached, d as handleRunRouteRequest, f as InMemoryDispatchQueue, i as failRecoveredRun, l as reserveDispatchAgentSession, n as createDirectAgentHandler, o as handleWorkflowRequest, p as assertCurrentDispatchInput, r as createDispatchAgentHandler, s as invokeDirectAttached, t as createAgentDispatchProcessor, u as validateAgentDispatchAdmission } from "./handle-agent-C2C75lpq.mjs";
import { c as registeredWorkflowsForTransport, l as resetFlueRuntimeForTests, n as createDefaultFlueApp, r as dispatch, s as registeredAgentsForTransport, t as configureFlueRuntime } from "./flue-app-NVPwXKDc.mjs";
import { i as parseWorkflowRunId, r as generateWorkflowRunId } from "./ids-DiSJFsle.mjs";
import { n as createCallHandle } from "./abort-B-q3xjj8.mjs";
import { n as decodeRunCursor, r as encodeRunCursor, t as MAX_LIST_LIMIT } from "./run-registry-CwGOiYQh.mjs";
import { n as parseAgentWebSocketMessage, r as parseWorkflowWebSocketMessage, t as createWebSocketErrorMessage } from "./websocket-protocol-BSIpcV6C.mjs";
import { getModel } from "@earendil-works/pi-ai";
import { Bash, InMemoryFs } from "just-bash";
//#region src/harness.ts
const DEFAULT_SESSION_NAME = "default";
var Harness = class {
	instanceId;
	name;
	config;
	env;
	store;
	eventCallback;
	agentTools;
	toolFactory;
	sessions = {
		get: (name) => this.openSession(name, "get"),
		create: (name) => this.openSession(name, "create"),
		delete: (name) => this.deleteSession(name)
	};
	fs;
	openSessions = /* @__PURE__ */ new Map();
	pendingSessionOperations = /* @__PURE__ */ new Map();
	constructor(instanceId, name, config, env, store, eventCallback, agentTools = [], toolFactory) {
		this.instanceId = instanceId;
		this.name = name;
		this.config = config;
		this.env = env;
		this.store = store;
		this.eventCallback = eventCallback;
		this.agentTools = agentTools;
		this.toolFactory = toolFactory;
		this.fs = createFlueFs(env);
	}
	async session(name) {
		return this.openSession(name, "get-or-create");
	}
	shell(command, options) {
		return createCallHandle(options?.signal, async (signal) => {
			const toolCallId = crypto.randomUUID();
			const startedAt = Date.now();
			const args = { command };
			if (options?.cwd !== void 0) args.cwd = options.cwd;
			if (options?.env !== void 0) args.env = redactEnvValues(options.env);
			this.emit({
				type: "tool_start",
				toolName: "bash",
				toolCallId,
				args
			});
			try {
				const result = await this.env.exec(command, {
					env: options?.env,
					cwd: options?.cwd,
					signal
				});
				const shellResult = {
					stdout: result.stdout,
					stderr: result.stderr,
					exitCode: result.exitCode
				};
				this.emit({
					type: "tool_call",
					toolName: "bash",
					toolCallId,
					isError: false,
					result: formatBashResult(shellResult, command),
					durationMs: Date.now() - startedAt
				});
				return shellResult;
			} catch (error) {
				const result = {
					content: [{
						type: "text",
						text: getErrorMessage(error)
					}],
					details: {
						command,
						exitCode: -1
					}
				};
				this.emit({
					type: "tool_call",
					toolName: "bash",
					toolCallId,
					isError: true,
					result,
					durationMs: Date.now() - startedAt
				});
				throw error;
			}
		});
	}
	openSession(name, mode) {
		const sessionName = normalizeSessionName(name);
		return this.runSessionOperation(sessionName, () => this.loadSession(sessionName, mode));
	}
	runSessionOperation(sessionName, operation) {
		const result = (this.pendingSessionOperations.get(sessionName) ?? Promise.resolve()).then(operation);
		const tail = result.then(() => {}, () => {});
		this.pendingSessionOperations.set(sessionName, tail);
		tail.then(() => {
			if (this.pendingSessionOperations.get(sessionName) === tail) this.pendingSessionOperations.delete(sessionName);
		});
		return result;
	}
	async loadSession(sessionName, mode) {
		const open = this.openSessions.get(sessionName);
		if (open) {
			if (mode === "create") throw new Error(`[flue] Session "${sessionName}" already exists in harness "${this.name}".`);
			return open;
		}
		const storageKey = createSessionStorageKey(this.instanceId, this.name, sessionName);
		const affinityKey = createSessionAffinityKey(this.instanceId, this.name, sessionName);
		const existingData = await this.store.load(storageKey);
		if (mode === "get" && !existingData) throw new Error(`[flue] Session "${sessionName}" does not exist in harness "${this.name}".`);
		if (mode === "create" && existingData) throw new Error(`[flue] Session "${sessionName}" already exists in harness "${this.name}".`);
		let data = existingData;
		if (!data && mode !== "get") {
			data = createEmptySessionData();
			await this.store.save(storageKey, data);
		}
		const session = new Session({
			name: sessionName,
			storageKey,
			affinityKey,
			config: this.config,
			env: this.env,
			store: this.store,
			existingData: data,
			onAgentEvent: this.decorateEventCallback(this.eventCallback),
			agentTools: this.agentTools,
			toolFactory: this.toolFactory,
			taskDepth: 0,
			createTaskSession: (taskOptions) => this.createTaskSession(taskOptions),
			onDelete: () => this.openSessions.delete(sessionName)
		});
		this.openSessions.set(sessionName, session);
		return session;
	}
	deleteSession(name) {
		const sessionName = normalizeSessionName(name);
		return this.runSessionOperation(sessionName, async () => {
			const open = this.openSessions.get(sessionName);
			if (open) {
				await open.delete();
				return;
			}
			await deleteSessionTree(this.store, createSessionStorageKey(this.instanceId, this.name, sessionName));
		});
	}
	async createTaskSession(options) {
		const sessionName = `task:${options.parentSession}:${options.taskId}`;
		const taskEnv = options.cwd ? createCwdSessionEnv(options.parentEnv, options.parentEnv.resolvePath(options.cwd)) : options.parentEnv;
		const taskAgent = options.agent;
		const localContext = await discoverSessionContext(taskEnv, taskAgent?.instructions ?? this.config.instructions, taskAgent?.skills ?? this.config.definitionSkills);
		const taskConfig = {
			...this.config,
			systemPrompt: localContext.systemPrompt,
			instructions: taskAgent?.instructions ?? this.config.instructions,
			definitionSkills: taskAgent?.skills ?? this.config.definitionSkills,
			skills: localContext.skills,
			subagents: taskAgent ? Object.fromEntries((taskAgent.subagents ?? []).filter((agent) => agent.name !== void 0).map((agent) => [agent.name, agent])) : this.config.subagents,
			model: taskAgent?.model !== void 0 ? this.config.resolveModel(taskAgent.model) : this.config.model,
			thinkingLevel: taskAgent?.thinkingLevel ?? this.config.thinkingLevel,
			compaction: taskAgent?.compaction ?? this.config.compaction
		};
		const storageKey = createSessionStorageKey(this.instanceId, this.name, sessionName);
		const affinityKey = createSessionAffinityKey(this.instanceId, this.name, sessionName);
		const data = createEmptySessionData();
		data.metadata = {
			parentSession: options.parentSession,
			taskId: options.taskId,
			cwd: taskEnv.cwd,
			agent: taskAgent?.name,
			depth: options.depth
		};
		await this.store.save(storageKey, data);
		const eventCallback = this.eventCallback ? (event) => {
			this.eventCallback?.({
				...event,
				harness: event.harness ?? this.name,
				parentSession: event.parentSession ?? options.parentSession,
				taskId: event.taskId ?? options.taskId
			});
		} : void 0;
		return new Session({
			name: sessionName,
			storageKey,
			affinityKey,
			config: taskConfig,
			env: taskEnv,
			store: this.store,
			existingData: data,
			onAgentEvent: eventCallback,
			agentTools: taskAgent?.tools ?? this.agentTools,
			toolFactory: this.toolFactory,
			taskDepth: options.depth,
			createTaskSession: (childOptions) => this.createTaskSession(childOptions)
		});
	}
	emit(event) {
		this.eventCallback?.({
			...event,
			harness: event.harness ?? this.name
		});
	}
	decorateEventCallback(callback) {
		return callback ? (event) => {
			callback({
				...event,
				harness: event.harness ?? this.name
			});
		} : void 0;
	}
};
function getErrorMessage(error) {
	return error instanceof Error ? error.message : String(error);
}
function redactEnvValues(env) {
	return Object.fromEntries(Object.keys(env).map((key) => [key, "<redacted>"]));
}
function normalizeSessionName(name) {
	return name ?? DEFAULT_SESSION_NAME;
}
function createSessionStorageKey(instanceId, harness, sessionName) {
	return `agent-session:${JSON.stringify([
		instanceId,
		harness,
		sessionName
	])}`;
}
function createSessionAffinityKey(instanceId, harness, sessionName) {
	return `${instanceId}::${harness}::${sessionName}`;
}
function createEmptySessionData() {
	const now = (/* @__PURE__ */ new Date()).toISOString();
	return {
		version: 4,
		entries: [],
		leafId: null,
		metadata: {},
		createdAt: now,
		updatedAt: now
	};
}
//#endregion
//#region src/client.ts
function createFlueContext(config) {
	const subscribers = /* @__PURE__ */ new Set();
	let handlerUnsubscribe;
	let eventIndex = config.initialEventIndex ?? 0;
	const initializedHarnessNames = /* @__PURE__ */ new Set();
	const emitEvent = (event) => {
		const decorated = {
			...event,
			...config.runId === void 0 ? { instanceId: config.id } : { runId: config.runId },
			...config.dispatchId === void 0 ? {} : { dispatchId: config.dispatchId },
			eventIndex: eventIndex++,
			timestamp: (/* @__PURE__ */ new Date()).toISOString()
		};
		for (const subscriber of subscribers) try {
			Promise.resolve(subscriber(decorated)).catch((error) => {
				console.error("[flue:subscriber] Event subscriber failed:", error);
			});
		} catch (error) {
			console.error("[flue:subscriber] Event subscriber failed:", error);
		}
		dispatchGlobalEvent(decorated, ctx);
		return decorated;
	};
	const ctx = {
		get id() {
			return config.id;
		},
		get runId() {
			return config.runId;
		},
		get payload() {
			return config.payload;
		},
		get env() {
			return config.env;
		},
		get req() {
			return config.req;
		},
		log: {
			info(message, attributes) {
				emitEvent({
					type: "log",
					level: "info",
					message,
					attributes: normalizeLogAttributes(attributes)
				});
			},
			warn(message, attributes) {
				emitEvent({
					type: "log",
					level: "warn",
					message,
					attributes: normalizeLogAttributes(attributes)
				});
			},
			error(message, attributes) {
				emitEvent({
					type: "log",
					level: "error",
					message,
					attributes: normalizeLogAttributes(attributes)
				});
			}
		},
		init(agent, options) {
			return ctx.initializeCreatedAgent(agent, config.payload, options);
		},
		async initializeCreatedAgent(agent, payload, options) {
			if (!agent || agent.__flueCreatedAgent !== true || typeof agent.initialize !== "function") throw new Error("[flue] init() requires an agent created with createAgent(...).");
			const resolvedOptions = await agent.initialize({
				id: config.id,
				env: config.env,
				payload
			});
			const definition = assertResolvedAgentProfile(extendAgentProfile(resolveAgentProfile(resolvedOptions), {
				tools: options?.tools,
				skills: options?.skills,
				subagents: options?.subagents
			}), "createAgent() and init()");
			if (!hasInitModel(resolvedOptions)) throw new Error("[flue] createAgent() requires a model. Return { model: \"provider-id/model-id\" }, { model: false }, or a profile with a model.");
			if (definition.model !== false && typeof definition.model !== "string") throw new Error("[flue] createAgent() model must be a model specifier or false.");
			const name = options?.name ?? "default";
			if (initializedHarnessNames.has(name)) throw new Error(`[flue] init() has already been called with name "${name}" in this request.`);
			initializedHarnessNames.add(name);
			try {
				const sandbox = resolvedOptions.sandbox;
				const { env: baseEnv, toolFactory } = await resolveSessionEnv(config.id, sandbox, config);
				const env = resolvedOptions.cwd ? createCwdSessionEnv(baseEnv, baseEnv.resolvePath(resolvedOptions.cwd)) : baseEnv;
				const store = resolvedOptions.persist ?? config.defaultStore;
				const localContext = await discoverSessionContext(env, definition.instructions, definition.skills);
				const agentModel = config.agentConfig.resolveModel(definition.model);
				const agentConfig = {
					...config.agentConfig,
					systemPrompt: localContext.systemPrompt,
					instructions: definition.instructions,
					definitionSkills: definition.skills,
					skills: localContext.skills,
					subagents: Object.fromEntries((definition.subagents ?? []).filter((agent) => agent.name !== void 0).map((agent) => [agent.name, agent])),
					model: agentModel,
					thinkingLevel: definition.thinkingLevel ?? config.agentConfig.thinkingLevel,
					compaction: definition.compaction ?? config.agentConfig.compaction
				};
				return new Harness(config.id, name, agentConfig, env, store, (event) => {
					emitEvent(event);
				}, definition.tools, toolFactory);
			} catch (error) {
				initializedHarnessNames.delete(name);
				throw error;
			}
		},
		emitEvent,
		subscribeEvent(callback) {
			subscribers.add(callback);
			return () => subscribers.delete(callback);
		},
		setEventCallback(callback) {
			handlerUnsubscribe?.();
			handlerUnsubscribe = callback ? ctx.subscribeEvent(callback) : void 0;
		}
	};
	return ctx;
}
function normalizeLogAttributes(attributes) {
	if (!attributes) return void 0;
	if (!(attributes.error instanceof Error)) return attributes;
	return {
		...attributes,
		error: serializeLogError(attributes.error)
	};
}
function serializeLogError(error) {
	return {
		name: error.name,
		message: error.message,
		stack: error.stack
	};
}
function hasInitModel(options) {
	return Boolean(options && ("model" in options || options.profile && "model" in options.profile));
}
function isBashFactory(value) {
	return typeof value === "function";
}
function isSandboxFactory(value) {
	return typeof value === "object" && value !== null && "createSessionEnv" in value && typeof value.createSessionEnv === "function";
}
/** Resolve sandbox option to its session environment and optional tool factory. */
async function resolveSessionEnv(id, sandbox, config) {
	if (sandbox === void 0 || sandbox === false) return { env: await config.createDefaultEnv() };
	if (sandbox === "empty") throw new Error("[flue] `sandbox: 'empty'` is no longer supported. Omit the `sandbox` option (or pass `false`) for the default in-memory sandbox.");
	if (sandbox === "local") throw new Error("[flue] `sandbox: 'local'` is no longer supported. Use the `local()` factory instead: `import { local } from '@flue/runtime/node'; createAgent(() => ({ sandbox: local(), model: false }))`. The factory accepts an `env` option for opting host env vars into the sandbox.");
	if (isBashFactory(sandbox)) return { env: await bashFactoryToSessionEnv(sandbox) };
	if (isBashLike(sandbox)) throw new Error("[flue] createAgent() sandbox no longer accepts a Bash-like object directly. Pass a BashFactory instead, e.g. `sandbox: () => new Bash({ fs })`.");
	if (config.resolveSandbox) {
		const resolved = await config.resolveSandbox(sandbox);
		if (resolved) return { env: resolved };
	}
	if (isSandboxFactory(sandbox)) return {
		env: await sandbox.createSessionEnv({ id }),
		toolFactory: sandbox.tools
	};
	throw new Error("[flue] Invalid sandbox option returned from createAgent().");
}
//#endregion
//#region src/runtime/run-store.ts
const MAX_EVENT_BYTES = 1024 * 1024;
const ENCODER = new TextEncoder();
function serializedEventForPersistence(event) {
	const payload = JSON.stringify(event);
	if (byteLength(payload) > MAX_EVENT_BYTES) throw new RunEventTooLargeError();
	return payload;
}
function byteLength(value) {
	return ENCODER.encode(value).byteLength;
}
//#endregion
//#region src/cloudflare/run-store.ts
function createDurableRunStore(sql) {
	ensureRunTables(sql);
	return new DurableRunStore(sql);
}
var DurableRunStore = class {
	sql;
	constructor(sql) {
		this.sql = sql;
	}
	async createRun(input) {
		if (input.owner.instanceId !== input.runId) throw new Error("[flue] Workflow run owners must use the same instanceId as the run record runId.");
		this.sql.exec(`INSERT OR REPLACE INTO flue_runs
			 (run_id, owner_kind, instance_id, workflow_name, status, started_at, payload, ended_at, is_error, duration_ms, result, error)
			 VALUES (?, 'workflow', ?, ?, ?, ?, ?, NULL, NULL, NULL, NULL, NULL)`, input.runId, input.owner.instanceId, input.owner.workflowName, "active", input.startedAt, JSON.stringify(input.payload ?? null));
	}
	async endRun(input) {
		this.sql.exec(`UPDATE flue_runs
			 SET status = ?, ended_at = ?, is_error = ?, duration_ms = ?, result = ?, error = ?
			 WHERE run_id = ?`, input.isError ? "errored" : "completed", input.endedAt, input.isError ? 1 : 0, input.durationMs, JSON.stringify(input.result ?? null), JSON.stringify(input.error ?? null), input.runId);
	}
	async appendEvent(runId, event) {
		const payload = serializedEventForPersistence(event);
		this.sql.exec(`INSERT OR REPLACE INTO flue_run_events
			 (run_id, event_index, type, payload, timestamp)
			 VALUES (?, ?, ?, ?, ?)`, runId, event.eventIndex ?? 0, event.type, payload, event.timestamp ?? (/* @__PURE__ */ new Date()).toISOString());
	}
	async getEvents(runId, fromIndex) {
		return this.sql.exec(fromIndex === void 0 ? "SELECT payload FROM flue_run_events WHERE run_id = ? ORDER BY event_index ASC" : "SELECT payload FROM flue_run_events WHERE run_id = ? AND event_index >= ? ORDER BY event_index ASC", ...fromIndex === void 0 ? [runId] : [runId, fromIndex]).toArray().flatMap((row) => typeof row.payload === "string" ? [JSON.parse(row.payload)] : []);
	}
	async getRun(runId) {
		const row = this.sql.exec("SELECT * FROM flue_runs WHERE run_id = ? AND owner_kind = 'workflow'", runId).toArray()[0];
		if (!row) return null;
		return rowToRunRecord(row);
	}
};
function ensureRunTables(sql) {
	sql.exec(`CREATE TABLE IF NOT EXISTS flue_runs (
		 run_id TEXT PRIMARY KEY,
		 owner_kind TEXT NOT NULL,
		 instance_id TEXT,
		 workflow_name TEXT,
			 status TEXT NOT NULL,
			 started_at TEXT NOT NULL,
			 payload TEXT,
			 ended_at TEXT,
			 is_error INTEGER,
			 duration_ms INTEGER,
			 result TEXT,
			 error TEXT
		)`);
	sql.exec(`CREATE TABLE IF NOT EXISTS flue_run_events (
		 run_id TEXT NOT NULL,
		 event_index INTEGER NOT NULL,
		 type TEXT NOT NULL,
		 payload TEXT NOT NULL,
		 timestamp TEXT NOT NULL,
		 PRIMARY KEY (run_id, event_index)
		)`);
	sql.exec("CREATE INDEX IF NOT EXISTS flue_runs_instance_started_idx ON flue_runs (owner_kind, instance_id, started_at DESC)");
	sql.exec("CREATE INDEX IF NOT EXISTS flue_runs_workflow_started_idx ON flue_runs (owner_kind, workflow_name, started_at DESC)");
	sql.exec("CREATE INDEX IF NOT EXISTS flue_run_events_run_idx ON flue_run_events (run_id, event_index ASC)");
}
function rowToRunRecord(row) {
	const payload = typeof row.payload === "string" ? JSON.parse(row.payload) : void 0;
	const result = typeof row.result === "string" ? JSON.parse(row.result) : void 0;
	const error = typeof row.error === "string" ? JSON.parse(row.error) : void 0;
	const runId = String(row.run_id);
	return {
		runId,
		owner: {
			kind: "workflow",
			workflowName: String(row.workflow_name),
			instanceId: String(row.instance_id ?? runId)
		},
		status: row.status,
		startedAt: String(row.started_at),
		payload,
		endedAt: typeof row.ended_at === "string" ? row.ended_at : void 0,
		isError: row.is_error === null || row.is_error === void 0 ? void 0 : Boolean(row.is_error),
		durationMs: typeof row.duration_ms === "number" ? row.duration_ms : void 0,
		result,
		error
	};
}
//#endregion
//#region src/node/run-registry.ts
/** In-memory `RunRegistry` for workflow runs on the Node target. */
var InMemoryRunRegistry = class {
	pointers = /* @__PURE__ */ new Map();
	async recordRunStart(input) {
		if (this.pointers.has(input.runId)) return;
		if (input.owner.instanceId !== input.runId) throw new Error("[flue] Workflow run owners must use the same instanceId as the pointer runId.");
		this.pointers.set(input.runId, {
			runId: input.runId,
			owner: input.owner,
			status: "active",
			startedAt: input.startedAt
		});
	}
	async recordRunEnd(input) {
		const pointer = this.pointers.get(input.runId);
		if (!pointer) return;
		this.pointers.set(input.runId, {
			...pointer,
			status: input.isError ? "errored" : "completed",
			endedAt: input.endedAt,
			durationMs: input.durationMs,
			isError: input.isError
		});
	}
	async lookupRun(runId) {
		return this.pointers.get(runId) ?? null;
	}
	async listRuns(opts = {}) {
		const limit = clampLimit(opts.limit);
		const cursor = decodeRunCursor(opts.cursor);
		const all = [...this.pointers.values()].filter((pointer) => matchesListFilter(pointer, opts)).sort(comparePointersDesc);
		const startIndex = cursor ? all.findIndex((pointer) => isAfterCursor(pointer, cursor)) : 0;
		if (startIndex === -1) return { runs: [] };
		const page = all.slice(startIndex, startIndex + limit);
		const last = page.at(-1);
		return {
			runs: page,
			nextCursor: startIndex + limit < all.length && last ? encodeRunCursor(last) : void 0
		};
	}
};
function matchesListFilter(pointer, opts) {
	if (opts.status && pointer.status !== opts.status) return false;
	if (opts.workflowName && pointer.owner.workflowName !== opts.workflowName) return false;
	return true;
}
function comparePointersDesc(a, b) {
	const byStarted = b.startedAt.localeCompare(a.startedAt);
	if (byStarted !== 0) return byStarted;
	return b.runId.localeCompare(a.runId);
}
function isAfterCursor(pointer, cursor) {
	if (pointer.startedAt < cursor.startedAt) return true;
	if (pointer.startedAt > cursor.startedAt) return false;
	return pointer.runId < cursor.runId;
}
function clampLimit(limit) {
	if (!limit || !Number.isFinite(limit) || limit <= 0) return 100;
	return Math.min(limit, MAX_LIST_LIMIT);
}
//#endregion
//#region src/node/run-store.ts
var InMemoryRunStore = class {
	instances = /* @__PURE__ */ new Map();
	async createRun(input) {
		if (input.owner.instanceId !== input.runId) throw new Error("[flue] Workflow run owners must use the same instanceId as the run record runId.");
		const instance = this.getInstance(ownerKey(input.owner));
		instance.runs.set(input.runId, {
			runId: input.runId,
			owner: input.owner,
			status: "active",
			startedAt: input.startedAt,
			payload: input.payload
		});
		instance.events.set(input.runId, []);
	}
	async endRun(input) {
		const existing = await this.getRun(input.runId);
		if (!existing) return;
		this.getInstance(ownerKey(existing.owner)).runs.set(input.runId, {
			...existing,
			status: input.isError ? "errored" : "completed",
			endedAt: input.endedAt,
			isError: input.isError,
			durationMs: input.durationMs,
			result: input.result,
			error: input.error
		});
	}
	async appendEvent(runId, event) {
		const run = await this.getRun(runId);
		if (!run) return;
		const instance = this.getInstance(ownerKey(run.owner));
		const events = instance.events.get(runId) ?? [];
		events.push({
			eventIndex: event.eventIndex,
			payload: serializedEventForPersistence(event)
		});
		instance.events.set(runId, events);
	}
	async getEvents(runId, fromIndex) {
		const run = await this.getRun(runId);
		if (!run) return [];
		return (this.getInstance(ownerKey(run.owner)).events.get(runId) ?? []).filter((event) => fromIndex === void 0 || typeof event.eventIndex === "number" && event.eventIndex >= fromIndex).map((event) => JSON.parse(event.payload));
	}
	async getRun(runId) {
		for (const instance of this.instances.values()) {
			const run = instance.runs.get(runId);
			if (run) return run;
		}
		return null;
	}
	getInstance(key) {
		let instance = this.instances.get(key);
		if (!instance) {
			instance = {
				runs: /* @__PURE__ */ new Map(),
				events: /* @__PURE__ */ new Map()
			};
			this.instances.set(key, instance);
		}
		return instance;
	}
};
function ownerKey(owner) {
	return `workflow\0${owner.workflowName}\0${owner.instanceId}`;
}
//#endregion
//#region src/runtime/run-subscribers.ts
function createRunSubscriberRegistry() {
	const listeners = /* @__PURE__ */ new Map();
	return {
		subscribe(runId, listener) {
			let bucket = listeners.get(runId);
			if (!bucket) {
				bucket = /* @__PURE__ */ new Set();
				listeners.set(runId, bucket);
			}
			bucket.add(listener);
			return () => {
				const current = listeners.get(runId);
				if (!current) return;
				current.delete(listener);
				if (current.size === 0) listeners.delete(runId);
			};
		},
		publish(runId, event) {
			const bucket = listeners.get(runId);
			if (!bucket || bucket.size === 0) return;
			for (const listener of [...bucket]) try {
				listener(event);
			} catch (error) {
				console.error("[flue:run-subscribers] listener threw:", error);
			}
		},
		complete(runId) {
			listeners.delete(runId);
		}
	};
}
//#endregion
//#region src/internal.ts
/**
* Internal runtime helpers consumed by the generated server entry point.
*
* This subpath is NOT part of the public API. It exists solely so the build
* plugins (Node, Cloudflare) can emit stable bare-specifier imports that
* resolve through normal package-exports resolution at both build time and
* runtime, for both workspace-linked and published-npm installs.
*
* User agent code should never import from here.
*/
/**
* Resolve a `provider-id/model-id` model specifier to a pi-ai Model.
* Registered provider IDs win over pi-ai's catalog; configured provider
* settings patch the resolved Model.
*/
function resolveModel(model) {
	if (model === false || model === void 0) return void 0;
	const modelSpecifier = model;
	const slash = modelSpecifier.indexOf("/");
	if (slash === -1) throw new Error(`[flue] Invalid model specifier "${modelSpecifier}". Use the "provider-id/model-id" format (e.g. "anthropic/claude-haiku-4-5").`);
	const providerId = modelSpecifier.slice(0, slash);
	const modelId = modelSpecifier.slice(slash + 1);
	const registered = resolveRegisteredModel(providerId, modelId);
	if (registered) {
		if (modelId === "") throw new Error(`[flue] Invalid model specifier "${modelSpecifier}". Provider ID "${providerId}" is registered via registerProvider(), but no model ID was given. Use "${providerId}/<model-id>".`);
		return applyProviderSettings(registered, getProviderConfiguration(providerId));
	}
	const resolved = getModel(providerId, modelId);
	if (!resolved) throw new Error(`[flue] Unknown model specifier "${modelSpecifier}". Provider ID "${providerId}" / model ID "${modelId}" is not registered with @earendil-works/pi-ai or via registerProvider().`);
	return applyProviderSettings(resolved, getProviderConfiguration(providerId));
}
function applyProviderSettings(model, providerSettings) {
	if (!providerSettings) return model;
	const hasBaseUrl = providerSettings.baseUrl !== void 0;
	const hasHeaders = providerSettings.headers !== void 0;
	if (!hasBaseUrl && !hasHeaders) return model;
	return {
		...model,
		baseUrl: providerSettings.baseUrl ?? model.baseUrl,
		headers: hasHeaders ? {
			...model.headers ?? {},
			...providerSettings.headers
		} : model.headers
	};
}
//#endregion
export { Bash, InMemoryDispatchQueue, InMemoryFs, InMemoryRunRegistry, InMemoryRunStore, InMemorySessionStore, assertCurrentDispatchInput, bashFactoryToSessionEnv, configureFlueRuntime, createAgentDispatchProcessor, createDefaultFlueApp, createDirectAgentHandler, createDispatchAgentHandler, createDurableRunStore, createFlueContext, createRunSubscriberRegistry, createWebSocketErrorMessage, dispatch, failRecoveredRun, generateWorkflowRunId, handleAgentRequest, handleRunRouteRequest, handleWorkflowRequest, hasRegisteredProvider, invokeDirectAttached, invokeWorkflowAttached, parseAgentWebSocketMessage, parseSkillMarkdown, parseWorkflowRunId, parseWorkflowWebSocketMessage, registeredAgentsForTransport, registeredWorkflowsForTransport, reserveDispatchAgentSession, resetFlueRuntimeForTests, resetProvidersForTests, resolveModel, validateAgentDispatchAdmission };
