// @flue/runtime@0.9.2/dist/routing.mjs
import { S as toHttpResponse, b as ValidationError, g as RouteNotFoundError, y as RunRegistryUnavailableError } from "./handle-agent-C2C75lpq.mjs";
import { a as getFlueRuntime, d as ErrorEnvelopeSchema, f as ListAgentsResponseSchema, h as RunRecordSchema, i as flue, m as RunIdParamSchema, o as handleRunById, p as ListRunsResponseSchema, u as AdminRunsQuerySchema } from "./flue-app-NVPwXKDc.mjs";
import { Hono } from "hono";
import { describeRoute, openAPIRouteHandler, resolver, validator } from "hono-openapi";
//#region src/runtime/admin-app.ts
/**
* Creates a mountable Hono sub-app for read-only deployment inspection.
* Mount it explicitly beneath an application-chosen prefix and protect that
* mount with application-owned authorization.
*
* The mounted sub-app exposes:
*
* - `GET /openapi.json`
* - `GET /agents`
* - `GET /runs`
* - `GET /runs/:runId`
*/
function admin() {
	const app = new Hono();
	app.get("/openapi.json", lazyOpenApiRouteHandler(app, adminOpenApiOptions));
	app.get("/agents", describeRoute(adminAgentsSpec()), listAgentsHandler);
	app.get("/runs", describeRoute(adminRunsSpec()), validated("query", AdminRunsQuerySchema), listRunsHandler);
	app.get("/runs/:runId", describeRoute(adminRunDetailSpec()), validated("param", RunIdParamSchema), runDetailHandler);
	app.onError((err) => toHttpResponse(err));
	return app;
}
function adminOpenApiOptions() {
	return { documentation: {
		info: {
			title: "Flue Admin API",
			version: getFlueRuntime()?.runtimeVersion ?? "0.0.0",
			description: "Read-only Flue deployment inspection API."
		},
		servers: []
	} };
}
function lazyOpenApiRouteHandler(app, getOptions) {
	return (c, next) => openAPIRouteHandler(app, getOptions())(c, next);
}
function validated(target, schema) {
	return validator(target, schema, (result) => {
		if (result.success) return;
		throw new ValidationError({
			details: `Invalid ${target} parameters.`,
			issues: result.error
		});
	});
}
function jsonResponse(schema, description) {
	return {
		description,
		content: { "application/json": { schema: resolver(schema) } }
	};
}
function errorResponses() {
	return {
		400: jsonResponse(ErrorEnvelopeSchema, "Validation or request-shape error."),
		404: jsonResponse(ErrorEnvelopeSchema, "Resource or route not found."),
		500: jsonResponse(ErrorEnvelopeSchema, "Internal server error."),
		501: jsonResponse(ErrorEnvelopeSchema, "Runtime feature is not configured.")
	};
}
function adminAgentsSpec() {
	return {
		tags: ["admin"],
		operationId: "adminListAgents",
		summary: "List built agents",
		responses: {
			200: jsonResponse(ListAgentsResponseSchema, "Unpaginated list of built agents."),
			...errorResponses()
		}
	};
}
function adminRunsSpec() {
	return {
		tags: ["admin"],
		operationId: "adminListRuns",
		summary: "List workflow runs across the deployment",
		responses: {
			200: jsonResponse(ListRunsResponseSchema, "Cursor-paginated list response."),
			...errorResponses()
		}
	};
}
function adminRunDetailSpec() {
	return {
		tags: ["admin"],
		operationId: "adminGetRun",
		summary: "Get a workflow run record",
		responses: {
			200: jsonResponse(RunRecordSchema, "Run record."),
			...errorResponses()
		}
	};
}
const listAgentsHandler = async (c) => {
	const rt = requireRuntime();
	return c.json({ items: rt.manifest?.agents ?? [] });
};
const listRunsHandler = async (c) => {
	const registry = requireRegistry(requireRuntime(), c.env);
	const url = new URL(c.req.url);
	const opts = {
		...parseListQuery(c.req.raw),
		status: statusFromRequest(c.req.raw)
	};
	const workflowName = url.searchParams.get("workflowName");
	if (workflowName) opts.workflowName = workflowName;
	const out = await registry.listRuns(opts);
	return c.json({
		items: out.runs,
		nextCursor: out.nextCursor
	});
};
const runDetailHandler = async (c) => {
	const rt = requireRuntime();
	const runId = c.req.param("runId") ?? "";
	if (!runId) throw new RouteNotFoundError({
		method: c.req.method,
		path: new URL(c.req.url).pathname
	});
	return handleRunById({
		rt,
		request: rewriteToPublicRunRequest(c.req.raw, runId),
		env: c.env,
		runId,
		action: "get"
	});
};
function rewriteToPublicRunRequest(request, runId) {
	const url = new URL(request.url);
	url.pathname = `/runs/${encodeURIComponent(runId)}`;
	return new Request(url, request);
}
function requireRuntime() {
	const rt = getFlueRuntime();
	if (!rt) throw new Error("[flue] admin() route invoked before runtime was configured. This usually means admin() was used outside a Flue-built server entry.");
	return rt;
}
function requireRegistry(rt, env) {
	if (rt.target === "cloudflare") {
		const registry = rt.createRunRegistryForRequest?.(env);
		if (!registry) throw new RunRegistryUnavailableError();
		return registry;
	}
	if (!rt.runRegistry) throw new RunRegistryUnavailableError();
	return rt.runRegistry;
}
function parseListQuery(request) {
	const params = new URL(request.url).searchParams;
	const out = {};
	const cursor = params.get("cursor");
	if (cursor) out.cursor = cursor;
	const limit = params.get("limit");
	if (limit) out.limit = Number.parseInt(limit, 10);
	return out;
}
function statusFromRequest(request) {
	const status = new URL(request.url).searchParams.get("status");
	return status === "active" || status === "completed" || status === "errored" ? status : void 0;
}
//#endregion
export { admin, flue };
