// @flue/runtime@0.9.2/dist/sandbox-C2jFycj2.mjs
import { n as generateTurnId, t as generateOperationId } from "./ids-DiSJFsle.mjs";
import { n as createCallHandle, t as abortErrorFor } from "./abort-B-q3xjj8.mjs";
import { Type, completeSimple, getModel, isContextOverflow, registerApiProvider, streamSimple } from "@earendil-works/pi-ai";
import * as v from "valibot";
import { toJsonSchema } from "@valibot/to-json-schema";
import { load } from "js-yaml";
import { Agent } from "@earendil-works/pi-agent-core";
//#region src/agent.ts
const MAX_READ_LINES = 2e3;
const MAX_READ_BYTES = 50 * 1024;
const MAX_GREP_MATCHES = 100;
const MAX_GREP_LINE_LENGTH = 500;
const MAX_GLOB_RESULTS = 1e3;
const BASE64_READ_LINE_LENGTH = 76;
const PACKAGED_SKILLS_ROOT = "/.flue/packaged-skills/";
function createTools(env, options) {
	const tools = [
		createReadTool(env, options?.packagedSkills ?? {}),
		createWriteTool(env),
		createEditTool(env),
		createBashTool(env),
		createGrepTool(env),
		createGlobTool(env)
	];
	if (options?.task) tools.push(createTaskTool(options.task, options.subagents ?? {}));
	return tools;
}
const ReadParams = Type.Object({
	path: Type.String({ description: "Path to the file to read" }),
	offset: Type.Optional(Type.Number({ description: "Line number to start from (1-indexed)" })),
	limit: Type.Optional(Type.Number({ description: "Maximum number of lines to read" }))
});
function createPackagedSkillReadTool(packagedSkills) {
	return {
		name: "read",
		label: "Read Packaged Skill File",
		description: "Read a packaged skill supporting file by its advertised path.",
		parameters: ReadParams,
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			const content = readPackagedSkillFile(packagedSkills, params.path);
			if (content === void 0) throw new Error(`[flue] Packaged skill file not found: ${params.path}`);
			return formatReadContent(params.path, content, params.offset, params.limit);
		}
	};
}
function createReadTool(env, packagedSkills) {
	return {
		name: "read",
		label: "Read File",
		description: "Read a file or list a directory. For files, output is truncated to 2000 lines or 50KB — use offset/limit for large files. For directories, returns the list of entries.",
		parameters: ReadParams,
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			const packagedFile = readPackagedSkillFile(packagedSkills, params.path);
			if (packagedFile !== void 0) return formatReadContent(params.path, packagedFile, params.offset, params.limit);
			if (params.path.startsWith(PACKAGED_SKILLS_ROOT)) throw new Error(`[flue] Packaged skill file not found: ${params.path}`);
			try {
				if ((await env.stat(params.path)).isDirectory) {
					const entries = await env.readdir(params.path);
					return {
						content: [{
							type: "text",
							text: entries.join("\n") || "(empty directory)"
						}],
						details: {
							path: params.path,
							isDirectory: true,
							entries: entries.length
						}
					};
				}
			} catch {}
			const content = await env.readFile(params.path);
			return formatReadContent(params.path, content, params.offset, params.limit);
		}
	};
}
const WriteParams = Type.Object({
	path: Type.String({ description: "Path to the file to write" }),
	content: Type.String({ description: "Content to write to the file" })
});
function createWriteTool(env) {
	return {
		name: "write",
		label: "Write File",
		description: "Write content to a file. Creates the file and parent directories if they do not exist.",
		parameters: WriteParams,
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			const resolved = env.resolvePath(params.path);
			const dir = resolved.replace(/\/[^/]*$/, "");
			if (dir && dir !== resolved) await env.mkdir(dir, { recursive: true });
			await env.writeFile(resolved, params.content);
			return {
				content: [{
					type: "text",
					text: `Successfully wrote ${params.content.length} bytes to ${params.path}`
				}],
				details: {
					path: params.path,
					size: params.content.length
				}
			};
		}
	};
}
const EditParams = Type.Object({
	path: Type.String({ description: "Path to the file to edit" }),
	oldText: Type.String({ description: "Exact text to find (must be unique)" }),
	newText: Type.String({ description: "Replacement text" }),
	replaceAll: Type.Optional(Type.Boolean({ description: "Replace all occurrences" }))
});
function createEditTool(env) {
	return {
		name: "edit",
		label: "Edit File",
		description: "Edit a file using exact text replacement. The oldText must match a unique region of the file. Use replaceAll to replace all occurrences.",
		parameters: EditParams,
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			const content = await env.readFile(params.path);
			if (params.replaceAll) {
				const newContent = content.replaceAll(params.oldText, params.newText);
				if (newContent === content) throw new Error(`Could not find the text in ${params.path}. No changes made.`);
				await env.writeFile(params.path, newContent);
				const count = content.split(params.oldText).length - 1;
				return {
					content: [{
						type: "text",
						text: `Replaced ${count} occurrences in ${params.path}`
					}],
					details: {
						path: params.path,
						replacements: count
					}
				};
			}
			const occurrences = countOccurrences(content, params.oldText);
			if (occurrences === 0) throw new Error(`Could not find the exact text in ${params.path}. Make sure your oldText matches exactly, including whitespace and indentation.`);
			if (occurrences > 1) throw new Error(`Found ${occurrences} occurrences of the text in ${params.path}. Provide more surrounding context to make the match unique, or use replaceAll.`);
			const newContent = content.replace(params.oldText, params.newText);
			await env.writeFile(params.path, newContent);
			return {
				content: [{
					type: "text",
					text: `Successfully edited ${params.path}`
				}],
				details: { path: params.path }
			};
		}
	};
}
const BashParams = Type.Object({
	command: Type.String({ description: "Bash command to execute" }),
	timeout: Type.Optional(Type.Number({ description: "Timeout in seconds" }))
});
function createBashTool(env) {
	return {
		name: "bash",
		label: "Run Command",
		description: "Execute a bash command. Returns stdout and stderr. Output is truncated to the last 2000 lines or 50KB.",
		parameters: BashParams,
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			const timeoutSignal = typeof params.timeout === "number" ? AbortSignal.timeout(params.timeout * 1e3) : void 0;
			const execSignal = signal && timeoutSignal ? AbortSignal.any([signal, timeoutSignal]) : signal ?? timeoutSignal;
			const timedOut = () => formatBashResult({
				stdout: "",
				stderr: `[flue] Command timed out after ${params.timeout} seconds.`,
				exitCode: 124
			}, params.command);
			try {
				const result = await env.exec(params.command, {
					timeout: params.timeout,
					signal: execSignal
				});
				if (timeoutSignal?.aborted && !signal?.aborted) return timedOut();
				return formatBashResult(result, params.command);
			} catch (err) {
				if (timeoutSignal?.aborted && !signal?.aborted) return timedOut();
				throw err;
			}
		}
	};
}
const TaskParams = Type.Object({
	description: Type.Optional(Type.String({ description: "Short human-readable label for the delegated work" })),
	prompt: Type.String({ description: "Focused instructions for the child agent" }),
	agent: Type.Optional(Type.String({ description: "Declared subagent to use for the child agent" })),
	cwd: Type.Optional(Type.String({ description: "Working directory for the child agent. AGENTS.md and skills are discovered from here." }))
});
/** Build Flue's framework-owned `task` tool. */
function createTaskTool(runTask, subagents) {
	const agentNames = Object.keys(subagents);
	return {
		name: "task",
		label: "Run Task",
		description: "Delegate a focused task to a detached child agent with its own context. Use this for independent research, file exploration, or parallel work. The task returns only its final answer to this conversation." + (agentNames.length > 0 ? ` Available agents: ${agentNames.join(", ")}.` : " No subagents are currently defined."),
		parameters: TaskParams,
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			return runTask(params, signal);
		}
	};
}
function createActivateSkillTool(skillNames, activate) {
	const sortedNames = [...skillNames].sort();
	const NameSchema = sortedNames.length === 1 ? Type.Literal(sortedNames[0] ?? "") : Type.Union(sortedNames.map((name) => Type.Literal(name)));
	return {
		name: "activate_skill",
		label: "Activate Skill",
		description: "Load the full instructions for one available skill before performing work that matches its description. Supporting resources remain lazy until explicitly read.",
		parameters: Type.Object({ name: NameSchema }),
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			const name = typeof params === "object" && params !== null && "name" in params && typeof params.name === "string" ? params.name : "";
			return {
				content: [{
					type: "text",
					text: await activate(name, signal)
				}],
				details: { skill: name }
			};
		}
	};
}
function formatBashResult(result, command) {
	const { text: output } = truncateTail((result.stdout + (result.stderr ? `\n${result.stderr}` : "")).trim(), MAX_READ_LINES, MAX_READ_BYTES);
	const exitLine = `Command exited with code ${result.exitCode}`;
	return {
		content: [{
			type: "text",
			text: result.exitCode === 0 ? output || "(no output)" : `${output || "(no output)"}\n\n${exitLine}`
		}],
		details: {
			command,
			exitCode: result.exitCode
		}
	};
}
const GrepParams = Type.Object({
	pattern: Type.String({ description: "Search pattern (regex)" }),
	path: Type.Optional(Type.String({ description: "Directory or file to search (default: .)" })),
	include: Type.Optional(Type.String({ description: "Glob filter, e.g. \"*.ts\"" }))
});
function createGrepTool(env) {
	return {
		name: "grep",
		label: "Search Files",
		description: "Search file contents for a regex pattern. Returns matching lines with file paths and line numbers.",
		parameters: GrepParams,
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			const searchPath = params.path || ".";
			let cmd = `grep -rn ${shellQuote(params.pattern)} ${shellQuote(searchPath)}`;
			if (params.include) cmd = `grep -rn --include=${shellQuote(params.include)} ${shellQuote(params.pattern)} ${shellQuote(searchPath)}`;
			const result = await env.exec(cmd);
			if (result.exitCode === 1 && !result.stdout.trim()) return {
				content: [{
					type: "text",
					text: "No matches found."
				}],
				details: { matchCount: 0 }
			};
			if (result.exitCode > 1) throw new Error(`grep failed: ${result.stderr}`);
			const lines = result.stdout.trim().split("\n");
			let finalOutput = lines.slice(0, MAX_GREP_MATCHES).map((line) => line.length > MAX_GREP_LINE_LENGTH ? `${line.slice(0, MAX_GREP_LINE_LENGTH)}...` : line).join("\n");
			if (lines.length > MAX_GREP_MATCHES) finalOutput += `\n\n[Showing ${MAX_GREP_MATCHES} of ${lines.length} matches. Narrow your search.]`;
			return {
				content: [{
					type: "text",
					text: finalOutput
				}],
				details: { matchCount: Math.min(lines.length, MAX_GREP_MATCHES) }
			};
		}
	};
}
const GlobParams = Type.Object({
	pattern: Type.String({ description: "Filename pattern, e.g. \"*.ts\"" }),
	path: Type.Optional(Type.String({ description: "Directory to search in (default: .)" }))
});
function createGlobTool(env) {
	return {
		name: "glob",
		label: "Find Files",
		description: "Find files by filename pattern using shell find -name semantics. Returns matching file paths.",
		parameters: GlobParams,
		async execute(_toolCallId, params, signal) {
			throwIfAborted(signal);
			const cmd = `find ${shellQuote(params.path || ".")} -type f -name ${shellQuote(params.pattern)} 2>/dev/null | head -${MAX_GLOB_RESULTS}`;
			const result = await env.exec(cmd);
			if (result.exitCode !== 0 && !result.stdout.trim()) return {
				content: [{
					type: "text",
					text: "No files found matching pattern."
				}],
				details: { matchCount: 0 }
			};
			const paths = result.stdout.trim().split("\n").filter(Boolean);
			if (paths.length === 0) return {
				content: [{
					type: "text",
					text: "No files found matching pattern."
				}],
				details: { matchCount: 0 }
			};
			return {
				content: [{
					type: "text",
					text: paths.join("\n")
				}],
				details: { matchCount: paths.length }
			};
		}
	};
}
function throwIfAborted(signal) {
	if (signal?.aborted) throw new Error("Operation aborted");
}
function readPackagedSkillFile(skills, path) {
	for (const skill of Object.values(skills)) for (const [filePath, file] of Object.entries(skill.files)) {
		if (path !== packagedSkillReadPath(skill.id, filePath)) continue;
		return file.kind === "binary" ? wrapBase64ForReading(file.content) : new TextDecoder().decode(Uint8Array.from(atob(file.content), (character) => character.charCodeAt(0)));
	}
}
function wrapBase64ForReading(content) {
	const lines = [];
	for (let offset = 0; offset < content.length; offset += BASE64_READ_LINE_LENGTH) lines.push(content.slice(offset, offset + BASE64_READ_LINE_LENGTH));
	return lines.join("\n");
}
function formatReadContent(path, content, offset, limit) {
	const allLines = content.split("\n");
	const startLine = offset ? Math.max(0, offset - 1) : 0;
	if (startLine >= allLines.length) throw new Error(`Offset ${offset} is beyond end of file (${allLines.length} lines total)`);
	const endLine = limit ? startLine + limit : allLines.length;
	const { text: truncatedText, wasTruncated } = truncateHead(allLines.slice(startLine, endLine), MAX_READ_LINES, MAX_READ_BYTES);
	let output = truncatedText;
	if (wasTruncated) {
		const shownEnd = startLine + truncatedText.split("\n").length;
		output += `\n\n[Showing lines ${startLine + 1}-${shownEnd} of ${allLines.length}. Use offset=${shownEnd + 1} to continue.]`;
	}
	return {
		content: [{
			type: "text",
			text: output
		}],
		details: {
			path,
			lines: allLines.length
		}
	};
}
function formatPackagedSkillFilePath(skillId, filePath) {
	return packagedSkillReadPath(skillId, filePath);
}
function packagedSkillReadPath(skillId, filePath) {
	return `/.flue/packaged-skills/${encodeURIComponent(skillId)}/${filePath}`;
}
function countOccurrences(str, substr) {
	let count = 0;
	let pos = str.indexOf(substr, 0);
	while (pos !== -1) {
		count++;
		pos = str.indexOf(substr, pos + substr.length);
	}
	return count;
}
function shellQuote(arg) {
	return `'${arg.replace(/'/g, `'\\''`)}'`;
}
function truncateHead(lines, maxLines, maxBytes) {
	let result = "";
	let lineCount = 0;
	let wasTruncated = false;
	for (const line of lines) {
		if (lineCount >= maxLines) {
			wasTruncated = true;
			break;
		}
		const next = lineCount === 0 ? line : `\n${line}`;
		if (result.length + next.length > maxBytes) {
			wasTruncated = true;
			break;
		}
		result += next;
		lineCount++;
	}
	return {
		text: result,
		wasTruncated
	};
}
function truncateTail(text, maxLines, maxBytes) {
	const lines = text.split("\n");
	if (lines.length <= maxLines && text.length <= maxBytes) return {
		text,
		wasTruncated: false
	};
	let result = lines.slice(-maxLines).join("\n");
	if (result.length > maxBytes) result = result.slice(-maxBytes);
	return {
		text: result,
		wasTruncated: true
	};
}
//#endregion
//#region src/skill-frontmatter.ts
function parseSkillMarkdown(content, options) {
	const match = content.match(/^---\s*\r?\n([\s\S]*?)\r?\n---\s*(?:\r?\n|$)([\s\S]*)$/);
	if (!match) throw new Error(`[flue] Skill ${options.path} is missing YAML frontmatter. Start SKILL.md with "---", include "name" and "description", then close the block with "---".`);
	let raw;
	try {
		raw = load(match[1] ?? "");
	} catch (error) {
		const detail = error instanceof Error ? ` ${error.message}` : "";
		throw new Error(`[flue] Skill ${options.path} has invalid YAML frontmatter.${detail}`);
	}
	if (!isRecord(raw)) throw new Error(`[flue] Skill ${options.path} frontmatter must be a YAML mapping.`);
	const name = requireString(raw.name, options.path, "name");
	validateSkillName(name, options);
	const description = requireString(raw.description, options.path, "description");
	if (description.length > 1024) throw new Error(`[flue] Skill ${options.path} frontmatter description exceeds the 1024-character Agent Skills limit. Shorten "description" to a concise one-line summary.`);
	const license = optionalString(raw.license, options.path, "license");
	const compatibility = optionalString(raw.compatibility, options.path, "compatibility");
	if (compatibility !== void 0 && compatibility.length > 500) throw new Error(`[flue] Skill ${options.path} compatibility must be at most 500 characters.`);
	return {
		name,
		description,
		body: (match[2] ?? "").trim(),
		license,
		compatibility,
		metadata: parseMetadata(raw.metadata, options.path),
		allowedTools: parseAllowedTools(raw["allowed-tools"], options.path)
	};
}
function validateSkillName(name, options) {
	if (name.length > 64) throw new Error(`[flue] Skill ${options.path} name must be at most 64 characters.`);
	if (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(name)) throw new Error(`[flue] Skill ${options.path} frontmatter name "${name}" must contain only lowercase letters, numbers, and single internal hyphens. Use a spec-compliant value such as "review-pr".`);
	if (name !== options.directoryName) throw new Error(`[flue] Skill ${options.path} declares frontmatter name "${name}", but Agent Skills requires it to match directory "${options.directoryName}"; names must match. Rename the directory or change "name" so they match.`);
}
function requireString(value, path, field) {
	if (typeof value !== "string" || value.trim().length === 0) throw new Error(`[flue] Skill ${path} must define frontmatter ${field} as a non-empty string.`);
	return value.trim();
}
function optionalString(value, path, field) {
	if (value === void 0 || value === null) return void 0;
	if (typeof value !== "string" || value.trim().length === 0) throw new Error(`[flue] Skill ${path} frontmatter ${field} must be a non-empty string when provided.`);
	return value.trim();
}
function parseMetadata(value, path) {
	if (value === void 0 || value === null) return void 0;
	if (!isRecord(value)) throw new Error(`[flue] Skill ${path} frontmatter metadata must be a string-to-string mapping.`);
	const entries = Object.entries(value);
	if (entries.some(([, metadataValue]) => typeof metadataValue !== "string")) throw new Error(`[flue] Skill ${path} frontmatter metadata must be a string-to-string mapping. Quote scalar values such as version: "1.0".`);
	return Object.fromEntries(entries);
}
function parseAllowedTools(value, path) {
	if (value === void 0 || value === null) return void 0;
	if (typeof value !== "string") throw new Error(`[flue] Skill ${path} frontmatter allowed-tools must be a string when provided.`);
	const tools = value.trim().split(/\s+/).filter(Boolean);
	return tools.length > 0 ? tools : void 0;
}
function isRecord(value) {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}
//#endregion
//#region src/result.ts
/**
* Names of the framework-injected tools used to capture structured results.
* Surfaced for diagnostics and logging; not part of the public API.
*/
const FINISH_TOOL_NAME = "finish";
const GIVE_UP_TOOL_NAME = "give_up";
/** Footer appended to user prompts/skill bodies when a `result` schema is set. */
function buildResultFooter() {
	return [
		"",
		`When the task is complete, call the \`${FINISH_TOOL_NAME}\` tool with your final answer as its arguments. The arguments are validated against the required schema; if validation fails you will receive an error and may try again.`,
		`If you determine that you cannot complete the task or cannot produce a result that conforms to the required schema, call the \`${GIVE_UP_TOOL_NAME}\` tool with a clear \`reason\`.`,
		`Do not respond with the answer in plain text — only a successful \`${FINISH_TOOL_NAME}\` (or \`${GIVE_UP_TOOL_NAME}\`) call counts.`
	].join("\n");
}
/** Follow-up prompt sent when the LLM ends a turn without calling `finish` or `give_up`. */
function buildResultFollowUpPrompt() {
	return [`You ended your turn without calling \`${FINISH_TOOL_NAME}\` or \`${GIVE_UP_TOOL_NAME}\`.`, `Either call \`${FINISH_TOOL_NAME}\` with your final answer, or call \`${GIVE_UP_TOOL_NAME}\` with a reason if you cannot determine the answer.`].join(" ");
}
function buildPackagedSkillPrompt(reference, directory, args, schema) {
	const skillFile = directory.files["SKILL.md"];
	if (!skillFile) throw new Error(`[flue] Packaged skill "${reference.name}" is missing SKILL.md.`);
	const skill = parseSkillMarkdown(new TextDecoder().decode(Uint8Array.from(atob(skillFile.content), (character) => character.charCodeAt(0))), {
		directoryName: reference.name,
		path: `${reference.name}/SKILL.md`
	});
	const parts = [
		`Run the skill named "${reference.name}".`,
		"",
		"<skill_instructions>",
		skill.body,
		"</skill_instructions>"
	];
	const resources = Object.keys(directory.files).filter((filePath) => filePath !== "SKILL.md").sort();
	if (resources.length > 0) parts.push("", "Supporting skill resources are available but are not loaded into context unless needed:", "<skill_resources>", ...resources.map((filePath) => `- ${filePath} → read ${formatPackagedSkillFilePath(reference.id, filePath)}`), "</skill_resources>");
	if (args && Object.keys(args).length > 0) parts.push("", "Arguments:", JSON.stringify(args, null, 2));
	if (schema) parts.push(buildResultFooter());
	return parts.join("\n");
}
function buildWorkspaceSkillPrompt(name, directory, skillMdPath, raw) {
	const skill = parseSkillMarkdown(raw, {
		directoryName: name,
		path: skillMdPath
	});
	return [
		`Run the skill named "${name}".`,
		"",
		"<skill_instructions>",
		skill.body,
		"</skill_instructions>",
		"",
		"Supporting skill resources are available relative to this workspace skill directory but are not loaded into context unless needed:",
		"<skill_resources>",
		`- Base directory: ${directory}`,
		"- Resolve relative resource paths from this directory and read only the files you need.",
		"</skill_resources>"
	].join("\n");
}
/** Build the existing name-only prompt for runtime-discovered sandbox skills. */
function buildSkillByPathlessNamePrompt(name, args, schema) {
	const parts = [`Run the skill named "${name}".`];
	if (args && Object.keys(args).length > 0) parts.push("", "Arguments:", JSON.stringify(args, null, 2));
	if (schema) parts.push(buildResultFooter());
	return parts.join("\n");
}
function buildPromptText(text, schema) {
	const parts = [text];
	if (schema) parts.push(buildResultFooter());
	return parts.join("\n");
}
/**
* Produce the per-call `finish` and `give_up` tool pair for a given valibot schema.
*
* - `finish`'s parameters are derived from the schema via `@valibot/to-json-schema`.
*   Non-object top-level schemas are wrapped in a `{ result: <schema> }` envelope
*   because every LLM provider expects tool arguments to be a top-level object.
* - Pi-agent-core validates args against the JSON Schema before calling `execute`.
*   Inside `execute` we additionally run `valibot.safeParse` to enforce
*   valibot-specific refinements and to obtain the parsed output (transforms,
*   defaults, coercion). On valibot failure we throw — pi-agent-core surfaces
*   the throw as a tool-error tool-result, so the LLM can self-correct.
* - First successful `finish` (or `give_up`) call wins. Subsequent calls return
*   a tool error rather than throwing, to keep the conversation transcript natural.
* - Successful calls set `terminate: true` so pi-agent-core ends the loop after
*   the current tool batch.
*/
function createResultTools(schema) {
	let outcome = { type: "pending" };
	const wrapped = needsEnvelope(schema);
	const innerJsonSchema = stripJsonSchemaMeta(toJsonSchema(schema, { errorMode: "ignore" }));
	return {
		tools: [{
			name: FINISH_TOOL_NAME,
			label: FINISH_TOOL_NAME,
			description: `Call this tool when the task is complete. Provide your final answer as the arguments. The arguments are validated against the required schema; if validation fails you will receive an error message and may try again. The first successful \`${FINISH_TOOL_NAME}\` call wins — once the task is finished, do not call \`${FINISH_TOOL_NAME}\` again.`,
			parameters: wrapped ? {
				type: "object",
				properties: { result: innerJsonSchema },
				required: ["result"],
				additionalProperties: false
			} : innerJsonSchema,
			async execute(_toolCallId, params) {
				if (outcome.type !== "pending") return alreadyDoneToolError(outcome);
				const candidate = wrapped ? params.result : params;
				const parsed = v.safeParse(schema, candidate);
				if (!parsed.success) {
					const issues = parsed.issues.map((i) => i.message + (i.path ? ` (at ${formatIssuePath(i.path)})` : "")).join("; ");
					throw new Error(`Result does not match the required schema: ${issues}. Please call \`${FINISH_TOOL_NAME}\` again with a corrected payload.`);
				}
				outcome = {
					type: "finished",
					value: parsed.output
				};
				return {
					content: [{
						type: "text",
						text: "Result accepted. The task is complete."
					}],
					details: {
						tool: FINISH_TOOL_NAME,
						result: parsed.output
					},
					terminate: true
				};
			}
		}, {
			name: GIVE_UP_TOOL_NAME,
			label: GIVE_UP_TOOL_NAME,
			description: "Call this tool only if you have determined that you cannot complete the task or cannot produce a result that conforms to the required schema. Provide a clear `reason`. This ends the task with a failure.",
			parameters: {
				type: "object",
				properties: { reason: {
					type: "string",
					minLength: 1,
					description: "A clear explanation of why the task cannot be completed."
				} },
				required: ["reason"],
				additionalProperties: false
			},
			async execute(_toolCallId, params) {
				if (outcome.type !== "pending") return alreadyDoneToolError(outcome);
				const reason = params.reason;
				if (typeof reason !== "string" || reason.trim().length === 0) throw new Error(`\`${GIVE_UP_TOOL_NAME}\` requires a non-empty \`reason\` string.`);
				outcome = {
					type: "gave_up",
					reason
				};
				return {
					content: [{
						type: "text",
						text: "Acknowledged."
					}],
					details: {
						tool: GIVE_UP_TOOL_NAME,
						reason
					},
					terminate: true
				};
			}
		}],
		getOutcome: () => outcome
	};
}
function needsEnvelope(schema) {
	return schema.type !== "object";
}
function stripJsonSchemaMeta(jsonSchema) {
	const { $schema: _schema, ...rest } = jsonSchema;
	return rest;
}
function formatIssuePath(path) {
	return path.map((p) => typeof p.key === "number" ? `[${p.key}]` : `.${String(p.key ?? "?")}`).join("").replace(/^\./, "");
}
function alreadyDoneToolError(outcome) {
	return {
		content: [{
			type: "text",
			text: `${outcome.type === "finished" ? "A result was already submitted; the task is complete." : "The task was already given up; it cannot be resumed."} Do not call this tool again.`
		}],
		details: { alreadyDone: true }
	};
}
/**
* Thrown when the LLM calls the `give_up` tool, indicating it cannot produce a
* result that conforms to the required schema. Carries the LLM-supplied
* `reason` and the assistant transcript leading up to the give-up.
*/
var ResultUnavailableError = class extends Error {
	reason;
	assistantText;
	constructor(reason, assistantText) {
		super(`The agent gave up: ${reason}`);
		this.reason = reason;
		this.assistantText = assistantText;
		this.name = "ResultUnavailableError";
	}
};
//#endregion
//#region src/cloudflare-model.ts
/** Pi-ai `Api` slug for the binding-backed Workers AI provider. */
const CLOUDFLARE_AI_BINDING_API = "cloudflare-ai-binding";
//#endregion
//#region src/runtime/providers.ts
/** Runtime provider registries consumed by `resolveModel` and Session. */
/**
* pi-ai's open-ended `Api` type prevents direct discriminator narrowing.
*/
function isCloudflareBindingRegistration(def) {
	return def.api === CLOUDFLARE_AI_BINDING_API;
}
/**
* Provider registry populated at module init by `app.ts` and generated
* server entries.
*/
const providersById = /* @__PURE__ */ new Map();
/**
* Register a model provider keyed by the provider ID used in model specifiers.
*
* Last-write-wins. On Cloudflare, registering the `cloudflare` provider ID in
* `app.ts` takes precedence over the generated Workers AI binding default.
*/
function registerProvider(providerId, registration) {
	providersById.set(providerId, registration);
}
function resetProvidersForTests() {
	providersById.clear();
	providerSettingsById.clear();
}
/** Whether a provider ID has already been registered. */
function hasRegisteredProvider(providerId) {
	return providersById.has(providerId);
}
/** Look up an API key registered for a provider ID. */
function getRegisteredApiKey(providerId) {
	const registration = providersById.get(providerId);
	if (!registration || isCloudflareBindingRegistration(registration)) return void 0;
	return registration.apiKey;
}
/**
* Re-export of pi-ai's `registerApiProvider`. Use to register a brand-new
* wire-protocol handler for an `api` slug pi-ai doesn't ship. Then call
* {@link registerProvider} to associate a provider ID with that api.
*
* ```ts
* registerApiProvider({ api: 'my-novel-api', stream, streamSimple });
* registerProvider('thing', { api: 'my-novel-api', baseUrl: '...', apiKey: '...' });
* ```
*
* pi-ai's registry is also module-scoped and last-write-wins. Calling
* `registerApiProvider` repeatedly with the same `api` string overwrites,
* so generated code can register on every isolate boot without dedupe
* bookkeeping.
*/
const registerApiProvider$1 = registerApiProvider;
const providerSettingsById = /* @__PURE__ */ new Map();
/**
* Configure transport-level settings on an existing provider while preserving
* its resolved Model metadata (cost, context window, token limits, etc.).
* Repeated calls for the same provider ID replace the previous settings object.
*
* ```ts
* import { configureProvider } from '@flue/runtime';
*
* configureProvider('anthropic', {
*   baseUrl: 'https://gateway.example.com/anthropic',
*   apiKey: process.env.GATEWAY_KEY,
* });
* ```
*
* Keyed by provider ID. Last-write-wins.
*/
function configureProvider(providerId, settings) {
	providerSettingsById.set(providerId, settings);
}
/** Internal read accessor for provider settings. */
function getProviderConfiguration(providerId) {
	return providerSettingsById.get(providerId);
}
/** Attach a Workers AI binding (and optional gateway options) to a Model literal. */
function attachModelBinding(model, binding, gateway) {
	return {
		...model,
		binding,
		gateway
	};
}
/**
* Read a Workers AI binding off a resolved Model, or `undefined` if no
* usable binding is attached.
*/
function getModelBinding(model) {
	const candidate = model.binding;
	if (!candidate || typeof candidate.run !== "function") return;
	return candidate;
}
/** Resolve `'provider-id/model-id'` against the provider registry. */
function resolveRegisteredModel(providerId, modelId) {
	const registration = providersById.get(providerId);
	if (!registration) return void 0;
	return buildModelFromRegistration(providerId, registration, modelId);
}
/**
* Construct a pi-ai Model from a registered provider template. Binding
* registrations hydrate metadata from pi-ai's `cloudflare-workers-ai`
* catalog; HTTP registrations supply their own metadata, with any unset
* fields defaulting to zero.
*/
function buildModelFromRegistration(providerId, registration, modelId) {
	if (isCloudflareBindingRegistration(registration)) {
		const catalog = getModel("cloudflare-workers-ai", modelId);
		return attachModelBinding(catalog ? {
			...catalog,
			api: CLOUDFLARE_AI_BINDING_API,
			provider: providerId,
			baseUrl: ""
		} : {
			id: modelId,
			name: modelId,
			api: CLOUDFLARE_AI_BINDING_API,
			provider: providerId,
			baseUrl: "",
			reasoning: false,
			input: ["text"],
			cost: {
				input: 0,
				output: 0,
				cacheRead: 0,
				cacheWrite: 0
			},
			contextWindow: 0,
			maxTokens: 0
		}, registration.binding, registration.gateway);
	}
	return {
		id: modelId,
		name: modelId,
		api: registration.api,
		provider: providerId,
		baseUrl: registration.baseUrl,
		reasoning: false,
		input: ["text"],
		cost: {
			input: 0,
			output: 0,
			cacheRead: 0,
			cacheWrite: 0
		},
		contextWindow: registration.models?.[modelId]?.contextWindow ?? registration.contextWindow ?? 0,
		maxTokens: registration.models?.[modelId]?.maxTokens ?? registration.maxTokens ?? 0,
		headers: registration.headers
	};
}
//#endregion
//#region src/usage.ts
/** All-zero `PromptUsage`. Identity element for `addUsage`. */
function emptyUsage() {
	return {
		input: 0,
		output: 0,
		cacheRead: 0,
		cacheWrite: 0,
		totalTokens: 0,
		cost: {
			input: 0,
			output: 0,
			cacheRead: 0,
			cacheWrite: 0,
			total: 0
		}
	};
}
/**
* Field-wise sum of two `PromptUsage` values, including the nested `cost`
* sub-object. Returns a fresh object; neither argument is mutated.
*/
function addUsage(a, b) {
	return {
		input: a.input + b.input,
		output: a.output + b.output,
		cacheRead: a.cacheRead + b.cacheRead,
		cacheWrite: a.cacheWrite + b.cacheWrite,
		totalTokens: a.totalTokens + b.totalTokens,
		cost: {
			input: a.cost.input + b.cost.input,
			output: a.cost.output + b.cost.output,
			cacheRead: a.cost.cacheRead + b.cost.cacheRead,
			cacheWrite: a.cost.cacheWrite + b.cost.cacheWrite,
			total: a.cost.total + b.cost.total
		}
	};
}
/**
* Convert pi-ai's `Usage` into Flue's public `PromptUsage`. The shapes are
* structurally identical today, but going through this normalizer keeps
* Flue's public types decoupled from pi-ai's so future divergence in
* pi-ai (e.g. additional fields) doesn't leak into the runtime package's public
* surface. Returns `undefined` when the input is `undefined`.
*/
function fromProviderUsage(usage) {
	if (!usage) return void 0;
	return {
		input: usage.input,
		output: usage.output,
		cacheRead: usage.cacheRead,
		cacheWrite: usage.cacheWrite,
		totalTokens: usage.totalTokens,
		cost: {
			input: usage.cost.input,
			output: usage.cost.output,
			cacheRead: usage.cost.cacheRead,
			cacheWrite: usage.cost.cacheWrite,
			total: usage.cost.total
		}
	};
}
//#endregion
//#region src/compaction.ts
/**
* Defaults applied when no user config and no model metadata are available.
* Real sessions construct settings via {@link deriveCompactionDefaults} so
* headroom tracks the active model instead of a fixed Sonnet-sized window.
*/
const DEFAULT_COMPACTION_SETTINGS = {
	enabled: true,
	reserveTokens: 2e4,
	keepRecentTokens: 8e3
};
/**
* Compute model-aware defaults. Reserve is capped at the model's max output
* because reserving more than the model can emit in one turn wastes context;
* the preserved tail stays flat because recent-context fidelity depends on
* the active work, not on the model's total window size.
*
* Caller may override either field after calling this.
*/
function deriveCompactionDefaults(input) {
	const reserveCap = input.maxTokens > 0 ? input.maxTokens : DEFAULT_COMPACTION_SETTINGS.reserveTokens;
	let reserveTokens = Math.min(DEFAULT_COMPACTION_SETTINGS.reserveTokens, reserveCap);
	if (input.contextWindow > 0 && reserveTokens * 2 >= input.contextWindow) reserveTokens = Math.max(1024, Math.floor(input.contextWindow / 3));
	return {
		enabled: true,
		reserveTokens,
		keepRecentTokens: DEFAULT_COMPACTION_SETTINGS.keepRecentTokens
	};
}
function calculateContextTokens(usage) {
	return usage.totalTokens || usage.input + usage.output + usage.cacheRead + usage.cacheWrite;
}
function getAssistantUsage(msg) {
	if (msg.role === "assistant" && "usage" in msg) {
		const assistantMsg = msg;
		if (assistantMsg.stopReason !== "aborted" && assistantMsg.stopReason !== "error" && assistantMsg.usage) return assistantMsg.usage;
	}
}
function getLastAssistantUsageInfo(messages) {
	for (let i = messages.length - 1; i >= 0; i--) {
		const msg = messages[i];
		if (!msg) continue;
		const usage = getAssistantUsage(msg);
		if (usage) return {
			usage,
			index: i
		};
	}
}
/** chars/4 heuristic. Conservative (overestimates). */
function estimateTokens(message) {
	let chars = 0;
	switch (message.role) {
		case "user": {
			const { content } = message;
			if (typeof content === "string") chars = content.length;
			else if (Array.isArray(content)) {
				for (const block of content) if (block.type === "text") chars += block.text.length;
			}
			return Math.ceil(chars / 4);
		}
		case "assistant": {
			const { content } = message;
			for (const block of content) if (block.type === "text") chars += block.text.length;
			else if (block.type === "thinking") chars += block.thinking.length;
			else if (block.type === "toolCall") chars += block.name.length + JSON.stringify(block.arguments).length;
			return Math.ceil(chars / 4);
		}
		case "toolResult": {
			const { content } = message;
			for (const block of content) if (block.type === "text") chars += block.text.length;
			else if (block.type === "image") chars += 4800;
			return Math.ceil(chars / 4);
		}
	}
	return 0;
}
function estimateContextTokens(messages) {
	const usageInfo = getLastAssistantUsageInfo(messages);
	if (!usageInfo) {
		let estimated = 0;
		for (const message of messages) estimated += estimateTokens(message);
		return {
			tokens: estimated,
			usageTokens: 0,
			trailingTokens: estimated,
			lastUsageIndex: null
		};
	}
	const usageTokens = calculateContextTokens(usageInfo.usage);
	let trailingTokens = 0;
	for (let i = usageInfo.index + 1; i < messages.length; i++) {
		const message = messages[i];
		if (message) trailingTokens += estimateTokens(message);
	}
	return {
		tokens: usageTokens + trailingTokens,
		usageTokens,
		trailingTokens,
		lastUsageIndex: usageInfo.index
	};
}
function shouldCompact(contextTokens, contextWindow, settings) {
	if (!settings.enabled) return false;
	if (contextWindow <= 0) return false;
	return contextTokens > contextWindow - settings.reserveTokens;
}
function createFileOps() {
	return {
		read: /* @__PURE__ */ new Set(),
		written: /* @__PURE__ */ new Set(),
		edited: /* @__PURE__ */ new Set()
	};
}
function extractFileOpsFromMessage(message, fileOps) {
	if (message.role !== "assistant") return;
	const assistant = message;
	if (!Array.isArray(assistant.content)) return;
	for (const block of assistant.content) {
		if (block.type !== "toolCall") continue;
		const args = block.arguments;
		if (!args) continue;
		const path = typeof args.path === "string" ? args.path : void 0;
		if (!path) continue;
		switch (block.name) {
			case "read":
				fileOps.read.add(path);
				break;
			case "write":
				fileOps.written.add(path);
				break;
			case "edit":
				fileOps.edited.add(path);
				break;
		}
	}
}
function computeFileLists(fileOps) {
	const modified = new Set([...fileOps.edited, ...fileOps.written]);
	return {
		readFiles: [...fileOps.read].filter((f) => !modified.has(f)).sort(),
		modifiedFiles: [...modified].sort()
	};
}
function formatFileOperations(readFiles, modifiedFiles) {
	const sections = [];
	if (readFiles.length > 0) sections.push(`<read-files>\n${readFiles.join("\n")}\n</read-files>`);
	if (modifiedFiles.length > 0) sections.push(`<modified-files>\n${modifiedFiles.join("\n")}\n</modified-files>`);
	if (sections.length === 0) return "";
	return `\n\n${sections.join("\n\n")}`;
}
const TOOL_RESULT_MAX_CHARS = 2e3;
function truncateForSummary(text, maxChars) {
	if (text.length <= maxChars) return text;
	const truncatedChars = text.length - maxChars;
	return `${text.slice(0, maxChars)}\n\n[... ${truncatedChars} more characters truncated]`;
}
/** Serialize messages to text so the summarization model doesn't treat it as a conversation to continue. */
function serializeConversation(messages) {
	const parts = [];
	for (const msg of messages) if (msg.role === "user") {
		const { content } = msg;
		const text = typeof content === "string" ? content : content.filter((c) => c.type === "text").map((c) => c.text).join("");
		if (text) parts.push(`[User]: ${text}`);
	} else if (msg.role === "assistant") {
		const { content } = msg;
		const textParts = [];
		const thinkingParts = [];
		const toolCalls = [];
		for (const block of content) if (block.type === "text") textParts.push(block.text);
		else if (block.type === "thinking") thinkingParts.push(block.thinking);
		else if (block.type === "toolCall") {
			const argsStr = Object.entries(block.arguments).map(([k, v]) => `${k}=${JSON.stringify(v)}`).join(", ");
			toolCalls.push(`${block.name}(${argsStr})`);
		}
		if (thinkingParts.length > 0) parts.push(`[Assistant thinking]: ${thinkingParts.join("\n")}`);
		if (textParts.length > 0) parts.push(`[Assistant]: ${textParts.join("\n")}`);
		if (toolCalls.length > 0) parts.push(`[Assistant tool calls]: ${toolCalls.join("; ")}`);
	} else if (msg.role === "toolResult") {
		const { content } = msg;
		const text = content.filter((c) => c.type === "text").map((c) => c.text).join("");
		if (text) parts.push(`[Tool result]: ${truncateForSummary(text, TOOL_RESULT_MAX_CHARS)}`);
	}
	return parts.join("\n\n");
}
const SUMMARIZATION_SYSTEM_PROMPT = "You are a context summarization assistant. Your task is to read a conversation between a user and an AI coding assistant, then produce a structured summary following the exact format specified.\n\nDo NOT continue the conversation. Do NOT respond to any questions in the conversation. ONLY output the structured summary.";
const SUMMARIZATION_PROMPT = `The messages above are a conversation to summarize. Create a structured context checkpoint summary that another LLM will use to continue the work.

Use this EXACT format:

## Goal
[What is the user trying to accomplish? Can be multiple items if the session covers different tasks.]

## Constraints & Preferences
- [Any constraints, preferences, or requirements mentioned by user]
- [Or "(none)" if none were mentioned]

## Progress
### Done
- [x] [Completed tasks/changes]

### In Progress
- [ ] [Current work]

### Blocked
- [Issues preventing progress, if any]

## Key Decisions
- **[Decision]**: [Brief rationale]

## Next Steps
1. [Ordered list of what should happen next]

## Critical Context
- [Any data, examples, or references needed to continue]
- [Or "(none)" if not applicable]

Keep each section concise. Preserve exact file paths, function names, and error messages.`;
const UPDATE_SUMMARIZATION_PROMPT = `The messages above are NEW conversation messages to incorporate into the existing summary provided in <previous-summary> tags.

Update the existing structured summary with new information. RULES:
- PRESERVE all existing information from the previous summary
- ADD new progress, decisions, and context from the new messages
- UPDATE the Progress section: move items from "In Progress" to "Done" when completed
- UPDATE "Next Steps" based on what was accomplished
- PRESERVE exact file paths, function names, and error messages
- If something is no longer relevant, you may remove it

Use this EXACT format:

## Goal
[Preserve existing goals, add new ones if the task expanded]

## Constraints & Preferences
- [Preserve existing, add new ones discovered]

## Progress
### Done
- [x] [Include previously done items AND newly completed items]

### In Progress
- [ ] [Current work - update based on progress]

### Blocked
- [Current blockers - remove if resolved]

## Key Decisions
- **[Decision]**: [Brief rationale] (preserve all previous, add new)

## Next Steps
1. [Update based on current state]

## Critical Context
- [Preserve important context, add new if needed]

Keep each section concise. Preserve exact file paths, function names, and error messages.`;
const TURN_PREFIX_SUMMARIZATION_PROMPT = `This is the PREFIX of a turn that was too large to keep. The SUFFIX (recent work) is retained.

Summarize the prefix to provide context for the retained suffix:

## Original Request
[What did the user ask for in this turn?]

## Early Progress
- [Key decisions and work done in the prefix]

## Context for Suffix
- [Information needed to understand the retained recent work]

Be concise. Focus on what's needed to understand the kept suffix.`;
/** Valid cut points: user or assistant messages. Never cut at toolResult. */
function findValidCutPoints(messages, start, end) {
	const cutPoints = [];
	for (let i = start; i < end; i++) {
		const role = messages[i]?.role;
		if (role === "user" || role === "assistant") cutPoints.push(i);
	}
	return cutPoints;
}
function findTurnStartIndex(messages, index, start) {
	for (let i = index; i >= start; i--) if (messages[i]?.role === "user") return i;
	return -1;
}
function findCutPoint(messages, start, end, keepRecentTokens) {
	const cutPoints = findValidCutPoints(messages, start, end);
	if (cutPoints.length === 0) return {
		firstKeptIndex: start,
		turnStartIndex: -1,
		isSplitTurn: false
	};
	let accumulatedTokens = 0;
	let cutIndex = cutPoints[0] ?? start;
	for (let i = end - 1; i >= start; i--) {
		const message = messages[i];
		if (!message) continue;
		const messageTokens = estimateTokens(message);
		accumulatedTokens += messageTokens;
		if (accumulatedTokens >= keepRecentTokens) {
			for (const cutPoint of cutPoints) if (cutPoint >= i) {
				cutIndex = cutPoint;
				break;
			}
			break;
		}
	}
	const isUserMessage = messages[cutIndex]?.role === "user";
	const turnStartIndex = isUserMessage ? -1 : findTurnStartIndex(messages, cutIndex, start);
	return {
		firstKeptIndex: cutIndex,
		turnStartIndex,
		isSplitTurn: !isUserMessage && turnStartIndex !== -1
	};
}
/** Pure function — no I/O. Finds cut point, extracts messages to summarize, tracks file ops. */
function prepareCompaction(messages, settings, previousCompaction) {
	if (messages.length === 0) return void 0;
	const boundaryStart = previousCompaction ? previousCompaction.firstKeptIndex : 0;
	const boundaryEnd = messages.length;
	const tokensBefore = estimateContextTokens(messages).tokens;
	const cutPoint = findCutPoint(messages, boundaryStart, boundaryEnd, settings.keepRecentTokens);
	if (cutPoint.firstKeptIndex <= boundaryStart) return void 0;
	const historyEnd = cutPoint.isSplitTurn ? cutPoint.turnStartIndex : cutPoint.firstKeptIndex;
	const messagesToSummarize = messages.slice(boundaryStart, historyEnd);
	const turnPrefixMessages = cutPoint.isSplitTurn ? messages.slice(cutPoint.turnStartIndex, cutPoint.firstKeptIndex) : [];
	const fileOps = createFileOps();
	if (previousCompaction?.details) {
		for (const f of previousCompaction.details.readFiles ?? []) fileOps.read.add(f);
		for (const f of previousCompaction.details.modifiedFiles ?? []) fileOps.edited.add(f);
	}
	for (const msg of messagesToSummarize) extractFileOpsFromMessage(msg, fileOps);
	for (const msg of turnPrefixMessages) extractFileOpsFromMessage(msg, fileOps);
	return {
		firstKeptIndex: cutPoint.firstKeptIndex,
		messagesToSummarize,
		turnPrefixMessages,
		isSplitTurn: cutPoint.isSplitTurn,
		tokensBefore,
		previousSummary: previousCompaction?.summary,
		fileOps,
		settings
	};
}
async function generateSummary(currentMessages, model, reserveTokens, apiKey, signal, previousSummary, observer) {
	const maxTokens = Math.min(Math.floor(.8 * reserveTokens), 16e3);
	const basePrompt = previousSummary ? UPDATE_SUMMARIZATION_PROMPT : SUMMARIZATION_PROMPT;
	let promptText = `<conversation>\n${serializeConversation(currentMessages)}\n</conversation>\n\n`;
	if (previousSummary) promptText += `<previous-summary>\n${previousSummary}\n</previous-summary>\n\n`;
	promptText += basePrompt;
	const summarizationMessages = [{
		role: "user",
		content: [{
			type: "text",
			text: promptText
		}],
		timestamp: Date.now()
	}];
	const completionOptions = {
		maxTokens,
		signal
	};
	if (apiKey) completionOptions.apiKey = apiKey;
	if (model.reasoning) completionOptions.reasoning = "high";
	const context = {
		systemPrompt: SUMMARIZATION_SYSTEM_PROMPT,
		messages: summarizationMessages
	};
	const handle = observer?.start("compaction", model, context, completionOptions);
	let response;
	try {
		response = await completeSimple(model, context, completionOptions);
		if (handle) observer?.end("compaction", handle, model, response, void 0);
	} catch (error) {
		if (handle) observer?.end("compaction", handle, model, void 0, error);
		throw error;
	}
	if (response.stopReason === "error") throw new Error(`Summarization failed: ${response.errorMessage || "Unknown error"}`);
	return {
		text: response.content.filter((c) => c.type === "text").map((c) => c.text).join("\n"),
		usage: response.usage
	};
}
async function generateTurnPrefixSummary(messages, model, reserveTokens, apiKey, signal, observer) {
	const maxTokens = Math.min(Math.floor(.5 * reserveTokens), 16e3);
	const summarizationMessages = [{
		role: "user",
		content: [{
			type: "text",
			text: `<conversation>\n${serializeConversation(messages)}\n</conversation>\n\n${TURN_PREFIX_SUMMARIZATION_PROMPT}`
		}],
		timestamp: Date.now()
	}];
	const completionOptions = {
		maxTokens,
		signal
	};
	if (apiKey) completionOptions.apiKey = apiKey;
	if (model.reasoning) completionOptions.reasoning = "high";
	const context = {
		systemPrompt: SUMMARIZATION_SYSTEM_PROMPT,
		messages: summarizationMessages
	};
	const handle = observer?.start("compaction_prefix", model, context, completionOptions);
	let response;
	try {
		response = await completeSimple(model, context, completionOptions);
		if (handle) observer?.end("compaction_prefix", handle, model, response, void 0);
	} catch (error) {
		if (handle) observer?.end("compaction_prefix", handle, model, void 0, error);
		throw error;
	}
	if (response.stopReason === "error") throw new Error(`Turn prefix summarization failed: ${response.errorMessage || "Unknown error"}`);
	return {
		text: response.content.filter((c) => c.type === "text").map((c) => c.text).join("\n"),
		usage: response.usage
	};
}
async function compact(preparation, model, apiKey, signal, observer) {
	const { firstKeptIndex, messagesToSummarize, turnPrefixMessages, isSplitTurn, tokensBefore, previousSummary, fileOps, settings } = preparation;
	let summary;
	let aggregateUsage;
	const addCallUsage = (usage) => {
		const normalized = fromProviderUsage(usage);
		if (!normalized) return;
		aggregateUsage = aggregateUsage ? addUsage(aggregateUsage, normalized) : normalized;
	};
	if (isSplitTurn && turnPrefixMessages.length > 0) {
		const [historyResult, turnPrefixResult] = await Promise.all([messagesToSummarize.length > 0 ? generateSummary(messagesToSummarize, model, settings.reserveTokens, apiKey, signal, previousSummary, observer) : Promise.resolve({
			text: "No prior history.",
			usage: void 0
		}), generateTurnPrefixSummary(turnPrefixMessages, model, settings.reserveTokens, apiKey, signal, observer)]);
		addCallUsage(historyResult.usage);
		addCallUsage(turnPrefixResult.usage);
		summary = `${historyResult.text}\n\n---\n\n**Turn Context (split turn):**\n\n${turnPrefixResult.text}`;
	} else {
		const historyResult = await generateSummary(messagesToSummarize, model, settings.reserveTokens, apiKey, signal, previousSummary, observer);
		addCallUsage(historyResult.usage);
		summary = historyResult.text;
	}
	const { readFiles, modifiedFiles } = computeFileLists(fileOps);
	summary += formatFileOperations(readFiles, modifiedFiles);
	return {
		summary,
		firstKeptIndex,
		tokensBefore,
		details: {
			readFiles,
			modifiedFiles
		},
		usage: aggregateUsage
	};
}
//#endregion
//#region src/context.ts
/**
* Context discovery: reads AGENTS.md and .agents/skills/ from a session's
* working directory. Used at runtime by the session initialisation path.
*/
function isWorkspaceSkill(skill) {
	const candidate = skill;
	return candidate.__flueWorkspaceSkill === true && typeof candidate.directory === "string" && typeof candidate.skillMdPath === "string";
}
/** Read AGENTS.md (and CLAUDE.md if present) from a directory. Returns concatenated contents. */
async function readAgentsMd(env, basePath) {
	const parts = [];
	for (const filename of ["AGENTS.md", "CLAUDE.md"]) {
		const filePath = basePath.endsWith("/") ? basePath + filename : `${basePath}/${filename}`;
		if (await env.exists(filePath)) {
			const content = await env.readFile(filePath);
			parts.push(content.trim());
		}
	}
	return parts.join("\n\n");
}
/** Path to the skills directory under a given base path. */
function skillsDirIn(basePath) {
	return basePath.endsWith("/") ? `${basePath}.agents/skills` : `${basePath}/.agents/skills`;
}
/**
* Discover skills from `.agents/skills/<name>/SKILL.md` under basePath.
*
* Skill bodies are intentionally not retained. Autonomous activation
* rereads SKILL.md before injecting its instructions, while direct name
* invocation lets the model read workspace files itself. This keeps
* relative references resolvable and picks up mid-session edits without
* re-initialising the agent. We parse the frontmatter here only to
* populate the system-prompt's "Available Skills" registry.
*/
async function discoverLocalSkills(env, basePath) {
	const skillsDir = skillsDirIn(basePath);
	if (!await env.exists(skillsDir)) return {};
	const skills = Object.create(null);
	const entries = await env.readdir(skillsDir);
	for (const entry of entries) {
		const skillDir = `${skillsDir}/${entry}`;
		try {
			if (!(await env.stat(skillDir)).isDirectory) continue;
		} catch {
			continue;
		}
		const skillMdPath = `${skillDir}/SKILL.md`;
		if (!await env.exists(skillMdPath)) continue;
		const parsed = parseSkillMarkdown(await env.readFile(skillMdPath), {
			directoryName: entry,
			path: skillMdPath
		});
		const workspaceSkill = {
			__flueWorkspaceSkill: true,
			name: parsed.name,
			description: parsed.description,
			directory: skillDir,
			skillMdPath
		};
		skills[parsed.name] = workspaceSkill;
	}
	return skills;
}
function mergeSkillCatalog(definitionSkills, discoveredSkills) {
	const merged = Object.create(null);
	for (const skill of definitionSkills) merged[skill.name] = skill;
	for (const [name, skill] of Object.entries(discoveredSkills)) {
		if (Object.hasOwn(merged, name)) throw new Error(`[flue] Skill name "${name}" appears in both agent definition and workspace discovery.`);
		merged[name] = skill;
	}
	return merged;
}
/**
* Headless-mode preamble. Included once at the top of every session's
* system prompt so the model knows it's running without a human operator
* before the first turn — and doesn't get reminded of it on every
* `prompt()` / `skill()` call. Previously this lived in
* `result.ts:buildPromptText` / `buildSkillPrompt` and was inlined into
* each per-call user message; that was redundant noise once the harness
* gained tool-call shape (it can't ask questions or wait for input
* regardless of what the user message says).
*/
const HEADLESS_PREAMBLE = "You are running in headless mode with no human operator. Work autonomously — never ask questions, never wait for user input. Make your best judgment and proceed independently.";
function composeSystemPrompt(agentsMd, skills, env, instructions) {
	const parts = [HEADLESS_PREAMBLE];
	if (instructions) parts.push("", instructions);
	if (agentsMd) parts.push("", agentsMd);
	const skillEntries = Object.values(skills);
	if (skillEntries.length > 0) {
		parts.push("", "## Available Skills", "", "The following skills provide specialized instructions for specific tasks. When a task matches a skill description, call the `activate_skill` tool with that skill name before proceeding so its full instructions are loaded. Skill instructions and supporting resources stay lazy until activation or explicit file reads.", "");
		for (const skill of skillEntries) {
			const desc = skill.description ? ` — ${skill.description}` : "";
			parts.push(`- **${skill.name}**${desc}`);
		}
	}
	if (env) {
		const date = (/* @__PURE__ */ new Date()).toLocaleDateString("en-US", {
			weekday: "short",
			year: "numeric",
			month: "short",
			day: "numeric"
		});
		parts.push("", `Date: ${date}`);
		parts.push(`Working directory: ${env.cwd}`);
		if (env.directoryListing && env.directoryListing.length > 0) parts.push("", "Directory structure:", env.directoryListing.join("\n"));
	}
	return parts.join("\n");
}
/** Discover AGENTS.md, local skills, and directory listing from the session's cwd. */
async function discoverSessionContext(env, instructions, definitionSkills = []) {
	const cwd = env.cwd;
	const agentsMd = await readAgentsMd(env, cwd);
	const skills = mergeSkillCatalog(definitionSkills, await discoverLocalSkills(env, cwd));
	let directoryListing;
	try {
		directoryListing = await env.readdir(cwd);
	} catch {}
	return {
		systemPrompt: composeSystemPrompt(agentsMd, skills, {
			cwd,
			directoryListing
		}, instructions),
		skills
	};
}
//#endregion
//#region src/session.ts
const MAX_TASK_DEPTH = 4;
function toTurnMessage(message) {
	if (message.role === "user") return {
		role: "user",
		content: typeof message.content === "string" ? message.content : message.content.map(toTurnContent)
	};
	if (message.role === "assistant") return {
		role: "assistant",
		content: message.content.map(toTurnContent)
	};
	return {
		role: "toolResult",
		toolCallId: message.toolCallId,
		toolName: message.toolName,
		content: message.content.map(toTurnContent),
		isError: message.isError
	};
}
function toTurnContent(block) {
	if (block.type === "text") return {
		type: "text",
		text: block.text,
		textSignature: block.textSignature
	};
	if (block.type === "image") return {
		type: "image",
		data: block.data,
		mimeType: block.mimeType
	};
	if (block.type === "thinking") return {
		type: "thinking",
		thinking: block.thinking,
		thinkingSignature: block.thinkingSignature,
		redacted: block.redacted
	};
	return {
		type: "toolCall",
		id: block.id,
		name: block.name,
		arguments: block.arguments,
		thoughtSignature: block.thoughtSignature
	};
}
function getRegisteredPackagedSkills(skills, packagedSkills) {
	const registered = {};
	for (const skill of Object.values(skills)) {
		if (!("__flueSkillReference" in skill)) continue;
		const packaged = packagedSkills?.[skill.id];
		if (packaged) registered[skill.id] = packaged;
	}
	return registered;
}
/** In-memory session store. Sessions persist for the lifetime of the process. */
var InMemorySessionStore = class {
	store = /* @__PURE__ */ new Map();
	async save(id, data) {
		this.store.set(id, data);
	}
	async load(id) {
		return this.store.get(id) ?? null;
	}
	async delete(id) {
		this.store.delete(id);
	}
};
var SessionHistory = class SessionHistory {
	entries;
	byId;
	leafId;
	constructor(entries, leafId) {
		this.entries = [...entries];
		this.leafId = leafId;
		this.byId = new Map(this.entries.map((entry) => [entry.id, entry]));
	}
	static empty() {
		return new SessionHistory([], null);
	}
	static fromData(data) {
		if (!data) return SessionHistory.empty();
		if (data.version !== 4) throw new Error(`[flue] Session data version ${String(data.version)} is unsupported. Clear persisted session state created by an earlier Flue beta.`);
		return new SessionHistory(data.entries, data.leafId);
	}
	getLeafId() {
		return this.leafId;
	}
	getActivePath() {
		const path = [];
		let current = this.leafId ? this.byId.get(this.leafId) : void 0;
		while (current) {
			path.push(current);
			current = current.parentId ? this.byId.get(current.parentId) : void 0;
		}
		return path.reverse();
	}
	/**
	* Active-path entries appended after `afterLeafId` (exclusive), in order.
	*
	* - `afterLeafId === null` means "from the start of the path" → returns
	*   the entire active path.
	* - When the id is found, returns entries strictly after it.
	* - When the id is *not* on the current active path (e.g. a branch
	*   switch happened mid-window), returns `[]`. Callers use this for
	*   bounded windowing — falling back to the full path would silently
	*   include unrelated history. An empty result is the safer answer
	*   for usage aggregation: zero is loud (sums won't match expectations)
	*   while full-history is silent overcounting.
	*/
	getActivePathSince(afterLeafId) {
		const path = this.getActivePath();
		if (afterLeafId === null) return path;
		const startIndex = path.findIndex((entry) => entry.id === afterLeafId);
		if (startIndex === -1) return [];
		return path.slice(startIndex + 1);
	}
	buildContextEntries() {
		const path = this.getActivePath();
		const latestCompactionIndex = findLatestCompactionIndex(path);
		if (latestCompactionIndex === -1) return pathToContextEntries(path);
		const compaction = path[latestCompactionIndex];
		const firstKeptIndex = path.findIndex((entry) => entry.id === compaction.firstKeptEntryId);
		const keptStart = firstKeptIndex >= 0 ? firstKeptIndex : latestCompactionIndex + 1;
		const context = [{
			message: createContextSummaryMessage(compaction.summary, compaction.timestamp),
			entry: compaction
		}];
		context.push(...pathToContextEntries(path.slice(keptStart, latestCompactionIndex)));
		context.push(...pathToContextEntries(path.slice(latestCompactionIndex + 1)));
		return context;
	}
	buildContext() {
		return this.buildContextEntries().map((entry) => entry.message);
	}
	getLatestCompaction() {
		const path = this.getActivePath();
		for (let i = path.length - 1; i >= 0; i--) {
			const entry = path[i];
			if (entry?.type === "compaction") return entry;
		}
	}
	appendMessage(message, source, dispatch) {
		const entry = {
			type: "message",
			id: generateEntryId(this.byId),
			parentId: this.leafId,
			timestamp: (/* @__PURE__ */ new Date()).toISOString(),
			message,
			source
		};
		if (dispatch) entry.dispatch = dispatch;
		this.appendEntry(entry);
		return entry.id;
	}
	findDispatchInput(dispatchId) {
		return this.getActivePath().find((entry) => entry.type === "message" && entry.dispatch?.dispatchId === dispatchId);
	}
	appendMessages(messages, source, dispatch) {
		let dispatchAttached = false;
		return messages.map((message) => {
			const messageDispatch = dispatch && !dispatchAttached && message.role === "user" ? dispatch : void 0;
			if (messageDispatch) dispatchAttached = true;
			return this.appendMessage(message, source, messageDispatch);
		});
	}
	appendCompaction(input) {
		if (!this.byId.has(input.firstKeptEntryId)) throw new Error(`[flue] Cannot compact: entry "${input.firstKeptEntryId}" does not exist.`);
		const entry = {
			type: "compaction",
			id: generateEntryId(this.byId),
			parentId: this.leafId,
			timestamp: (/* @__PURE__ */ new Date()).toISOString(),
			summary: input.summary,
			firstKeptEntryId: input.firstKeptEntryId,
			tokensBefore: input.tokensBefore,
			details: input.details,
			usage: input.usage
		};
		this.appendEntry(entry);
		return entry.id;
	}
	appendBranchSummary(summary, fromId, details) {
		const entry = {
			type: "branch_summary",
			id: generateEntryId(this.byId),
			parentId: this.leafId,
			timestamp: (/* @__PURE__ */ new Date()).toISOString(),
			fromId,
			summary,
			details
		};
		this.appendEntry(entry);
		return entry.id;
	}
	removeLeafMessage(message) {
		if (!this.leafId) return false;
		const leaf = this.byId.get(this.leafId);
		if (!leaf || leaf.type !== "message" || leaf.message !== message) return false;
		this.byId.delete(leaf.id);
		this.entries = this.entries.filter((entry) => entry.id !== leaf.id);
		this.leafId = leaf.parentId;
		return true;
	}
	toData(metadata, createdAt, updatedAt) {
		return {
			version: 4,
			entries: [...this.entries],
			leafId: this.leafId,
			metadata,
			createdAt,
			updatedAt
		};
	}
	appendEntry(entry) {
		this.entries.push(entry);
		this.byId.set(entry.id, entry);
		this.leafId = entry.id;
	}
};
function pathToContextEntries(path) {
	const context = [];
	for (const entry of path) if (entry.type === "message") context.push({
		message: entry.message,
		entry
	});
	else if (entry.type === "branch_summary") context.push({
		message: createUserContextMessage(`[Branch Summary]\n\n${entry.summary}`, entry.timestamp),
		entry
	});
	return context;
}
function findLatestCompactionIndex(path) {
	for (let i = path.length - 1; i >= 0; i--) if (path[i]?.type === "compaction") return i;
	return -1;
}
function createContextSummaryMessage(summary, timestamp) {
	return createUserContextMessage(summary.startsWith("[Context Summary]") ? summary : `[Context Summary]\n\n${summary}`, timestamp);
}
function createUserContextMessage(text, timestamp) {
	return {
		role: "user",
		content: [{
			type: "text",
			text
		}],
		timestamp: new Date(timestamp).getTime()
	};
}
function renderDispatchInput(input) {
	return [
		"[Dispatch Input]",
		`agent: ${input.agent}`,
		`id: ${input.id}`,
		`session: ${input.session}`,
		`dispatchId: ${input.dispatchId}`,
		`acceptedAt: ${input.acceptedAt}`,
		"",
		"input:",
		stableStringify(input.input)
	].join("\n");
}
function dispatchMetadata(input) {
	return {
		dispatchId: input.dispatchId,
		agent: input.agent,
		id: input.id,
		session: input.session,
		acceptedAt: input.acceptedAt,
		input: input.input
	};
}
function stableStringify(value) {
	return JSON.stringify(sortJsonLike(value), null, 2);
}
function sortJsonLike(value) {
	if (Array.isArray(value)) return value.map(sortJsonLike);
	if (!value || typeof value !== "object") return value;
	const sorted = {};
	for (const key of Object.keys(value).sort()) sorted[key] = sortJsonLike(value[key]);
	return sorted;
}
function generateEntryId(byId) {
	for (let i = 0; i < 100; i++) {
		const id = crypto.randomUUID().slice(0, 8);
		if (!byId.has(id)) return id;
	}
	return crypto.randomUUID();
}
var Session = class {
	name;
	fs;
	metadata;
	harness;
	storageKey;
	config;
	env;
	store;
	history;
	createdAt;
	overflowRecoveryAttempted = false;
	compactionAbortController;
	eventCallback;
	agentTools;
	toolFactory;
	deleted = false;
	deletionPromise;
	activeOperation;
	activeOperationId;
	toolStartTimes = /* @__PURE__ */ new Map();
	turnStartTime;
	activeTurnId;
	activeTasks = /* @__PURE__ */ new Set();
	taskDepth;
	createTaskSession;
	onDelete;
	emitTurnRequestAndStream = (model, context, options) => {
		if (this.activeTurnId === void 0) this.activeTurnId = generateTurnId();
		const turnId = this.activeTurnId;
		this.emitTurnRequest(turnId, "agent", model, context, options?.reasoning);
		return streamSimple(model, context, options);
	};
	emitTurnRequest(turnId, purpose, model, context, reasoning) {
		const tools = context.tools?.map((tool) => ({
			name: tool.name,
			description: tool.description,
			parameters: tool.parameters
		}));
		this.emit({
			type: "turn_request",
			turnId,
			purpose,
			model: model.id,
			provider: model.provider,
			api: model.api,
			input: {
				systemPrompt: context.systemPrompt,
				messages: context.messages.map(toTurnMessage),
				tools
			},
			reasoning
		});
	}
	constructor(options) {
		this.name = options.name;
		this.storageKey = options.storageKey;
		this.config = options.config;
		this.env = options.env;
		this.fs = createFlueFs(options.env);
		this.store = options.store;
		this.agentTools = options.agentTools ?? [];
		this.toolFactory = options.toolFactory;
		this.taskDepth = options.taskDepth ?? 0;
		this.createTaskSession = options.createTaskSession;
		this.onDelete = options.onDelete;
		this.metadata = options.existingData?.metadata ?? {};
		this.createdAt = options.existingData?.createdAt;
		this.history = SessionHistory.fromData(options.existingData);
		const systemPrompt = this.config.systemPrompt;
		const builtinTools = this.createBuiltinTools(this.env, []);
		const tools = [...builtinTools, ...this.createCustomTools(this.agentTools, builtinTools)];
		const previousMessages = this.history.buildContext();
		this.harness = new Agent({
			initialState: {
				systemPrompt,
				model: this.config.model,
				tools,
				messages: previousMessages,
				thinkingLevel: this.config.thinkingLevel ?? "medium"
			},
			getApiKey: (provider) => this.getProviderApiKey(provider),
			onPayload: (payload, model) => this.applyProviderPayloadOverrides(payload, model),
			streamFn: this.emitTurnRequestAndStream,
			toolExecution: "parallel",
			sessionId: options.affinityKey
		});
		this.eventCallback = options.onAgentEvent;
		this.harness.subscribe(async (event) => {
			switch (event.type) {
				case "agent_start":
					this.emit({ type: "agent_start" });
					break;
				case "turn_start":
					this.turnStartTime = Date.now();
					this.activeTurnId = generateTurnId();
					this.emit({
						type: "turn_start",
						turnId: this.activeTurnId,
						purpose: "agent"
					});
					break;
				case "message_start":
					this.emit({
						type: "message_start",
						message: event.message
					});
					break;
				case "message_update": {
					this.emit({
						type: "message_update",
						message: event.message,
						assistantMessageEvent: event.assistantMessageEvent
					});
					const aEvent = event.assistantMessageEvent;
					if (aEvent.type === "text_delta") this.emit({
						type: "text_delta",
						text: aEvent.delta
					});
					else if (aEvent.type === "thinking_start") this.emit({ type: "thinking_start" });
					else if (aEvent.type === "thinking_delta") this.emit({
						type: "thinking_delta",
						delta: aEvent.delta
					});
					else if (aEvent.type === "thinking_end") this.emit({
						type: "thinking_end",
						content: aEvent.content
					});
					break;
				}
				case "message_end":
					this.emit({
						type: "message_end",
						message: event.message
					});
					break;
				case "tool_execution_start":
					this.toolStartTimes.set(event.toolCallId, Date.now());
					this.emit({
						type: "tool_execution_start",
						toolName: event.toolName,
						toolCallId: event.toolCallId,
						args: event.args
					});
					this.emit({
						type: "tool_start",
						toolName: event.toolName,
						toolCallId: event.toolCallId,
						args: event.args
					});
					break;
				case "tool_execution_update":
					this.emit({
						type: "tool_execution_update",
						toolName: event.toolName,
						toolCallId: event.toolCallId,
						args: event.args,
						partialResult: event.partialResult
					});
					break;
				case "tool_execution_end":
					this.emit({
						type: "tool_execution_end",
						toolName: event.toolName,
						toolCallId: event.toolCallId,
						result: event.result,
						isError: event.isError
					});
					this.emit({
						type: "tool_call",
						toolName: event.toolName,
						toolCallId: event.toolCallId,
						isError: event.isError,
						result: event.result,
						durationMs: durationSince(this.toolStartTimes.get(event.toolCallId))
					});
					this.toolStartTimes.delete(event.toolCallId);
					break;
				case "turn_end": {
					const turnId = this.activeTurnId ?? generateTurnId();
					this.emit({
						type: "turn_end",
						turnId,
						purpose: "agent",
						message: event.message,
						toolResults: event.toolResults
					});
					const message = event.message;
					const assistant = message.role === "assistant" ? message : void 0;
					const output = assistant ? toTurnMessage(assistant) : void 0;
					const model = this.harness.state.model;
					this.emit({
						type: "turn",
						turnId,
						purpose: "agent",
						durationMs: durationSince(this.turnStartTime),
						model: model?.id,
						provider: model?.provider,
						api: model?.api,
						output,
						usage: fromProviderUsage(assistant?.usage),
						stopReason: assistant?.stopReason,
						isError: assistant?.stopReason === "error" || assistant?.stopReason === "aborted",
						error: assistant?.errorMessage
					});
					this.turnStartTime = void 0;
					this.activeTurnId = void 0;
					break;
				}
				case "agent_end":
					this.emit({
						type: "agent_end",
						messages: event.messages
					});
					this.turnStartTime = void 0;
					this.activeTurnId = void 0;
					break;
			}
		});
	}
	resolveCompactionSettings(model) {
		const cc = this.config.compaction;
		const defaults = model ? deriveCompactionDefaults({
			contextWindow: model.contextWindow ?? 0,
			maxTokens: model.maxTokens ?? 0
		}) : DEFAULT_COMPACTION_SETTINGS;
		if (cc === false) return {
			...defaults,
			enabled: false
		};
		if (!cc) return defaults;
		return {
			enabled: true,
			reserveTokens: cc.reserveTokens ?? defaults.reserveTokens,
			keepRecentTokens: cc.keepRecentTokens ?? defaults.keepRecentTokens
		};
	}
	prompt(text, options) {
		return createCallHandle(options?.signal, (signal) => this.runOperation("prompt", signal, async () => {
			const schema = options?.result;
			return this.runPromptCall({
				promptText: buildPromptText(text, schema),
				schema,
				tools: options?.tools,
				model: options?.model,
				thinkingLevel: options?.thinkingLevel,
				images: options?.images,
				source: "prompt",
				errorLabel: "prompt",
				callSite: "this prompt() call",
				signal
			});
		}));
	}
	processDirectInput(input) {
		return createCallHandle(void 0, (signal) => this.runOperation("prompt", signal, () => this.runPromptCall({
			promptText: input.message,
			schema: void 0,
			tools: void 0,
			model: void 0,
			thinkingLevel: void 0,
			images: void 0,
			source: "prompt",
			errorLabel: "prompt",
			callSite: "this direct input",
			signal
		})));
	}
	processDispatchInput(input) {
		return createCallHandle(void 0, (signal) => this.runOperation("prompt", signal, () => this.runPersistedDispatchInput(input)));
	}
	skill(skill, options) {
		return createCallHandle(options?.signal, (signal) => this.runOperation("skill", signal, async () => {
			const schema = options?.result;
			let promptText;
			let skillName;
			let activePackagedSkills;
			if (typeof skill === "string") {
				const registered = this.config.skills[skill];
				if (registered && "__flueSkillReference" in registered) {
					const packaged = this.resolvePackagedSkill(registered);
					promptText = buildPackagedSkillPrompt(registered, packaged, options?.args, schema);
					activePackagedSkills = { [registered.id]: packaged };
				} else if (registered) promptText = buildSkillByPathlessNamePrompt(skill, options?.args, schema);
				else this.throwMissingSkill(skill);
				skillName = skill;
			} else {
				const packaged = this.resolvePackagedSkill(skill);
				promptText = buildPackagedSkillPrompt(skill, packaged, options?.args, schema);
				activePackagedSkills = { [skill.id]: packaged };
				skillName = skill.name;
			}
			return this.runPromptCall({
				promptText,
				schema,
				tools: options?.tools,
				model: options?.model,
				thinkingLevel: options?.thinkingLevel,
				images: options?.images,
				source: "skill",
				errorLabel: `skill("${skillName}")`,
				callSite: `this skill("${skillName}") call`,
				activePackagedSkills,
				signal
			});
		}));
	}
	task(text, options) {
		return createCallHandle(options?.signal, async (signal) => {
			return (await this.runTask(text, options, signal)).output;
		});
	}
	shell(command, options) {
		return createCallHandle(options?.signal, (signal) => this.runOperation("shell", signal, async () => {
			const toolCallId = crypto.randomUUID();
			const toolStartMs = Date.now();
			const args = { command };
			if (options?.cwd !== void 0) args.cwd = options.cwd;
			if (options?.env !== void 0) args.env = redactEnvValues(options.env);
			this.emit({
				type: "tool_start",
				toolName: "bash",
				toolCallId,
				args
			});
			try {
				const result = await this.env.exec(command, {
					env: options?.env,
					cwd: options?.cwd,
					signal
				});
				const shellResult = {
					stdout: result.stdout,
					stderr: result.stderr,
					exitCode: result.exitCode
				};
				const toolResult = formatBashResult(shellResult, command);
				await this.appendShellTriple(toolCallId, args, toolResult, false);
				this.emit({
					type: "tool_call",
					toolName: "bash",
					toolCallId,
					isError: false,
					result: toolResult,
					durationMs: durationSince(toolStartMs)
				});
				return shellResult;
			} catch (error) {
				const errResult = {
					content: [{
						type: "text",
						text: getErrorMessage(error)
					}],
					details: {
						command,
						exitCode: -1
					}
				};
				await this.appendShellTriple(toolCallId, args, errResult, true);
				this.emit({
					type: "tool_call",
					toolName: "bash",
					toolCallId,
					isError: true,
					result: errResult,
					durationMs: durationSince(toolStartMs)
				});
				throw error;
			}
		}));
	}
	async compact() {
		await this.runOperation("compact", void 0, async () => {
			await this.runCompaction("manual", false);
		});
	}
	abort() {
		this.harness.abort();
		this.compactionAbortController?.abort();
		for (const task of this.activeTasks) task.abort();
	}
	close() {
		if (this.deleted) return;
		this.deleted = true;
		this.abort();
		this.onDelete?.();
	}
	delete() {
		if (this.deletionPromise) return this.deletionPromise;
		if (this.deleted) return Promise.resolve();
		if (this.activeOperation) return Promise.reject(/* @__PURE__ */ new Error(`[flue] Session "${this.name}" cannot be deleted while ${this.activeOperation} is running. Wait for the active operation to finish before deleting the session.`));
		this.deleted = true;
		this.deletionPromise = Promise.resolve().then(() => deleteSessionTree(this.store, this.storageKey)).then(() => {
			this.onDelete?.();
		});
		return this.deletionPromise;
	}
	/** Precedence: call-level > agent-level default. */
	resolveModelForCall(modelSpecifier, callSite) {
		const model = modelSpecifier ? this.config.resolveModel(modelSpecifier) : this.config.model;
		return this.requireModel(model, callSite);
	}
	/** Precedence: call-level > agent-level default > 'medium'. */
	resolveThinkingLevelForCall(callValue) {
		return callValue ?? this.config.thinkingLevel ?? "medium";
	}
	/**
	* Throws a clear, actionable error when no model is configured for a call.
	* Use with the resolved model (post-precedence) to guarantee we never hand
	* `undefined` to the underlying agent.
	*/
	requireModel(model, callSite) {
		if (model) return model;
		throw new Error(`[flue] No model configured for ${callSite}. Pass \`{ model: "provider-id/model-id" }\` to this call or configure an agent model.`);
	}
	getProviderApiKey(providerId) {
		const override = getProviderConfiguration(providerId)?.apiKey;
		if (override !== void 0) return override;
		return getRegisteredApiKey(providerId);
	}
	/**
	* Provider-specific payload overrides. Returning undefined keeps the
	* upstream-built payload as-is.
	*/
	applyProviderPayloadOverrides(payload, model) {
		if (model.api !== "openai-responses" && model.api !== "azure-openai-responses") return;
		if (getProviderConfiguration(model.provider)?.storeResponses !== true) return;
		return {
			...payload,
			store: true
		};
	}
	resolvePackagedSkill(reference) {
		const packaged = this.config.packagedSkills?.[reference.id];
		if (!packaged) throw new Error(`[flue] Packaged skill "${reference.name}" is unavailable for this application build.`);
		return packaged;
	}
	async activateSkillForTool(name) {
		const registered = this.config.skills[name];
		if (!registered) this.throwMissingSkill(name);
		if ("__flueSkillReference" in registered) return buildPackagedSkillPrompt(registered, this.resolvePackagedSkill(registered));
		if (isWorkspaceSkill(registered)) return buildWorkspaceSkillPrompt(registered.name, registered.directory, registered.skillMdPath, await this.env.readFile(registered.skillMdPath));
		return buildSkillByPathlessNamePrompt(name);
	}
	throwMissingSkill(skill) {
		const available = Object.keys(this.config.skills).join(", ") || "(none)";
		throw new Error(`[flue] Skill "${skill}" not registered. Available: ${available}.\n\nSkills are discovered at init() time from ${skillsDirIn(this.env.cwd)}/<name>/SKILL.md inside the session's sandbox. If you expected "${skill}" to be there, make sure the SKILL.md file exists at that path before calling init() — the default empty sandbox starts with no files, so it has no skills unless you put them there.\n\nPackaged skills can be imported from SKILL.md with { type: 'skill' } and passed directly to session.skill(skillReference).`);
	}
	createCustomTools(tools, builtinTools) {
		this.validateCustomToolNames(tools, builtinTools);
		return tools.map((toolDef) => ({
			name: toolDef.name,
			label: toolDef.name,
			description: toolDef.description,
			parameters: toolDef.parameters,
			async execute(_toolCallId, params, signal) {
				if (signal?.aborted) throw new Error("Operation aborted");
				return {
					content: [{
						type: "text",
						text: await toolDef.execute(params, signal)
					}],
					details: { customTool: toolDef.name }
				};
			}
		}));
	}
	/** Reject custom tools that collide with active built-ins or each other. */
	validateCustomToolNames(tools, builtinTools) {
		const reserved = new Set(builtinTools.map((t) => t.name));
		reserved.add("task");
		reserved.add("activate_skill");
		const names = /* @__PURE__ */ new Set();
		for (const toolDef of tools) {
			if (reserved.has(toolDef.name)) throw new Error(`[flue] Custom tool "${toolDef.name}" conflicts with a built-in tool. Built-in tools: ${[...reserved].join(", ")}`);
			if (names.has(toolDef.name)) throw new Error(`[flue] Duplicate custom tool name "${toolDef.name}". Tool names must be unique.`);
			names.add(toolDef.name);
		}
	}
	/** Build built-in tools from the connector or the framework defaults. */
	createBuiltinTools(env, tools, model, thinkingLevel, activePackagedSkills) {
		const runTask = (params, signal) => this.runTaskForTool(params, tools, model, thinkingLevel, signal);
		const packagedSkills = {
			...getRegisteredPackagedSkills(this.config.skills, this.config.packagedSkills),
			...activePackagedSkills
		};
		const skillNames = Object.keys(this.config.skills);
		const activateSkillTool = skillNames.length > 0 ? createActivateSkillTool(skillNames, (name) => this.activateSkillForTool(name)) : void 0;
		const appendActivateSkillTool = (builtinTools) => activateSkillTool ? [...builtinTools, activateSkillTool] : builtinTools;
		if (this.toolFactory) {
			let connectorTools = this.toolFactory(env, { subagents: this.config.subagents ?? {} });
			if (Object.keys(packagedSkills).length > 0) {
				const packagedRead = createPackagedSkillReadTool(packagedSkills);
				const connectorRead = connectorTools.find((tool) => tool.name === "read");
				if (connectorRead) connectorTools = connectorTools.map((tool) => tool !== connectorRead ? tool : {
					...tool,
					execute: (id, params, signal) => {
						const resourcePath = typeof params === "object" && params !== null && "path" in params ? params.path : void 0;
						return typeof resourcePath === "string" && resourcePath.startsWith("/.flue/packaged-skills/") ? packagedRead.execute(id, params, signal) : connectorRead.execute(id, params, signal);
					}
				});
				else connectorTools = [...connectorTools, packagedRead];
			}
			this.validateConnectorTools(connectorTools);
			return appendActivateSkillTool([...connectorTools, createTaskTool(runTask, this.config.subagents ?? {})]);
		}
		return appendActivateSkillTool(createTools(env, {
			subagents: this.config.subagents ?? {},
			packagedSkills,
			task: runTask
		}));
	}
	/** Validate connector tool names before handing them to the agent loop. */
	validateConnectorTools(tools) {
		const names = /* @__PURE__ */ new Set();
		for (const tool of tools) {
			if (tool.name === "task" || tool.name === "activate_skill") throw new Error(`[flue] Sandbox connector tools() returned a tool named "${tool.name}", which is framework-reserved. The framework appends \`${tool.name}\` automatically when appropriate; remove it from the connector.`);
			if (names.has(tool.name)) throw new Error(`[flue] Sandbox connector tools() returned duplicate tool name "${tool.name}". Connector tool names must be unique.`);
			names.add(tool.name);
		}
	}
	async withCallOverrides(options, fn) {
		const previousTools = this.harness.state.tools;
		const previousModel = this.harness.state.model;
		const previousSystemPrompt = this.harness.state.systemPrompt;
		const previousThinkingLevel = this.harness.state.thinkingLevel;
		const resolvedModel = this.resolveModelForCall(options.model, options.callSite);
		this.harness.state.model = resolvedModel;
		this.harness.state.thinkingLevel = this.resolveThinkingLevelForCall(options.thinkingLevel);
		const builtinTools = this.createBuiltinTools(this.env, options.tools, options.model, options.thinkingLevel, options.activePackagedSkills);
		const customTools = this.createCustomTools([...this.agentTools, ...options.tools], builtinTools);
		this.harness.state.tools = [
			...builtinTools,
			...customTools,
			...options.extraTools ?? []
		];
		try {
			return await fn({ resolvedModel });
		} finally {
			this.harness.state.tools = previousTools;
			this.harness.state.model = previousModel;
			this.harness.state.systemPrompt = previousSystemPrompt;
			this.harness.state.thinkingLevel = previousThinkingLevel;
		}
	}
	resolveDeclaredSubagent(name) {
		const subagents = this.config.subagents ?? {};
		const subagent = subagents[name];
		if (subagent) return subagent;
		const available = Object.keys(subagents).join(", ") || "(none)";
		throw new Error(`[flue] Subagent "${name}" is not declared. Available: ${available}.`);
	}
	async runTaskForTool(params, tools, inheritedModel, inheritedThinkingLevel, signal) {
		const result = await this.executeTask(params.prompt, {
			agent: params.agent,
			inheritedModel,
			inheritedThinkingLevel,
			cwd: params.cwd,
			tools
		}, signal);
		return {
			content: [{
				type: "text",
				text: result.text || "(task completed with no text)"
			}],
			details: {
				taskId: result.taskId,
				session: result.session,
				messageId: result.messageId,
				agent: result.agent,
				cwd: result.cwd
			}
		};
	}
	async runTask(text, options, signal) {
		return this.runExclusive("task", async () => {
			if (signal?.aborted) throw abortErrorFor(signal);
			this.activeOperationId = generateOperationId();
			const operationId = this.activeOperationId;
			const startedAt = Date.now();
			this.emit({
				type: "operation_start",
				operationId,
				operationKind: "task"
			});
			try {
				const result = await this.executeTask(text, options, signal);
				this.emit({
					type: "operation",
					operationId,
					operationKind: "task",
					durationMs: durationSince(startedAt),
					isError: false,
					result: result.output,
					usage: usageFromResult(result.output)
				});
				return result;
			} catch (error) {
				const surfaced = signal?.aborted ? abortErrorFor(signal) : error;
				this.emit({
					type: "operation",
					operationId,
					operationKind: "task",
					durationMs: durationSince(startedAt),
					isError: true,
					error: serializeError(surfaced)
				});
				throw surfaced;
			} finally {
				this.emit({ type: "idle" });
				this.activeOperationId = void 0;
			}
		});
	}
	async executeTask(text, options, signal) {
		this.assertActive();
		if (!this.createTaskSession) throw new Error("[flue] This session cannot create task sessions.");
		if (this.taskDepth >= MAX_TASK_DEPTH) throw new Error(`[flue] Maximum task depth (${MAX_TASK_DEPTH}) exceeded.`);
		if (signal?.aborted) throw abortErrorFor(signal);
		const taskId = crypto.randomUUID();
		const taskAgent = options?.agent ? this.resolveDeclaredSubagent(options.agent) : void 0;
		let child;
		let abortListener;
		this.emit({
			type: "task_start",
			taskId,
			prompt: text,
			agent: taskAgent?.name,
			cwd: options?.cwd,
			parentSession: this.name
		});
		const taskStartMs = Date.now();
		try {
			child = await this.createTaskSession({
				parentSession: this.name,
				taskId,
				parentEnv: this.env,
				cwd: options?.cwd,
				agent: taskAgent,
				depth: this.taskDepth + 1
			});
			await this.recordTaskSession(child.name, child.storageKey, taskId);
			this.activeTasks.add(child);
			if (signal) {
				abortListener = () => child?.abort();
				signal.addEventListener("abort", abortListener, { once: true });
			}
			const schema = options?.result;
			const childOptions = {
				model: options?.model ?? (taskAgent?.model !== void 0 ? void 0 : options?.inheritedModel),
				thinkingLevel: options?.thinkingLevel ?? (taskAgent?.thinkingLevel !== void 0 ? void 0 : options?.inheritedThinkingLevel),
				tools: options?.tools,
				images: options?.images,
				signal
			};
			if (schema) childOptions.result = schema;
			const output = await child.prompt(text, childOptions);
			const taskResult = {
				output,
				text: typeof output?.text === "string" ? output.text : child.getAssistantText(),
				taskId,
				session: child.name,
				messageId: child.getLatestAssistantMessageId(),
				agent: taskAgent?.name,
				cwd: options?.cwd
			};
			this.emit({
				type: "task",
				taskId,
				agent: taskAgent?.name,
				isError: false,
				result: taskResult.text,
				durationMs: durationSince(taskStartMs),
				parentSession: this.name
			});
			return taskResult;
		} catch (error) {
			this.emit({
				type: "task",
				taskId,
				agent: taskAgent?.name,
				isError: true,
				result: getErrorMessage(error),
				durationMs: durationSince(taskStartMs),
				parentSession: this.name
			});
			throw error;
		} finally {
			if (signal && abortListener) signal.removeEventListener("abort", abortListener);
			if (child) {
				this.activeTasks.delete(child);
				child.close();
			}
		}
	}
	async runOperation(operation, signal, fn) {
		return this.runExclusive(operation, async () => {
			if (signal?.aborted) throw abortErrorFor(signal);
			this.activeOperationId = generateOperationId();
			const operationId = this.activeOperationId;
			const startedAt = Date.now();
			this.emit({
				type: "operation_start",
				operationId,
				operationKind: operation
			});
			const onAbort = () => {
				this.harness.abort();
				this.compactionAbortController?.abort(signal?.reason);
				for (const task of this.activeTasks) task.abort();
			};
			signal?.addEventListener("abort", onAbort, { once: true });
			try {
				const result = await fn();
				this.emit({
					type: "operation",
					operationId,
					operationKind: operation,
					durationMs: durationSince(startedAt),
					isError: false,
					result,
					usage: usageFromResult(result)
				});
				return result;
			} catch (error) {
				const surfaced = signal?.aborted ? abortErrorFor(signal) : error;
				this.emit({
					type: "operation",
					operationId,
					operationKind: operation,
					durationMs: durationSince(startedAt),
					isError: true,
					error: serializeError(surfaced)
				});
				throw surfaced;
			} finally {
				signal?.removeEventListener("abort", onAbort);
				this.emit({ type: "idle" });
				this.activeOperationId = void 0;
			}
		});
	}
	async runExclusive(operation, fn) {
		this.assertActive();
		if (this.activeOperation) throw new Error(`[flue] Session "${this.name}" is already running ${this.activeOperation}. Start another session for parallel conversation branches.`);
		this.activeOperation = operation;
		try {
			return await fn();
		} finally {
			this.activeOperation = void 0;
		}
	}
	emit(event) {
		const decorated = {
			...event,
			session: event.session ?? this.name
		};
		const operationId = event.operationId ?? this.activeOperationId;
		if (operationId !== void 0) decorated.operationId = operationId;
		const turnId = event.turnId ?? this.activeTurnId;
		if (turnId !== void 0) decorated.turnId = turnId;
		this.eventCallback?.(decorated);
	}
	assertActive() {
		if (this.deleted) throw new Error(`[flue] Session "${this.name}" has been deleted.`);
	}
	/** Append a `session.shell()` call as an LLM-shaped bash tool exchange. */
	async appendShellTriple(toolCallId, args, toolResult, isError) {
		const timestamp = Date.now();
		const userMessage = {
			role: "user",
			content: `Run this shell command:\n\n\`\`\`bash\n${args.command}\n\`\`\``,
			timestamp
		};
		const assistantMessage = {
			role: "assistant",
			content: [{
				type: "toolCall",
				id: toolCallId,
				name: "bash",
				arguments: args
			}],
			api: "flue-shell",
			provider: "flue",
			model: "",
			usage: {
				input: 0,
				output: 0,
				cacheRead: 0,
				cacheWrite: 0,
				totalTokens: 0,
				cost: {
					input: 0,
					output: 0,
					cacheRead: 0,
					cacheWrite: 0,
					total: 0
				}
			},
			stopReason: "toolUse",
			timestamp
		};
		const toolResultMessage = {
			role: "toolResult",
			toolCallId,
			toolName: "bash",
			content: toolResult.content,
			details: toolResult.details,
			isError,
			timestamp
		};
		this.history.appendMessages([
			userMessage,
			assistantMessage,
			toolResultMessage
		], "shell");
		this.harness.state.messages = this.history.buildContext();
		await this.save();
	}
	async syncHarnessMessagesSince(index, source, firstDispatch) {
		const messages = this.harness.state.messages.slice(index);
		if (messages.length === 0) return;
		this.history.appendMessages(messages, source, firstDispatch);
		await this.save();
	}
	async save() {
		const now = (/* @__PURE__ */ new Date()).toISOString();
		const data = this.history.toData(this.metadata, this.createdAt ?? now, now);
		if (!this.createdAt) this.createdAt = now;
		await this.store.save(this.storageKey, data);
	}
	async recordTaskSession(session, storageKey, taskId) {
		const taskSessions = Array.isArray(this.metadata.taskSessions) ? this.metadata.taskSessions : [];
		if (!taskSessions.some((task) => task?.session === session)) {
			taskSessions.push({
				session,
				taskId,
				storageKey
			});
			this.metadata.taskSessions = taskSessions;
			await this.save();
		}
	}
	async checkLatestAssistantForCompaction() {
		const messages = this.harness.state.messages;
		const lastMsg = messages[messages.length - 1];
		if (lastMsg?.role === "assistant") await this.checkCompaction(lastMsg);
	}
	async checkCompaction(assistantMessage) {
		if (assistantMessage.stopReason === "aborted") return;
		const model = this.harness.state.model;
		const settings = this.resolveCompactionSettings(model);
		const contextWindow = model.contextWindow ?? 0;
		if (isContextOverflow(assistantMessage, contextWindow)) {
			if (this.overflowRecoveryAttempted) return;
			this.overflowRecoveryAttempted = true;
			this.internalLog("info", "[flue:compaction] Overflow detected, compacting and retrying...");
			const messages = this.harness.state.messages;
			const lastMsg = messages[messages.length - 1];
			if (lastMsg && lastMsg.role === "assistant") {
				this.harness.state.messages = messages.slice(0, -1);
				this.history.removeLeafMessage(lastMsg);
				await this.save();
			}
			try {
				await this.runCompaction("overflow", true);
			} finally {
				this.overflowRecoveryAttempted = false;
			}
			return;
		}
		if (!settings.enabled) return;
		if (assistantMessage.stopReason === "error") return;
		const contextTokens = calculateContextTokens(assistantMessage.usage);
		if (shouldCompact(contextTokens, contextWindow, settings)) {
			this.internalLog("info", `[flue:compaction] Threshold reached — ${contextTokens} tokens used, window ${contextWindow}, reserve ${settings.reserveTokens}, triggering compaction`);
			await this.runCompaction("threshold", false);
		}
	}
	/**
	* Runs a compaction pass. The summarization cost (1–2 internal LLM
	* calls) is persisted on the resulting `CompactionEntry.usage`, which
	* `aggregateUsageSince` later folds into the surrounding call's
	* `response.usage` — so users see the true cost of the call that
	* triggered compaction.
	*/
	async runCompaction(reason, willRetry) {
		this.compactionAbortController = new AbortController();
		const messagesBefore = this.harness.state.messages.length;
		const compactionStartMs = Date.now();
		try {
			const sessionModel = this.harness.state.model;
			const settings = this.resolveCompactionSettings(sessionModel);
			const compactionConfig = this.config.compaction === false ? void 0 : this.config.compaction;
			const summarizationModel = compactionConfig?.model ? this.config.resolveModel(compactionConfig.model) ?? sessionModel : sessionModel;
			const contextEntries = this.history.buildContextEntries();
			const messages = contextEntries.map((entry) => entry.message);
			const latestCompaction = this.history.getLatestCompaction();
			const preparation = prepareCompaction(messages, settings, latestCompaction ? {
				summary: latestCompaction.summary,
				firstKeptIndex: 1,
				details: latestCompaction.details
			} : void 0);
			if (!preparation) {
				this.internalLog("info", "[flue:compaction] Nothing to compact (no valid cut point found)");
				return;
			}
			const firstKeptEntry = contextEntries[preparation.firstKeptIndex]?.entry;
			if (!firstKeptEntry || firstKeptEntry.type !== "message") {
				this.internalLog("info", "[flue:compaction] Nothing to compact (first kept message has no entry)");
				return;
			}
			this.internalLog("info", `[flue:compaction] Summarizing ${preparation.messagesToSummarize.length} messages` + (preparation.isSplitTurn ? ` (split turn: ${preparation.turnPrefixMessages.length} prefix messages)` : "") + `, keeping messages from index ${preparation.firstKeptIndex}`);
			const estimatedTokens = preparation.tokensBefore;
			this.emit({
				type: "compaction_start",
				reason,
				estimatedTokens
			});
			const result = await compact(preparation, summarizationModel, this.getProviderApiKey(summarizationModel.provider), this.compactionAbortController.signal, {
				start: (purpose, model, context, options) => {
					const handle = {
						turnId: generateTurnId(),
						startedAt: Date.now()
					};
					this.emitTurnRequest(handle.turnId, purpose, model, context, options.reasoning);
					return handle;
				},
				end: (purpose, handle, model, response, error) => {
					const output = response ? toTurnMessage(response) : void 0;
					this.emit({
						type: "turn",
						turnId: handle.turnId,
						purpose,
						durationMs: durationSince(handle.startedAt),
						model: model.id,
						provider: model.provider,
						api: model.api,
						output,
						usage: fromProviderUsage(response?.usage),
						stopReason: response?.stopReason,
						isError: error !== void 0 || response?.stopReason === "error" || response?.stopReason === "aborted",
						error: error === void 0 ? response?.errorMessage : serializeError(error)
					});
				}
			});
			if (this.compactionAbortController.signal.aborted) return;
			this.history.appendCompaction({
				summary: result.summary,
				firstKeptEntryId: firstKeptEntry.id,
				tokensBefore: result.tokensBefore,
				details: result.details,
				usage: result.usage
			});
			this.harness.state.messages = this.history.buildContext();
			const messagesAfter = this.harness.state.messages.length;
			this.internalLog("info", `[flue:compaction] Complete — messages: ${messagesBefore} → ${messagesAfter}, tokens before: ${result.tokensBefore}`);
			this.emit({
				type: "compaction",
				messagesBefore,
				messagesAfter,
				durationMs: durationSince(compactionStartMs),
				usage: result.usage
			});
			await this.save();
			if (willRetry) {
				const msgs = this.harness.state.messages;
				const lastMsg = msgs[msgs.length - 1];
				if (lastMsg?.role === "assistant" && lastMsg.stopReason === "error") this.harness.state.messages = msgs.slice(0, -1);
				this.internalLog("info", "[flue:compaction] Retrying after overflow recovery...");
				const beforeRetry = this.harness.state.messages.length;
				await this.harness.continue();
				await this.harness.waitForIdle();
				await this.syncHarnessMessagesSince(beforeRetry, "retry");
			}
		} catch (error) {
			const errorMessage = error instanceof Error ? error.message : String(error);
			this.internalLog("error", `[flue:compaction] Failed: ${errorMessage}`, { error });
		} finally {
			this.compactionAbortController = void 0;
		}
	}
	internalLog(level, message, attributes) {
		if (level === "error") console.error(message);
		this.emit({
			type: "log",
			level,
			message,
			attributes: normalizeLogAttributes(attributes)
		});
	}
	throwIfError(context) {
		const errorMsg = this.harness.state.errorMessage;
		if (errorMsg) throw new Error(`[flue] ${context} failed: ${errorMsg}`);
	}
	/**
	* Sum the usage of every entry the call appended to the active path
	* after `beforeLeafId`: assistant messages contribute their per-turn
	* `usage` (provider-reported, normalized through `fromProviderUsage`),
	* and compaction entries contribute the aggregated cost of the
	* summarization call(s) they dispatched. Returns zeros when nothing
	* was appended (defensive — `throwIfError` normally fires first).
	*
	* Walks the durable, parent-linked active path rather than the volatile
	* flat `harness.state.messages` array, so the result is robust to
	* mid-call mutations (e.g. overflow recovery removing a failed
	* assistant turn before retry).
	*/
	aggregateUsageSince(beforeLeafId) {
		let totals = emptyUsage();
		for (const entry of this.history.getActivePathSince(beforeLeafId)) if (entry.type === "message" && entry.message.role === "assistant") {
			const usage = fromProviderUsage(entry.message.usage);
			if (usage) totals = addUsage(totals, usage);
		} else if (entry.type === "compaction" && entry.usage) totals = addUsage(totals, entry.usage);
		return totals;
	}
	getAssistantText() {
		const messages = this.harness.state.messages;
		for (let i = messages.length - 1; i >= 0; i--) {
			const msg = messages[i];
			if (msg?.role !== "assistant") continue;
			const content = msg.content;
			if (!Array.isArray(content)) continue;
			const textParts = [];
			for (const block of content) if (block.type === "text") textParts.push(block.text);
			return textParts.join("\n");
		}
		return "";
	}
	getLatestAssistantMessageId() {
		const path = this.history.getActivePath();
		for (let i = path.length - 1; i >= 0; i--) {
			const entry = path[i];
			if (entry?.type === "message" && entry.message.role === "assistant") return entry.id;
		}
	}
	async runPersistedDispatchInput(input) {
		return this.runPersistedContextInput({
			findInput: () => this.history.findDispatchInput(input.dispatchId),
			persistInput: () => this.history.appendMessage(createUserContextMessage(renderDispatchInput(input), (/* @__PURE__ */ new Date()).toISOString()), "dispatch", dispatchMetadata(input)),
			errorLabel: `dispatch(${input.dispatchId})`,
			outputSource: "dispatch",
			callSite: "this dispatched input",
			persistenceError: "[flue] Failed to persist dispatched input.",
			recoveryError: "[flue] Cannot recover dispatched input after the session has advanced."
		});
	}
	async runPersistedContextInput(options) {
		return this.withCallOverrides({
			tools: [],
			model: void 0,
			thinkingLevel: void 0,
			callSite: options.callSite
		}, async ({ resolvedModel }) => {
			let inputEntry = options.findInput();
			if (!inputEntry) {
				options.persistInput();
				this.harness.state.messages = this.history.buildContext();
				await this.save();
				inputEntry = options.findInput();
			}
			if (!inputEntry) throw new Error(options.persistenceError);
			const following = this.history.getActivePathSince(inputEntry.id);
			if (following.some((entry) => entry.type === "message" && entry.message.role === "user")) throw new Error(options.recoveryError);
			const persistedAssistant = [...following].reverse().find((entry) => entry.type === "message" && entry.message.role === "assistant");
			if (!persistedAssistant) {
				const beforeLength = this.harness.state.messages.length;
				await this.harness.continue();
				await this.harness.waitForIdle();
				await this.syncHarnessMessagesSince(beforeLength, options.outputSource);
				await this.checkLatestAssistantForCompaction();
				this.throwIfError(options.errorLabel);
			} else {
				const assistant = persistedAssistant.message;
				if (assistant.stopReason === "error" || assistant.stopReason === "aborted") throw new Error(`[flue] ${options.errorLabel} failed: ${assistant.errorMessage ?? assistant.stopReason}`);
			}
			return {
				text: this.getAssistantText(),
				usage: this.aggregateUsageSince(inputEntry.id),
				model: {
					provider: resolvedModel.provider,
					id: resolvedModel.id
				}
			};
		});
	}
	/**
	* Shared body of `prompt()` and `skill()`: scope the runtime, optionally
	* inject the result-tool pair, drive the harness, and aggregate usage.
	*
	* Returns `PromptResultResponse<T>` when a result schema is set, else `PromptResponse`.
	*/
	async runPromptCall(args) {
		const resultBundle = args.schema ? createResultTools(args.schema) : void 0;
		return this.withCallOverrides({
			tools: args.tools ?? [],
			model: args.model,
			thinkingLevel: args.thinkingLevel,
			callSite: args.callSite,
			extraTools: resultBundle?.tools,
			activePackagedSkills: args.activePackagedSkills
		}, async ({ resolvedModel }) => {
			const beforeLength = this.harness.state.messages.length;
			const beforeLeafId = this.history.getLeafId();
			const model = {
				provider: resolvedModel.provider,
				id: resolvedModel.id
			};
			if (resultBundle) return {
				data: await this.runWithResultTools(args.promptText, resultBundle, beforeLength, args.source, args.errorLabel, args.signal, args.images),
				usage: this.aggregateUsageSince(beforeLeafId),
				model
			};
			await this.harness.prompt(args.promptText, args.images);
			await this.harness.waitForIdle();
			await this.syncHarnessMessagesSince(beforeLength, args.source, args.dispatch);
			await this.checkLatestAssistantForCompaction();
			this.throwIfError(args.errorLabel);
			return {
				text: this.getAssistantText(),
				usage: this.aggregateUsageSince(beforeLeafId),
				model
			};
		});
	}
	/**
	* Drive the harness through one or more turns until the LLM either calls
	* the `finish` tool (success) or the `give_up` tool (typed error).
	*
	* If a turn ends with neither tool called, we send a brief reminder and
	* loop. There is no retry cap from the framework's perspective: the model has a
	* clear escape hatch via `give_up`, the user has cancellation via `signal`,
	* and pi-agent-core has its own iteration limits as the final ceiling.
	* `MAX_FOLLOWUPS` is a defense-in-depth ceiling against pathological loops.
	*
	* `beforeLength` is the harness-message-array length sampled by the caller
	* *before* the very first prompt; we keep advancing it across iterations so
	* `syncHarnessMessagesSince` only copies newly-produced messages each turn.
	*/
	async runWithResultTools(initialPrompt, bundle, beforeLength, source, errorLabel, signal, initialImages) {
		let nextPrompt = initialPrompt;
		let cursor = beforeLength;
		const MAX_FOLLOWUPS = 32;
		for (let attempt = 0; attempt <= MAX_FOLLOWUPS; attempt++) {
			if (signal.aborted) throw abortErrorFor(signal);
			await this.harness.prompt(nextPrompt, attempt === 0 ? initialImages : void 0);
			await this.harness.waitForIdle();
			await this.syncHarnessMessagesSince(cursor, source);
			cursor = this.harness.state.messages.length;
			await this.checkLatestAssistantForCompaction();
			this.throwIfError(errorLabel);
			const outcome = bundle.getOutcome();
			if (outcome.type === "finished") return outcome.value;
			if (outcome.type === "gave_up") throw new ResultUnavailableError(outcome.reason, this.getAssistantText());
			nextPrompt = buildResultFollowUpPrompt();
			source = "retry";
		}
		throw new ResultUnavailableError(`Agent did not call \`finish\` or \`give_up\` after 33 attempts.`, this.getAssistantText());
	}
};
function normalizePath(p) {
	const parts = p.split("/");
	const result = [];
	for (const part of parts) {
		if (part === "." || part === "") continue;
		if (part === "..") result.pop();
		else result.push(part);
	}
	return `/${result.join("/")}`;
}
async function deleteSessionTree(store, storageKey, seen = /* @__PURE__ */ new Set()) {
	if (seen.has(storageKey)) return;
	seen.add(storageKey);
	const data = await store.load(storageKey);
	const taskSessions = Array.isArray(data?.metadata?.taskSessions) ? data.metadata.taskSessions : [];
	for (const task of taskSessions) if (typeof task?.storageKey === "string") await deleteSessionTree(store, task.storageKey, seen);
	await store.delete(storageKey);
}
function getErrorMessage(error) {
	return error instanceof Error ? error.message : String(error);
}
function serializeError(error) {
	if (error instanceof Error) return {
		name: error.name,
		message: error.message
	};
	return error;
}
function normalizeLogAttributes(attributes) {
	if (!attributes) return void 0;
	if (!(attributes.error instanceof Error)) return attributes;
	return {
		...attributes,
		error: serializeError(attributes.error)
	};
}
function durationSince(start) {
	return start === void 0 ? 0 : Date.now() - start;
}
function usageFromResult(result) {
	if (typeof result !== "object" || result === null) return void 0;
	const usage = result.usage;
	return isPromptUsage(usage) ? usage : void 0;
}
function isPromptUsage(value) {
	return typeof value === "object" && value !== null && typeof value.input === "number" && typeof value.output === "number" && typeof value.totalTokens === "number";
}
function redactEnvValues(env) {
	return Object.fromEntries(Object.keys(env).map((key) => [key, "<redacted>"]));
}
//#endregion
//#region src/sandbox.ts
/**
* Sandbox adapters: wraps BashFactory or SandboxApi into SessionEnv.
*/
/** Adapt a SessionEnv to the public FlueFs surface. */
function createFlueFs(env) {
	return {
		readFile: (path) => env.readFile(path),
		readFileBuffer: (path) => env.readFileBuffer(path),
		writeFile: (path, content) => env.writeFile(path, content),
		stat: (path) => env.stat(path),
		readdir: (path) => env.readdir(path),
		exists: (path) => env.exists(path),
		mkdir: (path, options) => env.mkdir(path, options),
		rm: (path, options) => env.rm(path, options)
	};
}
function createCwdSessionEnv(parentEnv, cwd) {
	const scopedCwd = normalizePath(cwd);
	const resolvePath = (p) => {
		if (p.startsWith("/")) return normalizePath(p);
		if (scopedCwd === "/") return normalizePath(`/${p}`);
		return normalizePath(`${scopedCwd}/${p}`);
	};
	return {
		exec: (cmd, opts) => parentEnv.exec(cmd, {
			cwd: opts?.cwd ?? scopedCwd,
			env: opts?.env,
			timeout: opts?.timeout,
			signal: opts?.signal
		}),
		readFile: (p) => parentEnv.readFile(resolvePath(p)),
		readFileBuffer: (p) => parentEnv.readFileBuffer(resolvePath(p)),
		writeFile: (p, c) => parentEnv.writeFile(resolvePath(p), c),
		stat: (p) => parentEnv.stat(resolvePath(p)),
		readdir: (p) => parentEnv.readdir(resolvePath(p)),
		exists: (p) => parentEnv.exists(resolvePath(p)),
		mkdir: (p, o) => parentEnv.mkdir(resolvePath(p), o),
		rm: (p, o) => parentEnv.rm(resolvePath(p), o),
		cwd: scopedCwd,
		resolvePath
	};
}
async function bashFactoryToSessionEnv(factory) {
	const bash = await factory();
	assertBashLike(bash);
	return createBashSessionEnv(bash);
}
function createBashSessionEnv(bash) {
	const fs = bash.fs;
	const cwd = bash.getCwd();
	const resolve = (p) => p.startsWith("/") ? p : fs.resolvePath(cwd, p);
	return {
		exec: async (cmd, opts) => {
			const exec = bash.exec;
			const timeoutSignal = typeof opts?.timeout === "number" ? AbortSignal.timeout(opts.timeout * 1e3) : void 0;
			const mergedSignal = opts?.signal && timeoutSignal ? AbortSignal.any([opts.signal, timeoutSignal]) : opts?.signal ?? timeoutSignal;
			const result = await exec.call(bash, cmd, opts ? {
				cwd: opts.cwd,
				env: opts.env,
				signal: mergedSignal
			} : void 0);
			if (opts?.signal?.aborted) throw abortErrorFor(opts.signal);
			return result;
		},
		readFile: (p) => fs.readFile(resolve(p)),
		readFileBuffer: (p) => fs.readFileBuffer(resolve(p)),
		writeFile: async (p, content) => {
			const resolved = resolve(p);
			const dir = resolved.replace(/\/[^/]*$/, "");
			if (dir && dir !== resolved) try {
				await fs.mkdir(dir, { recursive: true });
			} catch {}
			await fs.writeFile(resolved, content);
		},
		stat: (p) => fs.stat(resolve(p)),
		readdir: (p) => fs.readdir(resolve(p)),
		exists: (p) => fs.exists(resolve(p)),
		mkdir: (p, o) => fs.mkdir(resolve(p), o),
		rm: (p, o) => fs.rm(resolve(p), o),
		cwd,
		resolvePath: resolve
	};
}
/**
* Duck-type detection for just-bash Bash instances. Shared with client.ts
* so the predicate doesn't drift between the throwing and boolean variants.
*/
function isBashLike(value) {
	return typeof value === "object" && value !== null && "exec" in value && "getCwd" in value && "fs" in value && typeof value.exec === "function" && typeof value.getCwd === "function" && typeof value.fs === "object" && value.fs !== null;
}
function assertBashLike(value) {
	if (!isBashLike(value)) throw new Error("[flue] BashFactory must return a Bash-like object.");
}
/** Wrap a SandboxApi into SessionEnv. No just-bash, no intermediate filesystem layer. */
function createSandboxSessionEnv(api, cwd) {
	const resolvePath = (p) => {
		if (p.startsWith("/")) return normalizePath(p);
		if (cwd === "/") return normalizePath(`/${p}`);
		return normalizePath(`${cwd}/${p}`);
	};
	return {
		async exec(command, options) {
			const signal = options?.signal;
			if (signal?.aborted) throw abortErrorFor(signal);
			const result = await api.exec(command, {
				cwd: options?.cwd ?? cwd,
				env: options?.env,
				timeout: options?.timeout,
				signal
			});
			if (signal?.aborted) throw abortErrorFor(signal);
			return result;
		},
		async readFile(path) {
			return api.readFile(resolvePath(path));
		},
		async readFileBuffer(path) {
			return api.readFileBuffer(resolvePath(path));
		},
		async writeFile(path, content) {
			return api.writeFile(resolvePath(path), content);
		},
		async stat(path) {
			return api.stat(resolvePath(path));
		},
		async readdir(path) {
			return api.readdir(resolvePath(path));
		},
		async exists(path) {
			return api.exists(resolvePath(path));
		},
		async mkdir(path, options) {
			return api.mkdir(resolvePath(path), options);
		},
		async rm(path, options) {
			return api.rm(resolvePath(path), options);
		},
		cwd,
		resolvePath
	};
}
//#endregion
export { resolveRegisteredModel as _, isBashLike as a, parseSkillMarkdown as b, deleteSessionTree as c, getModelBinding as d, getProviderConfiguration as f, resetProvidersForTests as g, registerProvider as h, createSandboxSessionEnv as i, discoverSessionContext as l, registerApiProvider$1 as m, createCwdSessionEnv as n, InMemorySessionStore as o, hasRegisteredProvider as p, createFlueFs as r, Session as s, bashFactoryToSessionEnv as t, configureProvider as u, CLOUDFLARE_AI_BINDING_API as v, formatBashResult as x, ResultUnavailableError as y };
