// @flue/runtime@0.9.2/dist/websocket-protocol-BSIpcV6C.mjs
import { C as toPublicError, m as InvalidRequestError } from "./handle-agent-C2C75lpq.mjs";
//#region src/runtime/websocket-protocol.ts
function parseAgentWebSocketMessage(raw) {
	const value = parseObject(raw);
	if (value.version !== 1 || value.type !== "prompt" && value.type !== "ping") throw new InvalidRequestError({ reason: "Agent WebSocket messages must use protocol version 1 and type \"prompt\" or \"ping\"." });
	if (value.type === "ping") {
		if (value.requestId !== void 0 && !isNonBlankString(value.requestId)) throw new InvalidRequestError({ reason: "Agent WebSocket ping requestId must be a string when provided." });
		return {
			version: 1,
			type: "ping",
			requestId: value.requestId
		};
	}
	if (!isNonBlankString(value.requestId) || typeof value.message !== "string") throw new InvalidRequestError({ reason: "Agent WebSocket prompt messages require string requestId and message values." });
	if (value.session !== void 0 && !isNonBlankString(value.session)) throw new InvalidRequestError({ reason: "Agent WebSocket prompt session must be a non-empty string when provided." });
	return {
		version: 1,
		type: "prompt",
		requestId: value.requestId,
		message: value.message,
		session: value.session
	};
}
function parseWorkflowWebSocketMessage(raw) {
	const value = parseObject(raw);
	if (value.version !== 1 || value.type !== "invoke" || !isNonBlankString(value.requestId)) throw new InvalidRequestError({ reason: "Workflow WebSocket messages require protocol version 1, type \"invoke\", and a string requestId." });
	return {
		version: 1,
		type: "invoke",
		requestId: value.requestId,
		payload: value.payload === void 0 ? {} : value.payload
	};
}
function createWebSocketErrorMessage(error, requestId, runId) {
	return runId === void 0 ? {
		version: 1,
		type: "error",
		requestId,
		error: toPublicError(error)
	} : {
		version: 1,
		type: "error",
		requestId,
		runId,
		error: toPublicError(error)
	};
}
function isNonBlankString(value) {
	return typeof value === "string" && value.trim() !== "";
}
function parseObject(raw) {
	let value;
	try {
		value = JSON.parse(raw);
	} catch {
		throw new InvalidRequestError({ reason: "WebSocket messages must be valid JSON objects." });
	}
	if (!value || typeof value !== "object" || Array.isArray(value)) throw new InvalidRequestError({ reason: "WebSocket messages must be valid JSON objects." });
	return value;
}
//#endregion
export { parseAgentWebSocketMessage as n, parseWorkflowWebSocketMessage as r, createWebSocketErrorMessage as t };
